{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e3d05ea9-1192-42c3-a418-2e04f969f707",
   "metadata": {},
   "source": [
    "<H1>Sulfite Oxidase Deficiency SUOX</H1>\n",
    "<P>Data from <a href=\"https://pubmed.ncbi.nlm.nih.gov/36303223/\" target=\"__blank\">Li JT, et al. (2022) Mutation analysis of SUOX in isolated sulfite oxidase deficiency with ectopia lentis as the presenting feature: insights into genotype-phenotype correlation. Orphanet J Rare Dis.17(1):392. PMID:36303223</a>.</P>\n",
    "<P>We transferred information from Additional Files 5, 6, and 7 to two Excel files to parse the data.</P>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4bb11342-95e1-42e6-82a2-cb91d853461e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pyphetools version 0.7.4\n"
     ]
    }
   ],
   "source": [
    "import phenopackets as php\n",
    "from google.protobuf.json_format import MessageToDict, MessageToJson\n",
    "from google.protobuf.json_format import Parse, ParseDict\n",
    "import pandas as pd\n",
    "import math\n",
    "from csv import DictReader\n",
    "pd.set_option('display.max_colwidth', None) # show entire column contents, important!\n",
    "from collections import defaultdict\n",
    "import re\n",
    "import pyphetools\n",
    "from pyphetools.creation import *\n",
    "from pyphetools.visualization import PhenopacketTable\n",
    "print(f\"pyphetools version {pyphetools.__version__}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ac59ce43-c3ba-4b45-a215-2cc651a9b6cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = HpoParser()\n",
    "hpo_cr = parser.get_hpo_concept_recognizer()\n",
    "hpo_version = parser.get_version()\n",
    "PMID = \"PMID:36303223\"\n",
    "title = \"Mutation analysis of SUOX in isolated sulfite oxidase deficiency with ectopia lentis as the presenting feature: insights into genotype-phenotype correlation\"\n",
    "metadata = MetaData(created_by=\"ORCID:0000-0003-2598-6622\", pmid=PMID, pubmed_title=title)\n",
    "metadata.default_versions_with_hpo(version=hpo_version)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96a05018-e6b1-4a64-bd69-dedc2c24b4f0",
   "metadata": {},
   "source": [
    "<H2>SUOX variants</H2>\n",
    "<P>The file Li-SUOX-Variants.xlsx has one variant per line, assigned to the patient ID.</P>\n",
    "<p>Note that one of the reported variants is erroneous according to Variant Validator:</p>\n",
    "<pre>NM_001032386.2:c.1355C>A: Variant reference (C) does not agree with reference sequence (G)</pre>\n",
    "<p>This is how the variant was reported in the original publication. We changed the C to a G and obtain the\n",
    "same amino acid change as reported in the original publication: <tt>NP_001027558.1:p.(G452D)</tt>.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f352f20b-09bc-45f9-8cab-ec508059bc7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Proband ID</th>\n",
       "      <th>ID</th>\n",
       "      <th>Nucleotide</th>\n",
       "      <th>Amino acid</th>\n",
       "      <th>Exon</th>\n",
       "      <th>Domain</th>\n",
       "      <th>Status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>M1</td>\n",
       "      <td>c.433delC</td>\n",
       "      <td>p.Q145Sfs*16</td>\n",
       "      <td>EX6</td>\n",
       "      <td>Cyt-b5 domain</td>\n",
       "      <td>Homo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>M2</td>\n",
       "      <td>c.650G&gt;A</td>\n",
       "      <td>p.R217Q</td>\n",
       "      <td>EX6</td>\n",
       "      <td>Moco domain</td>\n",
       "      <td>Homo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>M3</td>\n",
       "      <td>c.794C&gt;A</td>\n",
       "      <td>p.A265D</td>\n",
       "      <td>EX6</td>\n",
       "      <td>Moco domain</td>\n",
       "      <td>Com het</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>M4</td>\n",
       "      <td>c.1280C&gt;A</td>\n",
       "      <td>p.S427*</td>\n",
       "      <td>EX6</td>\n",
       "      <td>Homodimerization domain</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>M5</td>\n",
       "      <td>c.733_736delCTTT</td>\n",
       "      <td>p.L245Pfs*27</td>\n",
       "      <td>EX6</td>\n",
       "      <td>Moco domain</td>\n",
       "      <td>Homo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Proband ID  ID        Nucleotide    Amino acid Exon  \\\n",
       "0           1  M1         c.433delC  p.Q145Sfs*16  EX6   \n",
       "1           2  M2          c.650G>A       p.R217Q  EX6   \n",
       "2           3  M3          c.794C>A       p.A265D  EX6   \n",
       "3           3  M4         c.1280C>A       p.S427*  EX6   \n",
       "4           4  M5  c.733_736delCTTT  p.L245Pfs*27  EX6   \n",
       "\n",
       "                    Domain   Status  \n",
       "0            Cyt-b5 domain     Homo  \n",
       "1              Moco domain     Homo  \n",
       "2              Moco domain  Com het  \n",
       "3  Homodimerization domain      NaN  \n",
       "4              Moco domain     Homo  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_df = pd.read_excel('input/Li-SUOX-Variants.xlsx', na_values=['n.a.'])\n",
    "variant_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "ef99c452-08b0-46b9-9d96-5a1f932c0f2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "variant validator for c.1280C>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.1280C>A/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.1201A>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.1201A>G/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.649C>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.649C>G/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.1136A>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.1136A>G/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.1348T>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.1348T>C/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.1355G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.1355G>A/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.1376G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.1376G>A/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.1521_1524delTTGT\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.1521_1524delTTGT/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.284_285insC\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.284_285insC/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.520delG\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.520delG/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.182T>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.182T>C/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.1084G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.1084G>A/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.1261C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.1261C>T/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.1187A>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.1187A>G/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.475G>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.475G>T/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.400_403delGAGC\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.400_403delGAGC/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.1382A>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.1382A>T/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.733_736delCTTT\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.733_736delCTTT/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.1126C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.1126C>T/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.433delC\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.433delC/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.794C>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.794C>A/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.205G>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.205G>C/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.1096C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.1096C>T/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.1097G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.1097G>A/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.884G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.884G>A/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.772A>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.772A>C/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.803G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.803G>A/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.1200C>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.1200C>G/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.352C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.352C>T/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.650G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.650G>A/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.1549_1574dup\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.1549_1574dup/NM_001032386.2?content-type=application%2Fjson\n",
      "variant validator for c.427C>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032386.2%3Ac.427C>A/NM_001032386.2?content-type=application%2Fjson\n"
     ]
    }
   ],
   "source": [
    "hg38 = \"hg38\"\n",
    "SUOX_transcript = 'NM_001032386.2'\n",
    "vvalidator = VariantValidator(genome_build=hg38, transcript=SUOX_transcript)\n",
    "patient_id_to_variant_list_d = defaultdict(list)\n",
    "all_variants = set()\n",
    "for _, row in variant_df.iterrows():\n",
    "    proband = row['Proband ID']\n",
    "    var = row['Nucleotide']\n",
    "    if var == \"c.1355C>A\":\n",
    "        var = \"c.1355G>A\" # repair error (see above)\n",
    "    patient_id_to_variant_list_d[proband].append(var)\n",
    "    all_variants.add(var)\n",
    "variant_d = {}\n",
    "for v in all_variants:\n",
    "    print(f\"variant validator for {v}\")\n",
    "    var = vvalidator.encode_hgvs(v)\n",
    "    variant_d[v] = var"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "888490de-8920-43eb-8e9f-10d04d84181a",
   "metadata": {},
   "source": [
    "<H1>Clinical data</H1>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "339ca23e-ebcb-4dcb-8aee-c44fce6280e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Proband ID</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Parental consanguity</th>\n",
       "      <th>Age at onset (months)</th>\n",
       "      <th>Variants</th>\n",
       "      <th>Amino acid</th>\n",
       "      <th>status</th>\n",
       "      <th>Typical type/Mild type</th>\n",
       "      <th>...</th>\n",
       "      <th>Homocys (umol/L) NR: 5-15</th>\n",
       "      <th>Cys (umol/L) NR: 20-70</th>\n",
       "      <th>UA (umol/L) NR: 210-430</th>\n",
       "      <th>Sulfite (mg/L) NR: 0</th>\n",
       "      <th>Thiosulfate NR: 0</th>\n",
       "      <th>Urine SSC (umol/mmolCr) NR: 0.1-10</th>\n",
       "      <th>Urine Taurine (mmol/molCr) NR: 12-150</th>\n",
       "      <th>Urine XA NR: 0-0.46mmol/L or &lt;40umol/mmolCr or &lt;0.29XA/Cr</th>\n",
       "      <th>Urine HypoXA NR: 0-0.18mmol/L or &lt;8umol/mmolCr or &lt;0.5HypoXA/Cr</th>\n",
       "      <th>Urine UA NR: 0.44-4.50mmol/L or 50-980umol/mmolCr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>9050047</td>\n",
       "      <td>EUR</td>\n",
       "      <td>M</td>\n",
       "      <td>No</td>\n",
       "      <td>0</td>\n",
       "      <td>c.(433delC); (433delC)</td>\n",
       "      <td>p.(Q145Sfs*16); (Q145Sfs*16)</td>\n",
       "      <td>Homo</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>2</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>20-25</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>320umol/L</td>\n",
       "      <td>95</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>9600976</td>\n",
       "      <td>EUR</td>\n",
       "      <td>F</td>\n",
       "      <td>Yes</td>\n",
       "      <td>5</td>\n",
       "      <td>c.(650G&gt;A); (650G&gt;A)</td>\n",
       "      <td>p.(R217Q); (R217Q)</td>\n",
       "      <td>Homo</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>0.108-0.211</td>\n",
       "      <td>0.297-1.632mmol/L</td>\n",
       "      <td>240umol/L</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>0.04mmol/L</td>\n",
       "      <td>0.05mmol/L</td>\n",
       "      <td>0.14mmol/L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>10519592</td>\n",
       "      <td>NAM</td>\n",
       "      <td>M</td>\n",
       "      <td>No</td>\n",
       "      <td>0</td>\n",
       "      <td>c.(794C&gt;A); (1280C&gt;A)</td>\n",
       "      <td>p.(A265D); (S427*)</td>\n",
       "      <td>Com het</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>normal</td>\n",
       "      <td>80-100</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>690umol/L</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>normal</td>\n",
       "      <td>normal</td>\n",
       "      <td>normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>12112661</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Yes</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>c.(733_736delCTTT); (733_736delCTTT)</td>\n",
       "      <td>p.(L245Pfs*27); (L245Pfs*27)</td>\n",
       "      <td>Homo</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>12112661</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Yes</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>c.(284_285insC); (1126C&gt;T)</td>\n",
       "      <td>p.(E97*); (R376C)</td>\n",
       "      <td>Com het</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Proband ID      PMID Ethnicity Gender Parental consanguity  \\\n",
       "0           1   9050047       EUR      M                   No   \n",
       "1           2   9600976       EUR      F                  Yes   \n",
       "2           3  10519592       NAM      M                   No   \n",
       "3           4  12112661      n.a.   n.a.                  Yes   \n",
       "4           5  12112661      n.a.   n.a.                  Yes   \n",
       "\n",
       "  Age at onset (months)                              Variants  \\\n",
       "0                     0                c.(433delC); (433delC)   \n",
       "1                     5                  c.(650G>A); (650G>A)   \n",
       "2                     0                 c.(794C>A); (1280C>A)   \n",
       "3                  n.a.  c.(733_736delCTTT); (733_736delCTTT)   \n",
       "4                  n.a.            c.(284_285insC); (1126C>T)   \n",
       "\n",
       "                     Amino acid   status Typical type/Mild type  ...  \\\n",
       "0  p.(Q145Sfs*16); (Q145Sfs*16)     Homo                      T  ...   \n",
       "1            p.(R217Q); (R217Q)     Homo                      T  ...   \n",
       "2            p.(A265D); (S427*)  Com het                      T  ...   \n",
       "3  p.(L245Pfs*27); (L245Pfs*27)     Homo                      T  ...   \n",
       "4             p.(E97*); (R376C)  Com het                      T  ...   \n",
       "\n",
       "  Homocys (umol/L) NR: 5-15 Cys (umol/L) NR: 20-70 UA (umol/L) NR: 210-430  \\\n",
       "0                      n.a.                      2                    n.a.   \n",
       "1                      n.a.                   n.a.                    n.a.   \n",
       "2                      n.a.                   n.a.                  normal   \n",
       "3                      n.a.                   n.a.                    n.a.   \n",
       "4                      n.a.                   n.a.                    n.a.   \n",
       "\n",
       "  Sulfite (mg/L) NR: 0 Thiosulfate NR: 0  Urine SSC (umol/mmolCr) NR: 0.1-10  \\\n",
       "0                20-25               n.a.                          320umol/L   \n",
       "1          0.108-0.211  0.297-1.632mmol/L                          240umol/L   \n",
       "2               80-100               n.a.                          690umol/L   \n",
       "3                 n.a.               n.a.                               n.a.   \n",
       "4                 n.a.               n.a.                               n.a.   \n",
       "\n",
       "  Urine Taurine (mmol/molCr) NR: 12-150  \\\n",
       "0                                    95   \n",
       "1                                  n.a.   \n",
       "2                                  n.a.   \n",
       "3                                  n.a.   \n",
       "4                                  n.a.   \n",
       "\n",
       "  Urine XA NR: 0-0.46mmol/L or <40umol/mmolCr or <0.29XA/Cr  \\\n",
       "0                                                      n.a.   \n",
       "1                                                0.04mmol/L   \n",
       "2                                                    normal   \n",
       "3                                                      n.a.   \n",
       "4                                                      n.a.   \n",
       "\n",
       "  Urine HypoXA NR: 0-0.18mmol/L or <8umol/mmolCr or <0.5HypoXA/Cr  \\\n",
       "0                                                            n.a.   \n",
       "1                                                      0.05mmol/L   \n",
       "2                                                          normal   \n",
       "3                                                            n.a.   \n",
       "4                                                            n.a.   \n",
       "\n",
       "  Urine UA NR: 0.44-4.50mmol/L or 50-980umol/mmolCr  \n",
       "0                                              n.a.  \n",
       "1                                        0.14mmol/L  \n",
       "2                                            normal  \n",
       "3                                              n.a.  \n",
       "4                                              n.a.  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clinical = pd.read_excel(\"input/Li-Suox-Clinical.xlsx\")\n",
    "df_clinical.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "03e08fcd-12dc-4edb-9023-0b36642c5ec0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Proband ID', 'PMID', 'Ethnicity', 'Gender', 'Parental consanguity',\n",
       "       'Age at onset (months)', 'Variants', 'Amino acid', 'status',\n",
       "       'Typical type/Mild type', 'Death', 'Age of death (months)',\n",
       "       'Prodromal infection', 'Developmental delay', 'Regression', 'Seizure',\n",
       "       'Extrapyramidal symptoms', 'Hypertonia', 'Hypotonia', 'Microcephaly',\n",
       "       'Ectopia lentis', 'Age of diagnosis of ophthalmic manifestations',\n",
       "       'Proband ID.1', 'Resource (PMID)', 'SSC (umol/L) NR: 0',\n",
       "       'Taurine (umol/L) NR: 15-145', 'Homocys (umol/L) NR: 5-15',\n",
       "       'Cys (umol/L) NR: 20-70', 'UA (umol/L) NR: 210-430',\n",
       "       'Sulfite (mg/L) NR: 0', 'Thiosulfate NR: 0 ',\n",
       "       'Urine SSC (umol/mmolCr) NR: 0.1-10',\n",
       "       'Urine Taurine (mmol/molCr) NR: 12-150',\n",
       "       'Urine XA NR: 0-0.46mmol/L or <40umol/mmolCr or <0.29XA/Cr',\n",
       "       'Urine HypoXA NR: 0-0.18mmol/L or <8umol/mmolCr or <0.5HypoXA/Cr',\n",
       "       'Urine UA NR: 0.44-4.50mmol/L or 50-980umol/mmolCr'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clinical.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "4e2ac2d4-62f9-4671-a29b-12e753bf2286",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We created 8 simple column mappers\n"
     ]
    }
   ],
   "source": [
    "column_mapper_d = {}\n",
    "\n",
    "items = {\n",
    "    'Neurological manifestations_Developmental delay': ['Neurodevelopmental delay', 'HP:0012758'],\n",
    "    'Neurological manifestations_Regression': ['Cognitive regression', 'HP:0034332'],\n",
    "    'Neurological manifestations_Seizure': ['Seizure', 'HP:0001250'],\n",
    "    'Neurological manifestations_Extrapyramidal symptoms': ['Abnormality of extrapyramidal motor function', 'HP:0002071'],\n",
    "    'Neurological manifestations_Hypertonia':['Hypertonia','HP:0001276'],\n",
    "    'Neurological manifestations_Hypotonia': ['Hypotonia','HP:0001252'],\n",
    "    'Neurological manifestations_Microcephaly':['Microcephaly', 'HP:0000252'],\n",
    "    'Opthalmic manifestations_Ectopia lentis':['Ectopia lentis', 'HP:0001083'],\n",
    "}\n",
    "\n",
    "item_column_mapper_d = hpo_cr.initialize_simple_column_maps(column_name_to_hpo_label_map=items, \n",
    "                                                            observed='+',\n",
    "                                                            excluded='-')\n",
    "print(f\"We created {len(item_column_mapper_d)} simple column mappers\")\n",
    "# Transfor to column_mapper_d\n",
    "for k, v in item_column_mapper_d.items():\n",
    "    column_mapper_d[k] = v"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c24292e2-b9e6-4e3e-8014-aed58554bdb2",
   "metadata": {},
   "source": [
    "<H2>Threshold mappers</H2>\n",
    "<p>The data contain information about biochemical abnormalities framed as tests with reference ranges and values. We can capture this using threshold mappers</p>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "216aaca5-3cb0-495c-ab24-bf0d64980493",
   "metadata": {},
   "source": [
    "<h3>SSC (umol/L) NR: 0</h3>\n",
    "<p>SSC refers to S-sulfocysteine; The normal range is absent (not more than zero). The corresponding\n",
    "HPO term is Elevated circulating S-sulfocysteine concentration HP:0034745.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "51105641-67bc-4521-b73c-ceadd44e7b56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['n.a.', 28, 14], dtype=object)"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clinical['SSC (umol/L) NR: 0'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "1dd7f30b-0275-431d-be7f-26744a16b2f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "sscMapper = ThresholdedColumnMapper(hpo_id=\"HP:0034745\", \n",
    "                                    hpo_label=\"Elevated circulating S-sulfocysteine concentration\",\n",
    "                                    threshold=0,\n",
    "                                   call_if_above=True)\n",
    "sscMapper.preview_column(df_clinical['SSC (umol/L) NR: 0'])\n",
    "column_mapper_d['SSC (umol/L) NR: 0'] = sscMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "f2a8ef9e-2750-439a-8c8c-de16be4f5d19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([197, 'n.a.', 46], dtype=object)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clinical['Taurine (umol/L) NR: 15-145'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "4c7472bb-4808-47d2-a911-aec7b1e63b17",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Hypertaurinemia HP:0500181\n",
    "taurineMapper = ThresholdedColumnMapper(hpo_id=\"HP:0500181\",\n",
    "                                        hpo_label=\"Hypertaurinemia\",\n",
    "                                        threshold=145,\n",
    "                                        call_if_above=True)\n",
    "taurineMapper.preview_column(df_clinical['Taurine (umol/L) NR: 15-145'])\n",
    "column_mapper_d['Taurine (umol/L) NR: 15-145'] = taurineMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "987feeac-d3d3-4ba3-b5db-046f84b1a629",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 'Homocys (umol/L) NR: 5-15' -- Hyperhomocystinemia HP:0002160\n",
    "homocyteineMapper = ThresholdedColumnMapper(hpo_id=\"HP:0002160\",\n",
    "                                            hpo_label=\"Hyperhomocystinemia\",\n",
    "                                            threshold=15,\n",
    "                                            call_if_above=True)\n",
    "homocyteineMapper.preview_column(df_clinical['Homocys (umol/L) NR: 5-15'])\n",
    "column_mapper_d['Homocys (umol/L) NR: 5-15'] = homocyteineMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "95d8a51e-5f41-4f24-9f08-597f8b7a07be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 'Cys (umol/L) NR: 20-70' -- \n",
    "# Note this manifests as low circulate Cystine (not Cysteine)\n",
    "# Hypocystinemia HP:0500152\n",
    "\n",
    "cystineMapper = ThresholdedColumnMapper(hpo_id=\"HP:0500152\",\n",
    "                                        hpo_label=\"Hypocystinemia\",\n",
    "                                        threshold=20,\n",
    "                                        call_if_above=False)\n",
    "cystineMapper.preview_column(df_clinical['Cys (umol/L) NR: 20-70'])\n",
    "column_mapper_d['Cys (umol/L) NR: 20-70'] = cystineMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "89268985-bd47-4c46-8e18-f5febfe53b64",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 'UA (umol/L) NR: 210-430'  -- Hypouricemia HP:0003537\n",
    "uricAcidMapper = ThresholdedColumnMapper(hpo_id=\"HP:0003537\",\n",
    "                                         hpo_label=\"Hypouricemia\",\n",
    "                                         threshold=210,\n",
    "                                         call_if_above=False)\n",
    "uricAcidMapper.preview_column(df_clinical['UA (umol/L) NR: 210-430'])\n",
    "column_mapper_d['UA (umol/L) NR: 210-430'] = uricAcidMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "d0871aa9-eb8d-4def-a1aa-f4bc6ceab017",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 'Sulfite (mg/L) NR: 0' \n",
    "# df_clinical['Sulfite (mg/L) NR: 0']\n",
    "# requires new HPO term"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "9fbd8f6a-b715-4d10-8492-a50a5290d11d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 'Thiosulfate NR: 0 ' -- requires new HPO term"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "7b4f4501-5180-4f97-98c3-2403bd610291",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 'Urine SSC (umol/mmolCr) NR: 0.1-10' -- Sulfocysteinuria HP:0032350\n",
    "urineSscMapper = ThresholdedColumnMapper(hpo_id=\"HP:0032350\",\n",
    "                                         hpo_label=\"Sulfocysteinuria\",\n",
    "                                        threshold=10,\n",
    "                                        call_if_above=True)\n",
    "urineSscMapper.preview_column(df_clinical['Urine SSC (umol/mmolCr) NR: 0.1-10'])\n",
    "column_mapper_d['Urine SSC (umol/mmolCr) NR: 0.1-10'] = urineSscMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "5ce627da-72c2-47f6-9c78-da573d65e030",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 'Urine Taurine (mmol/molCr) NR: 12-150'  -- Increased urinary taurine HP:0003166\n",
    "urineTaurineMapper = ThresholdedColumnMapper(hpo_id=\"HP:0003166\",\n",
    "                                            hpo_label=\"Increased urinary taurine\",\n",
    "                                            threshold=150,\n",
    "                                            call_if_above=True)\n",
    "urineTaurineMapper.preview_column(df_clinical['Urine Taurine (mmol/molCr) NR: 12-150'])\n",
    "column_mapper_d['Urine Taurine (mmol/molCr) NR: 12-150'] = urineTaurineMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "ce135a0e-c1f4-411e-b5eb-85273bfeff24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>terms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HP:0010934 (Xanthinuria/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HP:0010934 (Xanthinuria/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>HP:0010934 (Xanthinuria/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>HP:0010934 (Xanthinuria/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>HP:0010934 (Xanthinuria/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>HP:0010934 (Xanthinuria/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>HP:0010934 (Xanthinuria/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>HP:0010934 (Xanthinuria/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>HP:0010934 (Xanthinuria/excluded)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                terms\n",
       "0                                 n/a\n",
       "1   HP:0010934 (Xanthinuria/excluded)\n",
       "2   HP:0010934 (Xanthinuria/excluded)\n",
       "3                                 n/a\n",
       "4                                 n/a\n",
       "5                                 n/a\n",
       "6                                 n/a\n",
       "7                                 n/a\n",
       "8                                 n/a\n",
       "9                                 n/a\n",
       "10                                n/a\n",
       "11                                n/a\n",
       "12                                n/a\n",
       "13                                n/a\n",
       "14                                n/a\n",
       "15                                n/a\n",
       "16                                n/a\n",
       "17                                n/a\n",
       "18  HP:0010934 (Xanthinuria/excluded)\n",
       "19                                n/a\n",
       "20                                n/a\n",
       "21  HP:0010934 (Xanthinuria/excluded)\n",
       "22                                n/a\n",
       "23                                n/a\n",
       "24                                n/a\n",
       "25                                n/a\n",
       "26                                n/a\n",
       "27  HP:0010934 (Xanthinuria/excluded)\n",
       "28  HP:0010934 (Xanthinuria/excluded)\n",
       "29  HP:0010934 (Xanthinuria/excluded)\n",
       "30  HP:0010934 (Xanthinuria/excluded)\n",
       "31                                n/a\n",
       "32                                n/a\n",
       "33                                n/a\n",
       "34  HP:0010934 (Xanthinuria/excluded)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 'Urine XA NR: 0-0.46mmol/L or <40umol/mmolCr or <0.29XA/Cr'\n",
    "# Here we need to use an OptionColumnMapper because three different measurement ranges are used\n",
    "df_clinical['Urine XA NR: 0-0.46mmol/L or <40umol/mmolCr or <0.29XA/Cr'].unique()\n",
    "urine_xa_d = { '11.7umol/mmolCr':\"Xanthinuria\",\n",
    "       '1.7umol/mmolCr':\"Xanthinuria\" }\n",
    "urine_not_xa_d = {'0.04mmol/L': \"Xanthinuria\", \n",
    "                  \"1.6umol/mmolCr\": \"Xanthinuria\", \n",
    "                  \"0.0214XA/Cr\": \"Xanthinuria\",\n",
    "                 \"normal\": \"Xanthinuria\"}\n",
    "urineXAmapper = OptionColumnMapper(concept_recognizer=hpo_cr, \n",
    "                                   option_d=urine_xa_d, \n",
    "                                   excluded_d=urine_not_xa_d)\n",
    "urineXAmapper.preview_column(df_clinical['Urine XA NR: 0-0.46mmol/L or <40umol/mmolCr or <0.29XA/Cr'])\n",
    "\n",
    "#df_clinical['Urine XA NR: 0-0.46mmol/L or <40umol/mmolCr or <0.29XA/Cr'].unique()\n",
    "# TODO CHECK ME -- thereshould be saome observed terms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "daeec93a-df9d-41be-a64a-d5c7f1509972",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>terms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HP:0011814 (Increased urinary hypoxanthine level/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HP:0011814 (Increased urinary hypoxanthine level/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>HP:0011814 (Increased urinary hypoxanthine level/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>HP:0011814 (Increased urinary hypoxanthine level/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>HP:0011814 (Increased urinary hypoxanthine level/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>HP:0011814 (Increased urinary hypoxanthine level/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>HP:0011814 (Increased urinary hypoxanthine level/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>HP:0011814 (Increased urinary hypoxanthine level/excluded)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         terms\n",
       "0                                                          n/a\n",
       "1   HP:0011814 (Increased urinary hypoxanthine level/excluded)\n",
       "2   HP:0011814 (Increased urinary hypoxanthine level/excluded)\n",
       "3                                                          n/a\n",
       "4                                                          n/a\n",
       "5                                                          n/a\n",
       "6                                                          n/a\n",
       "7                                                          n/a\n",
       "8                                                          n/a\n",
       "9                                                          n/a\n",
       "10                                                         n/a\n",
       "11                                                         n/a\n",
       "12                                                         n/a\n",
       "13                                                         n/a\n",
       "14                                                         n/a\n",
       "15                                                         n/a\n",
       "16                                                         n/a\n",
       "17                                                         n/a\n",
       "18  HP:0011814 (Increased urinary hypoxanthine level/excluded)\n",
       "19                                                         n/a\n",
       "20                                                         n/a\n",
       "21                                                         n/a\n",
       "22                                                         n/a\n",
       "23                                                         n/a\n",
       "24                                                         n/a\n",
       "25                                                         n/a\n",
       "26                                                         n/a\n",
       "27  HP:0011814 (Increased urinary hypoxanthine level/excluded)\n",
       "28  HP:0011814 (Increased urinary hypoxanthine level/excluded)\n",
       "29  HP:0011814 (Increased urinary hypoxanthine level/excluded)\n",
       "30  HP:0011814 (Increased urinary hypoxanthine level/excluded)\n",
       "31                                                         n/a\n",
       "32                                                         n/a\n",
       "33                                                         n/a\n",
       "34  HP:0011814 (Increased urinary hypoxanthine level/excluded)"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 'Urine HypoXA NR: 0-0.18mmol/L or <8umol/mmolCr or <0.5HypoXA/Cr',\n",
    "# Increased urinary hypoxanthine level HP:0011814\n",
    "urine_hxa_d = {\n",
    "    '8umol/mmolCr': \"Increased urinary hypoxanthine level\",\n",
    "}\n",
    "urine_hxa_excluded_d = {\n",
    "    'normal': \"Increased urinary hypoxanthine level\",\n",
    "    '0.05mmol/L': \"Increased urinary hypoxanthine level\",\n",
    "    '0.0264HypoXA/Cr': \"Increased urinary hypoxanthine level\",\n",
    "}\n",
    "urineHXAmapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=urine_hxa_d, excluded_d=urine_hxa_excluded_d)\n",
    "urineHXAmapper.preview_column(df_clinical['Urine HypoXA NR: 0-0.18mmol/L or <8umol/mmolCr or <0.5HypoXA/Cr'])\n",
    "\n",
    "# TODO CHECK"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "523a023f-cae7-4b04-b9cf-80520af4dc11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['n.a.', '0.14mmol/L', 'normal', '385umol/mmolCr', '21umol/mmolCr',\n",
       "       '430umol/mmolCr'], dtype=object)"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 'Urine UA NR: 0.44-4.50mmol/L or 50-980umol/mmolCr'\n",
    "# Hyperuricosuria HP:0003149\n",
    "# Decreased urinary urate HP:0011935\n",
    "# Abnormality of urinary uric acid level HP:0012610\n",
    "df_clinical['Urine UA NR: 0.44-4.50mmol/L or 50-980umol/mmolCr'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "15700414-70dc-4fac-afb0-a1f4ba554866",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>terms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HP:0012610 (Abnormality of urinary uric acid level/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HP:0012610 (Abnormality of urinary uric acid level/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>HP:0012610 (Abnormality of urinary uric acid level/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>HP:0012610 (Abnormality of urinary uric acid level/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>HP:0012610 (Abnormality of urinary uric acid level/excluded)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           terms\n",
       "0                                                            n/a\n",
       "1                                                            n/a\n",
       "2   HP:0012610 (Abnormality of urinary uric acid level/excluded)\n",
       "3                                                            n/a\n",
       "4                                                            n/a\n",
       "5                                                            n/a\n",
       "6                                                            n/a\n",
       "7                                                            n/a\n",
       "8                                                            n/a\n",
       "9                                                            n/a\n",
       "10                                                           n/a\n",
       "11                                                           n/a\n",
       "12                                                           n/a\n",
       "13  HP:0012610 (Abnormality of urinary uric acid level/excluded)\n",
       "14                                                           n/a\n",
       "15                                                           n/a\n",
       "16                                                           n/a\n",
       "17                                                           n/a\n",
       "18                                                           n/a\n",
       "19                                                           n/a\n",
       "20                                                           n/a\n",
       "21                                                           n/a\n",
       "22                                                           n/a\n",
       "23                                                           n/a\n",
       "24                                                           n/a\n",
       "25                                                           n/a\n",
       "26                                                           n/a\n",
       "27                                                           n/a\n",
       "28                                                           n/a\n",
       "29  HP:0012610 (Abnormality of urinary uric acid level/excluded)\n",
       "30  HP:0012610 (Abnormality of urinary uric acid level/excluded)\n",
       "31                                                           n/a\n",
       "32                                                           n/a\n",
       "33                                                           n/a\n",
       "34  HP:0012610 (Abnormality of urinary uric acid level/excluded)"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "urine_ua_d = {'0.14mmol/L': \"Decreased urinary urate\",\n",
    "             '21umol/mmolCr': \"Decreased urinary urate\",}\n",
    "urine_ua_excluded_d = {'normal' : \"Abnormality of urinary uric acid level\",\n",
    "                      '385umol/mmolCr': \"Abnormality of urinary uric acid level\",\n",
    "                      '430umol/mmolCr': \"Abnormality of urinary uric acid level\",}\n",
    "urineUaMapper = OptionColumnMapper(concept_recognizer=hpo_cr,\n",
    "                                  option_d=urine_ua_d,\n",
    "                                  excluded_d=urine_ua_excluded_d)\n",
    "urineUaMapper.preview_column(df_clinical['Urine UA NR: 0.44-4.50mmol/L or 50-980umol/mmolCr'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d3d623c-3e6f-427d-912a-541ea128435e",
   "metadata": {},
   "source": [
    "<H2>Putting it all together</H2>\n",
    "<p>First, let's create a new individual id column that shows the PMID</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "ae6282d2-d541-47a7-9581-7905102a7c19",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'Proband ID'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[90], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mdf_clinical\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43;01mlambda\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mrow\u001b[49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mindividual_\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mrow\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloc\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mProband ID\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/pstore_env/lib/python3.11/site-packages/pandas/core/frame.py:10037\u001b[0m, in \u001b[0;36mDataFrame.apply\u001b[0;34m(self, func, axis, raw, result_type, args, by_row, **kwargs)\u001b[0m\n\u001b[1;32m  10025\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mcore\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mapply\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m frame_apply\n\u001b[1;32m  10027\u001b[0m op \u001b[38;5;241m=\u001b[39m frame_apply(\n\u001b[1;32m  10028\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m  10029\u001b[0m     func\u001b[38;5;241m=\u001b[39mfunc,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m  10035\u001b[0m     kwargs\u001b[38;5;241m=\u001b[39mkwargs,\n\u001b[1;32m  10036\u001b[0m )\n\u001b[0;32m> 10037\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mop\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39m__finalize__(\u001b[38;5;28mself\u001b[39m, method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mapply\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/pstore_env/lib/python3.11/site-packages/pandas/core/apply.py:831\u001b[0m, in \u001b[0;36mFrameApply.apply\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    828\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mraw:\n\u001b[1;32m    829\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mapply_raw()\n\u001b[0;32m--> 831\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply_standard\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/pstore_env/lib/python3.11/site-packages/pandas/core/apply.py:957\u001b[0m, in \u001b[0;36mFrameApply.apply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    956\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mapply_standard\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m--> 957\u001b[0m     results, res_index \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply_series_generator\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    959\u001b[0m     \u001b[38;5;66;03m# wrap results\u001b[39;00m\n\u001b[1;32m    960\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mwrap_results(results, res_index)\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/pstore_env/lib/python3.11/site-packages/pandas/core/apply.py:973\u001b[0m, in \u001b[0;36mFrameApply.apply_series_generator\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    970\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m option_context(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmode.chained_assignment\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m    971\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m i, v \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(series_gen):\n\u001b[1;32m    972\u001b[0m         \u001b[38;5;66;03m# ignore SettingWithCopy here in case the user mutates\u001b[39;00m\n\u001b[0;32m--> 973\u001b[0m         results[i] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mv\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    974\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(results[i], ABCSeries):\n\u001b[1;32m    975\u001b[0m             \u001b[38;5;66;03m# If we have a view on v, we need to make a copy because\u001b[39;00m\n\u001b[1;32m    976\u001b[0m             \u001b[38;5;66;03m#  series_generator will swap out the underlying data\u001b[39;00m\n\u001b[1;32m    977\u001b[0m             results[i] \u001b[38;5;241m=\u001b[39m results[i]\u001b[38;5;241m.\u001b[39mcopy(deep\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "Cell \u001b[0;32mIn[90], line 1\u001b[0m, in \u001b[0;36m<lambda>\u001b[0;34m(row)\u001b[0m\n\u001b[0;32m----> 1\u001b[0m df_clinical\u001b[38;5;241m.\u001b[39mapply(\u001b[38;5;28;01mlambda\u001b[39;00m row: \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mindividual_\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[43mrow\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloc\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mProband ID\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/pstore_env/lib/python3.11/site-packages/pandas/core/indexing.py:1153\u001b[0m, in \u001b[0;36m_LocationIndexer.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1150\u001b[0m axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxis \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m0\u001b[39m\n\u001b[1;32m   1152\u001b[0m maybe_callable \u001b[38;5;241m=\u001b[39m com\u001b[38;5;241m.\u001b[39mapply_if_callable(key, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj)\n\u001b[0;32m-> 1153\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_axis\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmaybe_callable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/pstore_env/lib/python3.11/site-packages/pandas/core/indexing.py:1393\u001b[0m, in \u001b[0;36m_LocIndexer._getitem_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1391\u001b[0m \u001b[38;5;66;03m# fall thru to straight lookup\u001b[39;00m\n\u001b[1;32m   1392\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_key(key, axis)\n\u001b[0;32m-> 1393\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_label\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/pstore_env/lib/python3.11/site-packages/pandas/core/indexing.py:1343\u001b[0m, in \u001b[0;36m_LocIndexer._get_label\u001b[0;34m(self, label, axis)\u001b[0m\n\u001b[1;32m   1341\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_get_label\u001b[39m(\u001b[38;5;28mself\u001b[39m, label, axis: AxisInt):\n\u001b[1;32m   1342\u001b[0m     \u001b[38;5;66;03m# GH#5567 this will fail if the label is not present in the axis.\u001b[39;00m\n\u001b[0;32m-> 1343\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mxs\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlabel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/pstore_env/lib/python3.11/site-packages/pandas/core/generic.py:4236\u001b[0m, in \u001b[0;36mNDFrame.xs\u001b[0;34m(self, key, axis, level, drop_level)\u001b[0m\n\u001b[1;32m   4234\u001b[0m             new_index \u001b[38;5;241m=\u001b[39m index[loc]\n\u001b[1;32m   4235\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 4236\u001b[0m     loc \u001b[38;5;241m=\u001b[39m \u001b[43mindex\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4238\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(loc, np\u001b[38;5;241m.\u001b[39mndarray):\n\u001b[1;32m   4239\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m loc\u001b[38;5;241m.\u001b[39mdtype \u001b[38;5;241m==\u001b[39m np\u001b[38;5;241m.\u001b[39mbool_:\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/pstore_env/lib/python3.11/site-packages/pandas/core/indexes/range.py:418\u001b[0m, in \u001b[0;36mRangeIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    416\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m    417\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, Hashable):\n\u001b[0;32m--> 418\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key)\n\u001b[1;32m    419\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n\u001b[1;32m    420\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Proband ID'"
     ]
    }
   ],
   "source": [
    "df_clinical.apply(lambda row: f\"individual_{row.loc['Proband ID']}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "85819512-1e08-4d6f-a235-2815e92b42de",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_individual_id(arr):\n",
    "    iid = arr.iloc[0]\n",
    "    pmid = arr.iloc[1]\n",
    "    if pmid == \"our patient\":\n",
    "        return \"individual_35_PMID_36303223\" # from current manuscript\n",
    "    else:\n",
    "        return f\"individual_{iid}_PMID_{pmid}\"\n",
    "df_clinical['individual_id'] = df_clinical[['Proband ID', 'Resource (PMID)']].apply(lambda x: get_individual_id(x), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "89294943-d02b-4a61-aab2-404c531d3a26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Proband ID</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Parental consanguity</th>\n",
       "      <th>Age at onset (months)</th>\n",
       "      <th>Variants</th>\n",
       "      <th>Amino acid</th>\n",
       "      <th>status</th>\n",
       "      <th>Typical type/Mild type</th>\n",
       "      <th>...</th>\n",
       "      <th>Cys (umol/L) NR: 20-70</th>\n",
       "      <th>UA (umol/L) NR: 210-430</th>\n",
       "      <th>Sulfite (mg/L) NR: 0</th>\n",
       "      <th>Thiosulfate NR: 0</th>\n",
       "      <th>Urine SSC (umol/mmolCr) NR: 0.1-10</th>\n",
       "      <th>Urine Taurine (mmol/molCr) NR: 12-150</th>\n",
       "      <th>Urine XA NR: 0-0.46mmol/L or &lt;40umol/mmolCr or &lt;0.29XA/Cr</th>\n",
       "      <th>Urine HypoXA NR: 0-0.18mmol/L or &lt;8umol/mmolCr or &lt;0.5HypoXA/Cr</th>\n",
       "      <th>Urine UA NR: 0.44-4.50mmol/L or 50-980umol/mmolCr</th>\n",
       "      <th>individual_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>9050047</td>\n",
       "      <td>EUR</td>\n",
       "      <td>M</td>\n",
       "      <td>No</td>\n",
       "      <td>0</td>\n",
       "      <td>c.(433delC); (433delC)</td>\n",
       "      <td>p.(Q145Sfs*16); (Q145Sfs*16)</td>\n",
       "      <td>Homo</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>20-25</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>320umol/L</td>\n",
       "      <td>95</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>individual_1_PMID_9050047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>9600976</td>\n",
       "      <td>EUR</td>\n",
       "      <td>F</td>\n",
       "      <td>Yes</td>\n",
       "      <td>5</td>\n",
       "      <td>c.(650G&gt;A); (650G&gt;A)</td>\n",
       "      <td>p.(R217Q); (R217Q)</td>\n",
       "      <td>Homo</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>0.108-0.211</td>\n",
       "      <td>0.297-1.632mmol/L</td>\n",
       "      <td>240umol/L</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>0.04mmol/L</td>\n",
       "      <td>0.05mmol/L</td>\n",
       "      <td>0.14mmol/L</td>\n",
       "      <td>individual_2_PMID_9600976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>10519592</td>\n",
       "      <td>NAM</td>\n",
       "      <td>M</td>\n",
       "      <td>No</td>\n",
       "      <td>0</td>\n",
       "      <td>c.(794C&gt;A); (1280C&gt;A)</td>\n",
       "      <td>p.(A265D); (S427*)</td>\n",
       "      <td>Com het</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>normal</td>\n",
       "      <td>80-100</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>690umol/L</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>normal</td>\n",
       "      <td>normal</td>\n",
       "      <td>normal</td>\n",
       "      <td>individual_3_PMID_10519592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>12112661</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Yes</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>c.(733_736delCTTT); (733_736delCTTT)</td>\n",
       "      <td>p.(L245Pfs*27); (L245Pfs*27)</td>\n",
       "      <td>Homo</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>individual_4_PMID_12112661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>12112661</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>Yes</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>c.(284_285insC); (1126C&gt;T)</td>\n",
       "      <td>p.(E97*); (R376C)</td>\n",
       "      <td>Com het</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>n.a.</td>\n",
       "      <td>individual_5_PMID_12112661</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Proband ID      PMID Ethnicity Gender Parental consanguity  \\\n",
       "0           1   9050047       EUR      M                   No   \n",
       "1           2   9600976       EUR      F                  Yes   \n",
       "2           3  10519592       NAM      M                   No   \n",
       "3           4  12112661      n.a.   n.a.                  Yes   \n",
       "4           5  12112661      n.a.   n.a.                  Yes   \n",
       "\n",
       "  Age at onset (months)                              Variants  \\\n",
       "0                     0                c.(433delC); (433delC)   \n",
       "1                     5                  c.(650G>A); (650G>A)   \n",
       "2                     0                 c.(794C>A); (1280C>A)   \n",
       "3                  n.a.  c.(733_736delCTTT); (733_736delCTTT)   \n",
       "4                  n.a.            c.(284_285insC); (1126C>T)   \n",
       "\n",
       "                     Amino acid   status Typical type/Mild type  ...  \\\n",
       "0  p.(Q145Sfs*16); (Q145Sfs*16)     Homo                      T  ...   \n",
       "1            p.(R217Q); (R217Q)     Homo                      T  ...   \n",
       "2            p.(A265D); (S427*)  Com het                      T  ...   \n",
       "3  p.(L245Pfs*27); (L245Pfs*27)     Homo                      T  ...   \n",
       "4             p.(E97*); (R376C)  Com het                      T  ...   \n",
       "\n",
       "  Cys (umol/L) NR: 20-70 UA (umol/L) NR: 210-430 Sulfite (mg/L) NR: 0  \\\n",
       "0                      2                    n.a.                20-25   \n",
       "1                   n.a.                    n.a.          0.108-0.211   \n",
       "2                   n.a.                  normal               80-100   \n",
       "3                   n.a.                    n.a.                 n.a.   \n",
       "4                   n.a.                    n.a.                 n.a.   \n",
       "\n",
       "  Thiosulfate NR: 0  Urine SSC (umol/mmolCr) NR: 0.1-10  \\\n",
       "0               n.a.                          320umol/L   \n",
       "1  0.297-1.632mmol/L                          240umol/L   \n",
       "2               n.a.                          690umol/L   \n",
       "3               n.a.                               n.a.   \n",
       "4               n.a.                               n.a.   \n",
       "\n",
       "  Urine Taurine (mmol/molCr) NR: 12-150  \\\n",
       "0                                    95   \n",
       "1                                  n.a.   \n",
       "2                                  n.a.   \n",
       "3                                  n.a.   \n",
       "4                                  n.a.   \n",
       "\n",
       "  Urine XA NR: 0-0.46mmol/L or <40umol/mmolCr or <0.29XA/Cr  \\\n",
       "0                                                      n.a.   \n",
       "1                                                0.04mmol/L   \n",
       "2                                                    normal   \n",
       "3                                                      n.a.   \n",
       "4                                                      n.a.   \n",
       "\n",
       "  Urine HypoXA NR: 0-0.18mmol/L or <8umol/mmolCr or <0.5HypoXA/Cr  \\\n",
       "0                                                            n.a.   \n",
       "1                                                      0.05mmol/L   \n",
       "2                                                          normal   \n",
       "3                                                            n.a.   \n",
       "4                                                            n.a.   \n",
       "\n",
       "  Urine UA NR: 0.44-4.50mmol/L or 50-980umol/mmolCr  \\\n",
       "0                                              n.a.   \n",
       "1                                        0.14mmol/L   \n",
       "2                                            normal   \n",
       "3                                              n.a.   \n",
       "4                                              n.a.   \n",
       "\n",
       "                individual_id  \n",
       "0   individual_1_PMID_9050047  \n",
       "1   individual_2_PMID_9600976  \n",
       "2  individual_3_PMID_10519592  \n",
       "3  individual_4_PMID_12112661  \n",
       "4  individual_5_PMID_12112661  \n",
       "\n",
       "[5 rows x 37 columns]"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "e25b9519-f015-4e72-9f06-dda3cff04385",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'NewAgeCol'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/GIT/phenopacket-store/pstore_env/lib/python3.11/site-packages/pandas/core/indexes/base.py:3790\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3789\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3790\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3791\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32mindex.pyx:152\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:181\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7080\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7088\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'NewAgeCol'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[106], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m ageMapper \u001b[38;5;241m=\u001b[39m AgeColumnMapper(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mAgeEncodingType.YEAR_AND_MONTH\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mNewAgeCol\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m----> 2\u001b[0m ageMapper\u001b[38;5;241m.\u001b[39mpreview_column(\u001b[43mdf\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mNewAgeCol\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m)\n\u001b[1;32m      4\u001b[0m sexMapper \u001b[38;5;241m=\u001b[39m SexColumnMapper(male_symbol\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mM\u001b[39m\u001b[38;5;124m'\u001b[39m, female_symbol\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mF\u001b[39m\u001b[38;5;124m'\u001b[39m, column_name\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGender\u001b[39m\u001b[38;5;124m'\u001b[39m, unknown_symbol\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      5\u001b[0m sexMapper\u001b[38;5;241m.\u001b[39mpreview_column(df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGender\u001b[39m\u001b[38;5;124m'\u001b[39m])\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/pstore_env/lib/python3.11/site-packages/pandas/core/frame.py:3896\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3894\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   3895\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 3896\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3897\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   3898\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/pstore_env/lib/python3.11/site-packages/pandas/core/indexes/base.py:3797\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3792\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[1;32m   3793\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[1;32m   3794\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[1;32m   3795\u001b[0m     ):\n\u001b[1;32m   3796\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[0;32m-> 3797\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3798\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3799\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3800\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3801\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3802\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'NewAgeCol'"
     ]
    }
   ],
   "source": [
    "ageMapper = AgeColumnMapper('AgeEncodingType.YEAR_AND_MONTH', 'NewAgeCol')\n",
    "ageMapper.preview_column(df['NewAgeCol'])\n",
    "\n",
    "sexMapper = SexColumnMapper(male_symbol='M', female_symbol='F', column_name='Gender', unknown_symbol='')\n",
    "sexMapper.preview_column(df['Gender'])\n",
    "\n",
    "individual_column_name = 'individual_id'\n",
    "\n",
    "\n",
    "encoder = CohortEncoder(df=df, \n",
    "                        hpo_cr=hpo_cr, \n",
    "                        column_mapper_d=column_mapper_d, \n",
    "                        individual_column_name=individual_column_name,\n",
    "                        agemapper=ageMapper, \n",
    "                        sexmapper=sexMapper,\n",
    "                        metadata=metadata,\n",
    "                        pmid=PMID)\n",
    "\n",
    "encoder.set_disease(disease_id='OMIM:272300', label='Sulfite oxidase deficiency')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "01873645-d89a-4b79-9467-bc9d5129114a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        0\n",
       "1        5\n",
       "2        0\n",
       "3     n.a.\n",
       "4     n.a.\n",
       "5     n.a.\n",
       "6     n.a.\n",
       "7     n.a.\n",
       "8     n.a.\n",
       "9     n.a.\n",
       "10    n.a.\n",
       "11    n.a.\n",
       "12    n.a.\n",
       "13       5\n",
       "14       0\n",
       "15     0.8\n",
       "16     0.1\n",
       "17      12\n",
       "18      12\n",
       "19       0\n",
       "20     0.5\n",
       "21     1.3\n",
       "22     0.7\n",
       "23       0\n",
       "24      12\n",
       "25      16\n",
       "26      14\n",
       "27       0\n",
       "28     0.2\n",
       "29       9\n",
       "30      23\n",
       "31     0.5\n",
       "32       0\n",
       "33     0.1\n",
       "34    n.a.\n",
       "Name: Age at onset (months), dtype: object"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clinical['Age at onset (months)']\n",
    "ageMapper = AgeColumnMapper."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a536f82-744c-472b-aa31-d008b86790ba",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pstore_env",
   "language": "python",
   "name": "pstore_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
