{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GNAS\n",
    "Guanine Nucleotide-Binding Protein G(S) Subunit Alpha is a complex imprinted locus that produces multiple transcripts through the use of alternative promoters and alternative splicing. The most well-characterized transcript derived from GNAS, Gs-alpha, encodes the alpha subunit of the stimulatory guanine nucleotide-binding protein (G protein) ([OMIM:139320](https://omim.org/entry/139320)).\n",
    "\n",
    "Pseudohypoparathyroidism type Ia (PHP Ia; [OMIM:103580](https://omim.org/entry/103580)) is caused by mutation resulting in loss of function of the Gs-alpha isoform of the GNAS gene (139320) on the maternal allele. This results in expression of the Gs-alpha protein only from the paternal allele.\n",
    "\n",
    "Pseudopseudohypoparathyroidism (PPHP; [OMIM:612463](https://omim.org/entry/612463)) is caused by a mutation resulting in loss of function of the Gs-alpha isoform of the GNAS gene (139320) on the paternal allele. This results in expression of the Gs-alpha protein only from the maternal allele.\n",
    "\n",
    "[Thiel et al. (2015)](https://pubmed.ncbi.nlm.nih.gov/25802881/) report\n",
    "\n",
    "> Mutations are distributed throughout the Gsα coding exons of GNAS and there is a lack of genotype-phenotype correlation. The authors identified 58 different mutations in 88 patients and 27 relatives, and found a significantly higher occurrence of subcutaneous calcifications in patients harboring truncating versus missense mutations was demonstrated. \n",
    "\n",
    "In this notebook, we extract phenopackets from the Supplemental table provided by Thiel et al.\n",
    "\n",
    "Note that the original publication reports variants according to Refseq: NM_001077488.1. We have translated coordinates to the MANE transcript is NM_000516.7, both are shown in the input Excel file that we created based on the original Word supplemental file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using pyphetools version 0.9.108\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "pd.set_option('display.max_colwidth', None) # show entire column contents, important!\n",
    "import typing\n",
    "import os\n",
    "import re\n",
    "from google.protobuf.json_format import MessageToJson\n",
    "from pyphetools.creation import Measurements, VariantManager, PromoterVariant, StructuralVariant, OntologyTerms\n",
    "from pyphetools.pp.v202 import *\n",
    "import pyphetools\n",
    "print(f\"Using pyphetools version {pyphetools.__version__}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Index No</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age</th>\n",
       "      <th>PHP Type</th>\n",
       "      <th>AHOsigns</th>\n",
       "      <th>PTH[pg/ml]</th>\n",
       "      <th>Ca[mmol/l]</th>\n",
       "      <th>Ph[mmol/l]</th>\n",
       "      <th>Gsalpha activity[%]</th>\n",
       "      <th>Endocrinopathies</th>\n",
       "      <th>NM_001077488.1</th>\n",
       "      <th>NM_000516.7</th>\n",
       "      <th>variant_comment</th>\n",
       "      <th>Location</th>\n",
       "      <th>Inheritance</th>\n",
       "      <th>Previously described</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P1</td>\n",
       "      <td>F</td>\n",
       "      <td>P10Y</td>\n",
       "      <td>PHPIa</td>\n",
       "      <td>o,sst,bm,scc,mr</td>\n",
       "      <td>669</td>\n",
       "      <td>1.75</td>\n",
       "      <td>2.62</td>\n",
       "      <td>62.6</td>\n",
       "      <td>PTH, TSH</td>\n",
       "      <td>c.93dupG</td>\n",
       "      <td>c.93dup</td>\n",
       "      <td>p.K32EfsX22</td>\n",
       "      <td>Exon 1</td>\n",
       "      <td>de novo</td>\n",
       "      <td>(Thiele et al., 2010)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P2</td>\n",
       "      <td>F</td>\n",
       "      <td>P15Y10M</td>\n",
       "      <td>PHPIa</td>\n",
       "      <td>o,bm</td>\n",
       "      <td>43</td>\n",
       "      <td>2.33</td>\n",
       "      <td>1.69</td>\n",
       "      <td>50.8</td>\n",
       "      <td>TSH, GHRH</td>\n",
       "      <td>c.89T&gt;C</td>\n",
       "      <td>c.89T&gt;C</td>\n",
       "      <td>p.L30P</td>\n",
       "      <td>Exon 1</td>\n",
       "      <td>de novo</td>\n",
       "      <td>This study</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P3</td>\n",
       "      <td>F</td>\n",
       "      <td>P21Y8M</td>\n",
       "      <td>PHPIa</td>\n",
       "      <td>o,bm</td>\n",
       "      <td>nr</td>\n",
       "      <td>nr</td>\n",
       "      <td>nr</td>\n",
       "      <td>45.6</td>\n",
       "      <td>TSH</td>\n",
       "      <td>c.94_96delAAG</td>\n",
       "      <td>c.94_96del</td>\n",
       "      <td>p.K32del</td>\n",
       "      <td>Exon 1</td>\n",
       "      <td>de novo</td>\n",
       "      <td>This study</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P4</td>\n",
       "      <td>M</td>\n",
       "      <td>P1Y</td>\n",
       "      <td>PHPIa</td>\n",
       "      <td>scc</td>\n",
       "      <td>el</td>\n",
       "      <td>nr</td>\n",
       "      <td>nr</td>\n",
       "      <td>49.4</td>\n",
       "      <td>PTH</td>\n",
       "      <td>c.136_138dupCTG</td>\n",
       "      <td>c.136_138dup</td>\n",
       "      <td>p.L46dup</td>\n",
       "      <td>Exon 1</td>\n",
       "      <td>de novo</td>\n",
       "      <td>(Fernández Rebollo et al., 2013; Lim et al., 2002)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P5</td>\n",
       "      <td>F</td>\n",
       "      <td>P16Y3M</td>\n",
       "      <td>PPHP</td>\n",
       "      <td>bm,scc,mr</td>\n",
       "      <td>nr</td>\n",
       "      <td>nr</td>\n",
       "      <td>nr</td>\n",
       "      <td>49</td>\n",
       "      <td>no</td>\n",
       "      <td>c.136_138dupCTG</td>\n",
       "      <td>c.136_138dup</td>\n",
       "      <td>p.L46dup</td>\n",
       "      <td>NaN</td>\n",
       "      <td>de novo</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Index No Sex      Age PHP Type         AHOsigns PTH[pg/ml] Ca[mmol/l]  \\\n",
       "0       P1   F     P10Y    PHPIa  o,sst,bm,scc,mr        669       1.75   \n",
       "1       P2   F  P15Y10M    PHPIa             o,bm         43       2.33   \n",
       "2       P3   F   P21Y8M    PHPIa             o,bm         nr         nr   \n",
       "3       P4   M      P1Y    PHPIa              scc         el         nr   \n",
       "4       P5   F   P16Y3M     PPHP        bm,scc,mr         nr         nr   \n",
       "\n",
       "  Ph[mmol/l] Gsalpha activity[%] Endocrinopathies   NM_001077488.1  \\\n",
       "0       2.62                62.6         PTH, TSH         c.93dupG   \n",
       "1       1.69                50.8        TSH, GHRH          c.89T>C   \n",
       "2         nr                45.6              TSH    c.94_96delAAG   \n",
       "3         nr                49.4              PTH  c.136_138dupCTG   \n",
       "4         nr                  49               no  c.136_138dupCTG   \n",
       "\n",
       "    NM_000516.7 variant_comment Location Inheritance  \\\n",
       "0       c.93dup     p.K32EfsX22   Exon 1     de novo   \n",
       "1       c.89T>C          p.L30P   Exon 1     de novo   \n",
       "2    c.94_96del        p.K32del   Exon 1     de novo   \n",
       "3  c.136_138dup        p.L46dup   Exon 1     de novo   \n",
       "4  c.136_138dup        p.L46dup      NaN     de novo   \n",
       "\n",
       "                                 Previously described  \n",
       "0                               (Thiele et al., 2010)  \n",
       "1                                          This study  \n",
       "2                                          This study  \n",
       "3  (Fernández Rebollo et al., 2013; Lim et al., 2002)  \n",
       "4                                                 NaN  "
      ]
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_excel(\"input/Thiele_2015_GNAS.xlsx\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "PMID = \"PMID:25802881\"\n",
    "title = \"A positive genotype-phenotype correlation in a large cohort of patients with Pseudohypoparathyroidism Type Ia and Pseudo-pseudohypoparathyroidism and 33 newly identified mutations in the GNAS gene\"\n",
    "created_by=\"ORCID:0000-0002-5648-2155\"\n",
    "metadata = MetaData.metadata_for_pmid(created_by=created_by, pmid=PMID, citation_title=title, include_loinc=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parsing\n",
    "\n",
    "\n",
    "The suppl. table 1 summarizes the available data of our cohort of PHPIa, PPHP, and POH patients, including the age at diagnosis, the clinical signs, laboratory results, peptide hormone resistance. the underlying genetic defect, and the inheritance. P: patient, M: mother, S: sister, B: brother, f: female, m: male, nr: normal range; el: elevated; nk: not known, \n",
    "Reference values: PTH: 15-55 pg/ml, Ca: 2.1-2.6 mmol/l, Ph 0.9-1.75mmol/l, Gsalpha activity 85-115%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_aho_hpos(aho_signs):\n",
    "    \"\"\"\n",
    "    Signs related to Albright hereditary osteodystrophy (AHO).\n",
    "    Abbreviations used in the original publication: o: obesity, sst: short stature, bm: brachmetacarpia, scc: subcutanous calicifications, mr: mental redardation\n",
    "    \"\"\"\n",
    "    aho_list = aho_signs.split(\",\")\n",
    "    hpo_term_d = {\n",
    "        \"o\": [\"Obesity\", \"HP:0001513\"],\n",
    "        \"sst\": [\"Short stature\", \"HP:0004322\"],\n",
    "        \"bm\": [\"Short metacarpal\", \"HP:0010049\"],\n",
    "        \"scc\": [\"Subcutaneous ossification\", \"HP:0034282\"],\n",
    "        \"mr\": [\"Intellectual disability\", \"HP:0001249\"],\n",
    "        \"bd\": [\"Brachydactyly\", \"HP:0001156\"]\n",
    "    }\n",
    "    phenotypic_feature_list = list()\n",
    "    for aho in aho_list:\n",
    "        if aho == \"no signs\" or aho == \"\":\n",
    "            continue\n",
    "        if aho == \"nk\":\n",
    "            continue # not known\n",
    "        elif aho in hpo_term_d:\n",
    "            arr = hpo_term_d.get(aho)\n",
    "            label = arr[0]\n",
    "            hpo_id = arr[1]\n",
    "            term = OntologyClass(id=hpo_id, label=label)\n",
    "            phenotypic_feature_list.append(PhenotypicFeature(type=term))\n",
    "        else:\n",
    "            raise ValueError(f\"Did not recognize AHO sign: \\\"{aho}\\\"\")\n",
    "    return phenotypic_feature_list       \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_disease(diagnosis_type):\n",
    "    if diagnosis_type == \"PHPIa\": # PHPIa\n",
    "        diseaseClass = OntologyClass(id=\"OMIM:103580\", label=\"Pseudohypoparathyroidism Ia\")\n",
    "    elif diagnosis_type == \"PPHP\":\n",
    "        diseaseClass = OntologyClass(id=\"OMIM:612463\", label=\"Pseudopseudohypoparathyroidism\")\n",
    "    elif diagnosis_type == \"POH\":\n",
    "        diseaseClass = OntologyClass(id=\"OMIM:166350\", label=\"Osseous heteroplasia, progressive\")\t\n",
    "    else:\n",
    "        raise ValueError(f\"Did not recognize diagnosis \\\"{diagnosis_type}\\\"\")\n",
    "    return Disease(term=diseaseClass) \n",
    "    \t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parathormone(row:pd.Series, \n",
    "        measurement_list: typing.List[Measurement],\n",
    "        phenotypic_feature_list: typing.List[PhenotypicFeature]):\n",
    "    \"\"\"\n",
    "    PTH: 15-55 pg/ml\n",
    "    LOINC:35566-9, Parathyrin [Mass/volume] in Serum or Plasma --baseline\n",
    "    \"\"\"\n",
    "    value = row[\"PTH[pg/ml]\"]\n",
    "    if value == \"na\" or value == \"nk\":\n",
    "        return None\n",
    "    elevated_pth = PhenotypicFeature(type=OntologyClass(id=\"HP:0003165\", label=\"Elevated circulating parathyroid hormone level\"))\n",
    "    diminished_pth = PhenotypicFeature(type=OntologyClass(id=\"HP:0031817\", label=\"Decreased circulating parathyroid hormone level\"))\n",
    "    not_elevated_pth = PhenotypicFeature(type=OntologyClass(id=\"HP:0003165\", label=\"Elevated circulating parathyroid hormone level\"), excluded=True)\n",
    "    if value == \"nr\":\n",
    "        phenotypic_feature_list.append(not_elevated_pth)\n",
    "        return\n",
    "    if value == \"el\":\n",
    "        phenotypic_feature_list.append(elevated_pth)\n",
    "        return\n",
    "    if value == \"low\":\n",
    "        phenotypic_feature_list.append(diminished_pth)\n",
    "        return\n",
    "    if isinstance(value,float): # this happens if there is no value in the input table cell\n",
    "        return None\n",
    "    if isinstance(value,str) and value.endswith(\" \"):\n",
    "        raise ValueError(f\"Maformed PTH: \\\"{value}\\\"\")\n",
    "    lower_limit_of_normal = 15\n",
    "    upper_limit_of_normal = 55\n",
    "    try:\n",
    "        concentration = float(value)\n",
    "        m = Measurements.picogram_per_milliliter(code=\"LOINC:35566-9\",\n",
    "                                      label=\"Parathyrin [Mass/volume] in Serum or Plasma --baseline\",\n",
    "                                      concentration=concentration,\n",
    "                                      low=lower_limit_of_normal,\n",
    "                                      high=upper_limit_of_normal)\n",
    "        measurement_list.append(m)\n",
    "        if concentration > upper_limit_of_normal:\n",
    "            pf = elevated_pth\n",
    "        elif concentration < lower_limit_of_normal:\n",
    "            pf = diminished_pth\n",
    "        else:\n",
    "            pf = not_elevated_pth\n",
    "        phenotypic_feature_list.append(pf)\n",
    "    except ValueError:\n",
    "        print(f\"Could not parse \\\"{value}\\\"\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calcium(row:pd.Series, \n",
    "        measurement_list: typing.List[Measurement],\n",
    "        phenotypic_feature_list: typing.List[PhenotypicFeature]):\n",
    "    \"\"\"\n",
    "    Ca: 2.1-2.6 mmol/l\n",
    "    LOINC: 17861-6  Calcium [Mass/volume] in Serum or Plasma\n",
    "    \"\"\"\n",
    "    value = row[\"Ca[mmol/l]\"]\n",
    "    if value == \"na\" or value == \"nk\":\n",
    "        return None\n",
    "    if isinstance(value,float): # this happens if there is no value in the input table cell\n",
    "        return None\n",
    "    if value == \"nr\":\n",
    "        pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0004363\", label=\"Abnormal circulating calcium concentration\"), excluded=True)\n",
    "        phenotypic_feature_list.append(pf)\n",
    "        return\n",
    "    if value == \"el\":\n",
    "        pf =  PhenotypicFeature(type=OntologyClass(id=\"HP:0003072\", label=\"Hypercalcemia\"))\n",
    "        phenotypic_feature_list.append(pf)\n",
    "        return\n",
    "    if value == \"low\":\n",
    "        pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0002901\", label=\"Hypocalcemia\"))\n",
    "        phenotypic_feature_list.append(pf)\n",
    "        return\n",
    "    if isinstance(value,str) and value.endswith(\" \"):\n",
    "        raise ValueError(f\"Maformed PTH: \\\"{value}\\\"\")\n",
    "    lower_limit_of_normal = 2.1\n",
    "    upper_limit_of_normal = 2.6\n",
    "    try:\n",
    "        concentration = float(value)\n",
    "        m = Measurements.millimole_per_liter(code=\"LOINC:17861-6\",\n",
    "                        label=\"Calcium [Mass/volume] in Serum or Plasma\",\n",
    "                        concentration=concentration,\n",
    "                        low=lower_limit_of_normal,\n",
    "                        high=upper_limit_of_normal)\n",
    "        measurement_list.append(m)\n",
    "        if concentration > upper_limit_of_normal:\n",
    "            pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0003072\", label=\"Hypercalcemia\"))\n",
    "        elif concentration < lower_limit_of_normal:\n",
    "            pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0002901\", label=\"Hypocalcemia\"))\n",
    "        else:\n",
    "            pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0004363\", label=\"Abnormal circulating calcium concentration\"), excluded=True)\n",
    "        phenotypic_feature_list.append(pf)\n",
    "    except ValueError:\n",
    "        print(f\"Could not parse \\\"{value}\\\"\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [],
   "source": [
    "def phosphate(row:pd.Series, \n",
    "        measurement_list: typing.List[Measurement],\n",
    "        phenotypic_feature_list: typing.List[PhenotypicFeature]):\n",
    "    \"\"\"\n",
    "    Ph 0.9-1.75mmol/l\n",
    "    LOINC: 2777-1 Phosphate [Mass/volume] in Serum or Plasma\n",
    "    \"\"\"\n",
    "    value = row[\"Ph[mmol/l]\"]\n",
    "    if value == \"na\" or value == \"nk\":\n",
    "        return None\n",
    "    if isinstance(value,float): # this happens if there is no value in the input table cell\n",
    "        return None\n",
    "    if value == \"nr\":\n",
    "        pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0100529\", label=\"Abnormal blood phosphate concentration\"), excluded=True)\n",
    "        phenotypic_feature_list.append(pf)\n",
    "        return\n",
    "    if value == \"el\":\n",
    "        pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0002905\", label=\"Hyperphosphatemia\"))\n",
    "        phenotypic_feature_list.append(pf)\n",
    "        return\n",
    "    if value == \"low\":\n",
    "        pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0002148\", label=\"Hypophosphatemia\"))\n",
    "        phenotypic_feature_list.append(pf)\n",
    "        return\n",
    "    if isinstance(value,str) and value.endswith(\" \"):\n",
    "        raise ValueError(f\"Maformed PTH: \\\"{value}\\\"\")\n",
    "    lower_limit_of_normal = 0.9\n",
    "    upper_limit_of_normal = 1.75\n",
    "    try:\n",
    "        concentration = float(value)\n",
    "        m = Measurements.millimole_per_liter(code=\"LOINC:2777-1\",\n",
    "                        label=\"Phosphate [Mass/volume] in Serum or Plasma\",\n",
    "                        concentration=concentration,\n",
    "                        low=lower_limit_of_normal,\n",
    "                        high=upper_limit_of_normal)\n",
    "        measurement_list.append(m)\n",
    "        if concentration > upper_limit_of_normal:\n",
    "            pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0002905\", label=\"Hyperphosphatemia\"))\n",
    "        elif concentration < lower_limit_of_normal:\n",
    "            pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0002148\", label=\"Hypophosphatemia\"))\n",
    "        else:\n",
    "            pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0100529\", label=\"Abnormal blood phosphate concentration\"), excluded=True)\n",
    "        phenotypic_feature_list.append(pf)\n",
    "    except ValueError:\n",
    "        print(f\"Could not parse \\\"{value}\\\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [],
   "source": [
    "gnas_symbol = \"GNAS\"\n",
    "gnas_id = \"HGNC:4392\"\n",
    "gnas_MANE_transcript = \"NM_000516.7\"\n",
    "vmanager = VariantManager(df=df, \n",
    "    individual_column_name=\"Index No\", \n",
    "    transcript=gnas_MANE_transcript, \n",
    "    gene_id=gnas_id, \n",
    "    gene_symbol=gnas_symbol, \n",
    "    allele_1_column_name=gnas_MANE_transcript)\n",
    "variant_d = vmanager.get_variant_d()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [],
   "source": [
    "def row_to_phenopacket(row:pd.Series):\n",
    "    individual_id = row[\"Index No\"]\n",
    "    phenopacket_id = \"PMID_25802881_{}\".format(individual_id.replace(\" \", \"_\"))\n",
    "    sex = row[\"Sex\"]\n",
    "    age_at_dx = row[\"Age\"]\n",
    "    age_of_onset = Age(iso8601duration=age_at_dx)\n",
    "    i = Individual(id=individual_id)\n",
    "    if sex == \"M\":\n",
    "        i.sex = Sex.MALE\n",
    "    elif sex == \"F\":\n",
    "        i.sex = Sex.FEMALE\n",
    "    else:\n",
    "        print(f\"Warning count not identify sex for {individual_id}\")\n",
    "    allele_1 = row[gnas_MANE_transcript]\n",
    "    \n",
    "    var = variant_d.get(allele_1)\n",
    "    var_obj = var.to_variant_interpretation_202()\n",
    "    var_obj.variation_descriptor.allelic_state = OntologyTerms.heterozygous()\n",
    "    ## create genomic interpretation\n",
    "    interpretation_list = list()\n",
    "    genomic_interpretation = GenomicInterpretation(subject_or_biosample_id=individual_id, \n",
    "                                                       interpretation_status=GenomicInterpretation.InterpretationStatus.CAUSATIVE,\n",
    "                                                       call=var_obj)\n",
    "    interpretation_list.append(genomic_interpretation)\n",
    "    diagnosis_type = row[\"PHP Type\"]\n",
    "    disease = get_disease(diagnosis_type=diagnosis_type)\n",
    "    diagnosis = Diagnosis(disease=disease.term)\n",
    "    interpretation = Interpretation(id=individual_id, progress_status=Interpretation.ProgressStatus.SOLVED, diagnosis=diagnosis)\n",
    "    aho_signs = row[\"AHOsigns\"]\n",
    "    phenotypic_features = get_aho_hpos(aho_signs)\n",
    "    measurements = list()\n",
    "    parathormone(row=row, phenotypic_feature_list=phenotypic_features, measurement_list=measurements) \n",
    "    calcium(row=row, measurement_list=measurements, phenotypic_feature_list=phenotypic_features)\n",
    "    phosphate(row=row, measurement_list=measurements, phenotypic_feature_list=phenotypic_features)\n",
    "   \n",
    "    ppkt = Phenopacket(id=phenopacket_id, \n",
    "                       subject=i, \n",
    "                       diseases=[disease], \n",
    "                       phenotypic_features=phenotypic_features, \n",
    "                       measurements=measurements, \n",
    "                       interpretations=[interpretation], \n",
    "                       meta_data=metadata)\n",
    "    return ppkt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenopacket_list = list()\n",
    "firstRow = None\n",
    "for _,row in df.iterrows():\n",
    "    if firstRow is None:\n",
    "        firstRow = row\n",
    "        continue ## first row has additional definitions such as HPO identifiers\n",
    "    item = row_to_phenopacket(row)\n",
    "    phenopacket_list.append(item.to_message())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We output 114 GA4GH phenopackets to the directory phenopackets\n"
     ]
    }
   ],
   "source": [
    "outdir = \"phenopackets\"\n",
    "if not os.path.isdir(outdir):\n",
    "    os.makedirs(outdir)\n",
    "written = 0\n",
    "json_list = list()\n",
    "for ppkt in phenopacket_list:    \n",
    "    json_string = MessageToJson(ppkt)\n",
    "    fname = re.sub('[^A-Za-z0-9_-]', '', ppkt.id)  # remove any illegal characters from filename\n",
    "    fname = fname.replace(\" \", \"_\") + \".json\"\n",
    "    outpth = os.path.join(outdir, fname)\n",
    "    with open(outpth, \"wt\") as fh:\n",
    "        fh.write(json_string)\n",
    "        json_list.append(json_string)\n",
    "        written += 1\n",
    "print(f\"We output {written} GA4GH phenopackets to the directory {outdir}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ps24venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
