{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0d6ffbc5",
   "metadata": {},
   "source": [
    "<h1>SMARCC2: Bosch et al 2023 </h1>\n",
    "<p>Extract the clinical data from <a href=\"https://pubmed.ncbi.nlm.nih.gov/37551667/\"target=\"__blank\">Bosch E, et al. (2023) Elucidating the clinical and molecular spectrum of SMARCC2-associated NDD in a cohort of 65 affected individuals. Genet Med.  PMID:37551667</a>.<p>\n",
    "<p>The authors report that clinical presentation differed significantly, with LGD variants being predominantly inherited and associated with mildly reduced or normal cognitive development, while non-truncating variants were mostly de novo and presented with severe developmental delay. </p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7c95c46e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using pyphetools version 0.4.8\n"
     ]
    }
   ],
   "source": [
    "import phenopackets as php\n",
    "from google.protobuf.json_format import MessageToDict, MessageToJson\n",
    "from google.protobuf.json_format import Parse, ParseDict\n",
    "import pandas as pd\n",
    "import math\n",
    "from csv import DictReader\n",
    "pd.set_option('display.max_colwidth', None) # show entire column contents, important!\n",
    "from collections import defaultdict\n",
    "import re\n",
    "from pyphetools.creation import *\n",
    "from pyphetools.output import PhenopacketTable\n",
    "import importlib.metadata\n",
    "__version__ = importlib.metadata.version(\"pyphetools\")\n",
    "print(f\"Using pyphetools version {__version__}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b959053c",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = HpoParser()\n",
    "hpo_cr = parser.get_hpo_concept_recognizer()\n",
    "hpo_version = parser.get_version()\n",
    "metadata = MetaData(created_by=\"ORCID:0000-0002-0736-9199\")\n",
    "metadata.default_versions_with_hpo(version=hpo_version)\n",
    "pmid=\"37551667\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f044ce22",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel(\"input/FileS2_cases_clinical-table.xlsx\", index_col =\"Patient_ID (in Project)\", comment=\"##\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8fede285",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HPO</th>\n",
       "      <th>Ind-01</th>\n",
       "      <th>Ind-02</th>\n",
       "      <th>Ind-03</th>\n",
       "      <th>Ind-04</th>\n",
       "      <th>Ind-05</th>\n",
       "      <th>Ind-06</th>\n",
       "      <th>Ind-07</th>\n",
       "      <th>Ind-08</th>\n",
       "      <th>Ind-09</th>\n",
       "      <th>...</th>\n",
       "      <th>Machol_Ind_15</th>\n",
       "      <th>Li_Ind_1</th>\n",
       "      <th>Chen_Pat_123</th>\n",
       "      <th>Chen_Pat_124</th>\n",
       "      <th>Chen_Pat_126</th>\n",
       "      <th>Sun_case</th>\n",
       "      <th>Yi_case</th>\n",
       "      <th>Lo_twin_1</th>\n",
       "      <th>Lo_twin_2</th>\n",
       "      <th>Gofin_Subject_5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID (in Project)</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>#family</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Fam-01</td>\n",
       "      <td>Fam-02</td>\n",
       "      <td>Fam-03</td>\n",
       "      <td>Fam-04</td>\n",
       "      <td>Fam-05</td>\n",
       "      <td>Fam-06</td>\n",
       "      <td>Fam-07</td>\n",
       "      <td>Fam-08</td>\n",
       "      <td>Fam-09</td>\n",
       "      <td>...</td>\n",
       "      <td>Fam-45</td>\n",
       "      <td>Fam-46</td>\n",
       "      <td>Fam-47</td>\n",
       "      <td>Fam-48</td>\n",
       "      <td>Fam-49</td>\n",
       "      <td>Fam-50</td>\n",
       "      <td>Fam-51</td>\n",
       "      <td>Fam-52</td>\n",
       "      <td>Fam-52</td>\n",
       "      <td>Fam-53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#group</th>\n",
       "      <td>NaN</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>novel</td>\n",
       "      <td>...</td>\n",
       "      <td>literature</td>\n",
       "      <td>literature</td>\n",
       "      <td>literature</td>\n",
       "      <td>literature</td>\n",
       "      <td>literature</td>\n",
       "      <td>literature</td>\n",
       "      <td>literature</td>\n",
       "      <td>literature</td>\n",
       "      <td>literature</td>\n",
       "      <td>literature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#analysis</th>\n",
       "      <td>NaN</td>\n",
       "      <td>include</td>\n",
       "      <td>include</td>\n",
       "      <td>exclude</td>\n",
       "      <td>include</td>\n",
       "      <td>include</td>\n",
       "      <td>exclude</td>\n",
       "      <td>include</td>\n",
       "      <td>include</td>\n",
       "      <td>include</td>\n",
       "      <td>...</td>\n",
       "      <td>include</td>\n",
       "      <td>include</td>\n",
       "      <td>include</td>\n",
       "      <td>include</td>\n",
       "      <td>include</td>\n",
       "      <td>include</td>\n",
       "      <td>include</td>\n",
       "      <td>include</td>\n",
       "      <td>include</td>\n",
       "      <td>exclude</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#Case published previously</th>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#Literature reference</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PMID:30580808</td>\n",
       "      <td>PMID:34881817</td>\n",
       "      <td>PMID:34906496</td>\n",
       "      <td>PMID:34906496</td>\n",
       "      <td>PMID:34906496</td>\n",
       "      <td>PMID:35241061</td>\n",
       "      <td>PMID:35536477</td>\n",
       "      <td>PMID:35699097</td>\n",
       "      <td>PMID:35699097</td>\n",
       "      <td>PMID:35796094</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 66 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            HPO   Ind-01   Ind-02   Ind-03   Ind-04   Ind-05  \\\n",
       "Patient_ID (in Project)                                                        \n",
       "#family                     NaN   Fam-01   Fam-02   Fam-03   Fam-04   Fam-05   \n",
       "#group                      NaN    novel    novel    novel    novel    novel   \n",
       "#analysis                   NaN  include  include  exclude  include  include   \n",
       "#Case published previously  NaN       no       no       no       no       no   \n",
       "#Literature reference       NaN      NaN      NaN      NaN      NaN      NaN   \n",
       "\n",
       "                             Ind-06   Ind-07   Ind-08   Ind-09  ...  \\\n",
       "Patient_ID (in Project)                                         ...   \n",
       "#family                      Fam-06   Fam-07   Fam-08   Fam-09  ...   \n",
       "#group                        novel    novel    novel    novel  ...   \n",
       "#analysis                   exclude  include  include  include  ...   \n",
       "#Case published previously       no       no       no       no  ...   \n",
       "#Literature reference           NaN      NaN      NaN      NaN  ...   \n",
       "\n",
       "                            Machol_Ind_15       Li_Ind_1   Chen_Pat_123  \\\n",
       "Patient_ID (in Project)                                                   \n",
       "#family                            Fam-45         Fam-46         Fam-47   \n",
       "#group                         literature     literature     literature   \n",
       "#analysis                         include        include        include   \n",
       "#Case published previously            yes            yes            yes   \n",
       "#Literature reference       PMID:30580808  PMID:34881817  PMID:34906496   \n",
       "\n",
       "                             Chen_Pat_124   Chen_Pat_126       Sun_case  \\\n",
       "Patient_ID (in Project)                                                   \n",
       "#family                            Fam-48         Fam-49         Fam-50   \n",
       "#group                         literature     literature     literature   \n",
       "#analysis                         include        include        include   \n",
       "#Case published previously            yes            yes            yes   \n",
       "#Literature reference       PMID:34906496  PMID:34906496  PMID:35241061   \n",
       "\n",
       "                                  Yi_case      Lo_twin_1      Lo_twin_2  \\\n",
       "Patient_ID (in Project)                                                   \n",
       "#family                            Fam-51         Fam-52         Fam-52   \n",
       "#group                         literature     literature     literature   \n",
       "#analysis                         include        include        include   \n",
       "#Case published previously            yes            yes            yes   \n",
       "#Literature reference       PMID:35536477  PMID:35699097  PMID:35699097   \n",
       "\n",
       "                           Gofin_Subject_5  \n",
       "Patient_ID (in Project)                     \n",
       "#family                             Fam-53  \n",
       "#group                          literature  \n",
       "#analysis                          exclude  \n",
       "#Case published previously             yes  \n",
       "#Literature reference        PMID:35796094  \n",
       "\n",
       "[5 rows x 66 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75c61540",
   "metadata": {},
   "source": [
    "<H2>Parse strategy</H2>\n",
    "<p>This supplemental file already has HPO ids and all needed information for each patient."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f133395c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pat_id_list = df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e82a8511",
   "metadata": {},
   "outputs": [],
   "source": [
    "sex_pd_series = df.loc[\"#Sex\"]\n",
    "sex_pd_series = sex_pd_series.drop(labels=['HPO'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ad83361d",
   "metadata": {},
   "outputs": [],
   "source": [
    "hg38_var_list = df.loc[\"#Variant(s) in SMARCC2 (genomic hg38/GRCh38)\"]\n",
    "hg38_var_list = hg38_var_list.drop(labels=['HPO'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9b7a0559",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([nan, 'missense', 'missense; confirmed protein loss', 'truncating',\n",
       "       'splice; potentially inframe', 'splice; confirmed inframe',\n",
       "       'inframe', 'splice; potentially truncating',\n",
       "       'splice; confirmed NMD'], dtype=object)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant_type_list = df.loc[\"#variant type\"]\n",
    "variant_type_list.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d68b7347",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['heterozygous', 'heterozygous '], dtype=object)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allelic_state_list = df.loc[\"#Allelic state\"]\n",
    "allelic_state_list = allelic_state_list.drop(labels=['HPO'])\n",
    "allelic_state_list.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "13315bf6-2a3e-4929-95c9-ac9d4d3fa73c",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_hgvs_pd_series = df.loc[\"#Variant(s) in SMARCC2 (NM_003075.5: coding and protein)\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86812634",
   "metadata": {},
   "source": [
    "<h2>HPO data</h2>\n",
    "<p>The file contains rows that are already encoded as HPO terms. We look for a yes or no at the beginning of each cell, and do not parse any other text giving more detail.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2098d94e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_hpo = df[df['HPO'].fillna('').str.contains(\"HP:\")]\n",
    "hpo_ids = df_hpo[\"HPO\"].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d9c25112",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HPO</th>\n",
       "      <th>Ind-01</th>\n",
       "      <th>Ind-02</th>\n",
       "      <th>Ind-03</th>\n",
       "      <th>Ind-04</th>\n",
       "      <th>Ind-05</th>\n",
       "      <th>Ind-06</th>\n",
       "      <th>Ind-07</th>\n",
       "      <th>Ind-08</th>\n",
       "      <th>Ind-09</th>\n",
       "      <th>...</th>\n",
       "      <th>Machol_Ind_15</th>\n",
       "      <th>Li_Ind_1</th>\n",
       "      <th>Chen_Pat_123</th>\n",
       "      <th>Chen_Pat_124</th>\n",
       "      <th>Chen_Pat_126</th>\n",
       "      <th>Sun_case</th>\n",
       "      <th>Yi_case</th>\n",
       "      <th>Lo_twin_1</th>\n",
       "      <th>Lo_twin_2</th>\n",
       "      <th>Gofin_Subject_5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID (in Project)</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Microcephaly</th>\n",
       "      <td>HP:0000252</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>...</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Macrocephaly</th>\n",
       "      <td>HP:0000256</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Abnormal facial shape</th>\n",
       "      <td>HP:0001999</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes; occipital plagiocephaly</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes; macrocephaly</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>no</td>\n",
       "      <td>yes; macrocephaly</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Abnormality of the eye</th>\n",
       "      <td>HP:0000478</td>\n",
       "      <td>no</td>\n",
       "      <td>yes; strabismus</td>\n",
       "      <td>yes; ptosis at 6y</td>\n",
       "      <td>no</td>\n",
       "      <td>yes; strabismus</td>\n",
       "      <td>yes; strabismus</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Abnormality of the hand</th>\n",
       "      <td>HP:0001155</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes; clinodactyly</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no; slender appearance of hands and fingers without true arachnodactyly</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>no</td>\n",
       "      <td>yes; fetal finger pads bilaterally</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 66 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                HPO Ind-01           Ind-02  \\\n",
       "Patient_ID (in Project)                                       \n",
       "Microcephaly             HP:0000252     no              yes   \n",
       "Macrocephaly             HP:0000256     no               no   \n",
       "Abnormal facial shape    HP:0001999    yes               no   \n",
       "Abnormality of the eye   HP:0000478     no  yes; strabismus   \n",
       "Abnormality of the hand  HP:0001155     no               no   \n",
       "\n",
       "                                    Ind-03             Ind-04  \\\n",
       "Patient_ID (in Project)                                         \n",
       "Microcephaly                           yes                yes   \n",
       "Macrocephaly                            no                 no   \n",
       "Abnormal facial shape                  yes                yes   \n",
       "Abnormality of the eye   yes; ptosis at 6y                 no   \n",
       "Abnormality of the hand                 no  yes; clinodactyly   \n",
       "\n",
       "                                               Ind-05           Ind-06  \\\n",
       "Patient_ID (in Project)                                                  \n",
       "Microcephaly                                       no               no   \n",
       "Macrocephaly                                       no               no   \n",
       "Abnormal facial shape    yes; occipital plagiocephaly               no   \n",
       "Abnormality of the eye                yes; strabismus  yes; strabismus   \n",
       "Abnormality of the hand                            no               no   \n",
       "\n",
       "                                                                                          Ind-07  \\\n",
       "Patient_ID (in Project)                                                                            \n",
       "Microcephaly                                                                                  no   \n",
       "Macrocephaly                                                                                  no   \n",
       "Abnormal facial shape                                                                         no   \n",
       "Abnormality of the eye                                                                        no   \n",
       "Abnormality of the hand  no; slender appearance of hands and fingers without true arachnodactyly   \n",
       "\n",
       "                                    Ind-08 Ind-09  ... Machol_Ind_15  \\\n",
       "Patient_ID (in Project)                            ...                 \n",
       "Microcephaly                            no    yes  ...            no   \n",
       "Macrocephaly                           yes     no  ...            no   \n",
       "Abnormal facial shape    yes; macrocephaly     no  ...            no   \n",
       "Abnormality of the eye                  no     no  ...            no   \n",
       "Abnormality of the hand                 no     no  ...            no   \n",
       "\n",
       "                                                   Li_Ind_1 Chen_Pat_123  \\\n",
       "Patient_ID (in Project)                                                    \n",
       "Microcephaly                                             no          NaN   \n",
       "Macrocephaly                                            yes          NaN   \n",
       "Abnormal facial shape                     yes; macrocephaly          NaN   \n",
       "Abnormality of the eye                                  NaN          NaN   \n",
       "Abnormality of the hand  yes; fetal finger pads bilaterally          NaN   \n",
       "\n",
       "                        Chen_Pat_124 Chen_Pat_126 Sun_case Yi_case Lo_twin_1  \\\n",
       "Patient_ID (in Project)                                                        \n",
       "Microcephaly                     NaN          NaN      NaN     NaN       NaN   \n",
       "Macrocephaly                     NaN          NaN      NaN     NaN       NaN   \n",
       "Abnormal facial shape            NaN          NaN      NaN      no       NaN   \n",
       "Abnormality of the eye           NaN          NaN      NaN     NaN       NaN   \n",
       "Abnormality of the hand          NaN          NaN      NaN     NaN       NaN   \n",
       "\n",
       "                        Lo_twin_2 Gofin_Subject_5  \n",
       "Patient_ID (in Project)                            \n",
       "Microcephaly                  NaN             NaN  \n",
       "Macrocephaly                  NaN             NaN  \n",
       "Abnormal facial shape         NaN             NaN  \n",
       "Abnormality of the eye        NaN             NaN  \n",
       "Abnormality of the hand       NaN             NaN  \n",
       "\n",
       "[5 rows x 66 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_hpo.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "007b8968",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ind-01</th>\n",
       "      <th>Ind-02</th>\n",
       "      <th>Ind-03</th>\n",
       "      <th>Ind-04</th>\n",
       "      <th>Ind-05</th>\n",
       "      <th>Ind-06</th>\n",
       "      <th>Ind-07</th>\n",
       "      <th>Ind-08</th>\n",
       "      <th>Ind-09</th>\n",
       "      <th>Ind-10</th>\n",
       "      <th>...</th>\n",
       "      <th>Machol_Ind_15</th>\n",
       "      <th>Li_Ind_1</th>\n",
       "      <th>Chen_Pat_123</th>\n",
       "      <th>Chen_Pat_124</th>\n",
       "      <th>Chen_Pat_126</th>\n",
       "      <th>Sun_case</th>\n",
       "      <th>Yi_case</th>\n",
       "      <th>Lo_twin_1</th>\n",
       "      <th>Lo_twin_2</th>\n",
       "      <th>Gofin_Subject_5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HPO</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HP:0000252</th>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0000256</th>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0001999</th>\n",
       "      <td>yes</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes; occipital plagiocephaly</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes; macrocephaly</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>no</td>\n",
       "      <td>yes; macrocephaly</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0000478</th>\n",
       "      <td>no</td>\n",
       "      <td>yes; strabismus</td>\n",
       "      <td>yes; ptosis at 6y</td>\n",
       "      <td>no</td>\n",
       "      <td>yes; strabismus</td>\n",
       "      <td>yes; strabismus</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>no</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0001155</th>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>yes; clinodactyly</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no; slender appearance of hands and fingers without true arachnodactyly</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>no</td>\n",
       "      <td>...</td>\n",
       "      <td>no</td>\n",
       "      <td>yes; fetal finger pads bilaterally</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Ind-01           Ind-02             Ind-03             Ind-04  \\\n",
       "HPO                                                                        \n",
       "HP:0000252     no              yes                yes                yes   \n",
       "HP:0000256     no               no                 no                 no   \n",
       "HP:0001999    yes               no                yes                yes   \n",
       "HP:0000478     no  yes; strabismus  yes; ptosis at 6y                 no   \n",
       "HP:0001155     no               no                 no  yes; clinodactyly   \n",
       "\n",
       "                                  Ind-05           Ind-06  \\\n",
       "HPO                                                         \n",
       "HP:0000252                            no               no   \n",
       "HP:0000256                            no               no   \n",
       "HP:0001999  yes; occipital plagiocephaly               no   \n",
       "HP:0000478               yes; strabismus  yes; strabismus   \n",
       "HP:0001155                            no               no   \n",
       "\n",
       "                                                                             Ind-07  \\\n",
       "HPO                                                                                   \n",
       "HP:0000252                                                                       no   \n",
       "HP:0000256                                                                       no   \n",
       "HP:0001999                                                                       no   \n",
       "HP:0000478                                                                       no   \n",
       "HP:0001155  no; slender appearance of hands and fingers without true arachnodactyly   \n",
       "\n",
       "                       Ind-08 Ind-09 Ind-10  ... Machol_Ind_15  \\\n",
       "HPO                                          ...                 \n",
       "HP:0000252                 no    yes     no  ...            no   \n",
       "HP:0000256                yes     no     no  ...            no   \n",
       "HP:0001999  yes; macrocephaly     no     no  ...            no   \n",
       "HP:0000478                 no     no     no  ...            no   \n",
       "HP:0001155                 no     no     no  ...            no   \n",
       "\n",
       "                                      Li_Ind_1 Chen_Pat_123 Chen_Pat_124  \\\n",
       "HPO                                                                        \n",
       "HP:0000252                                  no          NaN          NaN   \n",
       "HP:0000256                                 yes          NaN          NaN   \n",
       "HP:0001999                   yes; macrocephaly          NaN          NaN   \n",
       "HP:0000478                                 NaN          NaN          NaN   \n",
       "HP:0001155  yes; fetal finger pads bilaterally          NaN          NaN   \n",
       "\n",
       "           Chen_Pat_126 Sun_case Yi_case Lo_twin_1 Lo_twin_2 Gofin_Subject_5  \n",
       "HPO                                                                           \n",
       "HP:0000252          NaN      NaN     NaN       NaN       NaN             NaN  \n",
       "HP:0000256          NaN      NaN     NaN       NaN       NaN             NaN  \n",
       "HP:0001999          NaN      NaN      no       NaN       NaN             NaN  \n",
       "HP:0000478          NaN      NaN     NaN       NaN       NaN             NaN  \n",
       "HP:0001155          NaN      NaN     NaN       NaN       NaN             NaN  \n",
       "\n",
       "[5 rows x 65 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_hpo = df_hpo.set_index('HPO')\n",
    "df_hpo.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "38e2e1e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Got 626 observed HPO terms and 2821 excluded terms for 65 individuals\n"
     ]
    }
   ],
   "source": [
    "patient_id_to_hpo_observation_list_d = defaultdict(list)\n",
    "hpo_observed = 0\n",
    "hpo_excluded = 0\n",
    "for hpo_term_id, row in df_hpo.iterrows():\n",
    "    # deal with multiple terms, e.g., HP:0001263; HP:0001249\n",
    "    fields = hpo_term_id.split(\";\")\n",
    "    for term_id in fields:\n",
    "        term_id = term_id.strip()\n",
    "        hpo_term = hpo_cr.get_term_from_id(term_id)\n",
    "        hpo_term_label = hpo_term.label\n",
    "        if hpo_term_label is None or len(hpo_term_label) < 5:\n",
    "            raise ValueError(f\"Could not find HPO term label for {hpo_term_id}\")\n",
    "        for patient_id, item in row.iteritems():\n",
    "            if item == \"yes\":\n",
    "                hterm = HpTerm(label=hpo_term_label, hpo_id=term_id)\n",
    "                hpo_observed += 1\n",
    "            elif item == \"no\":\n",
    "                hterm = HpTerm(label=hpo_term_label, hpo_id=term_id, observed=False)\n",
    "                hpo_excluded += 1\n",
    "            # skip if not yes/no -- in this case the term was not measured\n",
    "            patient_id_to_hpo_observation_list_d[patient_id].append(hterm)\n",
    "print(f\"Got {hpo_observed} observed HPO terms and {hpo_excluded} excluded terms for {len(patient_id_to_hpo_observation_list_d)} individuals\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "264ab841-8100-4b5f-95c7-95c134f814f9",
   "metadata": {},
   "source": [
    "<h2>Variants</h2>\n",
    "<p>All variants in this data set are heterozygous and refer to NM_003075.5.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2e8ac9d5-1ea7-47f0-9a9d-9fa52aa15780",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_patient_to_hgvs_d = defaultdict()\n",
    "for idx, item in variant_hgvs_pd_series.iteritems():\n",
    "    if idx == \"HPO\": continue # skip index column\n",
    "    fields = item.split(\",\")\n",
    "    variant_patient_to_hgvs_d[idx] = fields[0].strip()\n",
    "#for k, v in variant_str_d.items():\n",
    "#    print(f\"\\\"{k}\\\" \\\"{v}\\\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7b7c80cf-5e15-4049-b8f8-5a75613f8db9",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = 'hg38'\n",
    "transcript='NM_003075.5'\n",
    "vv = VariantValidator(genome_build=genome, transcript=transcript)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d6403819-4ad5-4d56-b653-b9e93884e5fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Encoding c.1826T>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.1826T>C/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.(?_-22)_(*1330_?)del\n",
      "Encoding c.1838T>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.1838T>C/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.2074G>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.2074G>C/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.880_881del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.880_881del/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.2686A>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.2686A>G/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.326dup\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.326dup/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.3561del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.3561del/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.888_889del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.888_889del/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.3129del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.3129del/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.1849_1854del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.1849_1854del/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.3094dup\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.3094dup/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.1833+1G>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.1833+1G>T/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.327C>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.327C>G/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.1834-7C>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.1834-7C>G/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.999dup\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.999dup/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.574C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.574C>T/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.1311-1G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.1311-1G>A/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.1327C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.1327C>T/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.1829T>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.1829T>C/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.805C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.805C>T/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.3456_3458del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.3456_3458del/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.3135_3139dup\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.3135_3139dup/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.317+2T>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.317+2T>A/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.3524C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.3524C>T/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.2699A>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.2699A>G/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.2678A>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.2678A>G/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.3279del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.3279del/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.723G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.723G>A/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.1919T>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.1919T>C/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.1833+2T>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.1833+2T>C/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.1094_1097del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.1094_1097del/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.1311-3C>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.1311-3C>G/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.2059C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.2059C>T/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.172C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.172C>T/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.1903T>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.1903T>C/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.1796G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.1796G>A/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.1824_1826del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.1824_1826del/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.640A>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.640A>G/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.1651-2A>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.1651-2A>G/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.2697G>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.2697G>T/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.2639A>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.2639A>G/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.743T>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.743T>C/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.2602G>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.2602G>C/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.230C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.230C>T/NM_003075.5?content-type=application%2Fjson\n",
      "Encoding c.400A>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003075.5%3Ac.400A>G/NM_003075.5?content-type=application%2Fjson\n"
     ]
    }
   ],
   "source": [
    "variant_d = defaultdict(Variant)\n",
    "variant_set = set(variant_patient_to_hgvs_d.values())\n",
    "# Note that c.(?_-22)_(*1330_?)del is chr12-55960344-56192514-DEL\n",
    "for hgvs in variant_set:\n",
    "    print(f\"Encoding {hgvs}\")\n",
    "    if hgvs == \"c.(?_-22)_(*1330_?)del\":\n",
    "        variant = StructuralVariant.chromosomal_deletion(cell_contents=\"chr12-55960344-56192514-DEL\", gene_symbol=\"SMARCC2\",\n",
    "                                                   gene_id=\"SMARCC2\")\n",
    "    else:\n",
    "        variant = vv.encode_hgvs(hgvs)\n",
    "    variant.set_heterozygous()\n",
    "    variant_d[hgvs] = variant"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03dd0f0d-8a1b-464f-a435-04e2ad85cd8f",
   "metadata": {},
   "source": [
    "<H2>Putting it all together</H2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "271d4765-4d5c-43d9-a2a9-dd708a932c75",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We created 65 individual objects\n"
     ]
    }
   ],
   "source": [
    "# patient_id_to_hpo_observation_list_d\n",
    "# variant_str_d and variant_d\n",
    "# pat_id_list\n",
    "# sex_list\n",
    "disease_id =  \"OMIM:618362\"\n",
    "disease_label = \"Coffin-Siris syndrome 8\"\n",
    "individual_list = []\n",
    "for pat_id in pat_id_list:\n",
    "    if pat_id == \"HPO\": continue\n",
    "    hpo_terms = patient_id_to_hpo_observation_list_d.get(pat_id)\n",
    "    sex = sex_pd_series[pat_id]\n",
    "    variant_hgvs = variant_patient_to_hgvs_d.get(pat_id)\n",
    "    if variant_hgvs is None:\n",
    "        raise ValueError(f\"Could not find variant hgvs for patient id \\\"{pat_id}\\\"\")\n",
    "    variant = variant_d.get(variant_hgvs)\n",
    "    if variant is None:\n",
    "        raise ValueError(f\"Could not find variant for {variant_hgvs}\")\n",
    "    indiv = Individual(individual_id=pat_id, \n",
    "                       hpo_terms=hpo_terms, \n",
    "                       sex=sex,\n",
    "                       interpretation_list=[variant.to_ga4gh_variant_interpretation()],\n",
    "                       disease_id=disease_id, \n",
    "                       disease_label=disease_label)\n",
    "    individual_list.append(indiv)\n",
    "print(f\"We created {len(individual_list)} individual objects\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b2053c2-eb61-417a-a4b5-d52b2c5c4e85",
   "metadata": {},
   "source": [
    "<h3>Output</h3>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a83c1f9d-f89e-4ca7-8e1d-466d11f3b65d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "65"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Individual.output_individuals_as_phenopackets(individual_list=individual_list, \n",
    "                                              pmid=pmid, \n",
    "                                              metadata=metadata.to_ga4gh(), \n",
    "                                              outdir=\"phenopackets\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31884d4b-ade0-49b6-90b8-b79ce3c6911d",
   "metadata": {},
   "source": [
    "<h2>Validation</h2>\n",
    "<p>Set up phenopacket-tools as described <a href=\"http://phenopackets.org/phenopacket-tools/stable/cli.html\">here</a></p>\n",
    "<p>Then execute the following command in the terminal: pxf validate *.json</p>\n",
    "<p>No errors were detected.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5846e08b-cceb-4d22-b195-f46174488e38",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pstore_env",
   "language": "python",
   "name": "pstore_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
