{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8612e396-9a71-48ba-9d9e-ef2727df8154",
   "metadata": {},
   "source": [
    "# KDM6B, Rots D et al.\n",
    "\n",
    "Data taken from  [Rots D, The clinical and molecular spectrum of the KDM6B-related neurodevelopmental disorder. Am J Hum Genet. 2023 ](https://pubmed.ncbi.nlm.nih.gov/37196654/)\n",
    "Data extracted from Table S1. Detailed clinical information of the cases with the (likely) pathogenic KDM6B variants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "afa4ac8e-bd70-4590-9107-8640c4d0119b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using pyphetools version 0.8.3\n"
     ]
    }
   ],
   "source": [
    "import phenopackets as PPkt\n",
    "from google.protobuf.json_format import MessageToDict, MessageToJson\n",
    "from google.protobuf.json_format import Parse, ParseDict\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_colwidth', None) # show entire column contents, important!\n",
    "from collections import defaultdict\n",
    "from pyphetools.creation import *\n",
    "from pyphetools.visualization import *\n",
    "import importlib.metadata\n",
    "from IPython.display import HTML, display\n",
    "__version__ = importlib.metadata.version(\"pyphetools\")\n",
    "print(f\"Using pyphetools version {__version__}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "88f3a65a-fef2-4cbe-bdd1-01ccef7f9a90",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = HpoParser()\n",
    "hpo_cr = parser.get_hpo_concept_recognizer()\n",
    "hpo_version = parser.get_version()\n",
    "PMID = \"PMID:37196654\"\n",
    "title = \"The clinical and molecular spectrum of the KDM6B-related neurodevelopmental disorder\"\n",
    "metadata = MetaData(created_by=\"ORCID:0000-0002-0736-9199\", pmid=PMID, pubmed_title=title)\n",
    "metadata.default_versions_with_hpo(version=hpo_version)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e0439b7e-6d67-4191-96b1-c4b90f9635c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel('input/Rots_2023_PMID_37196654.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6c6dfb2e-90f7-453b-ae98-a7127c4fece1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Field</th>\n",
       "      <th>Individual 1</th>\n",
       "      <th>Individual 2</th>\n",
       "      <th>Individual 3</th>\n",
       "      <th>Individual 5</th>\n",
       "      <th>Individual 6</th>\n",
       "      <th>Individual 7</th>\n",
       "      <th>Individual 8</th>\n",
       "      <th>Individual 9</th>\n",
       "      <th>Individual 11</th>\n",
       "      <th>...</th>\n",
       "      <th>Individual 4</th>\n",
       "      <th>Individual 10</th>\n",
       "      <th>Individual 34</th>\n",
       "      <th>Individual 38</th>\n",
       "      <th>Individual 44 (DDD_286674)</th>\n",
       "      <th>Individual 49 (DEASD_0146_001)</th>\n",
       "      <th>Individual 50 (DEASD_0129_001)</th>\n",
       "      <th>Individual 54 (SSC_13675.p1)</th>\n",
       "      <th>Individual 58 (DDD_305030)</th>\n",
       "      <th>Individual 59 (DDD_306396)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sex</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>F</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>F</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>F</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Age, years</td>\n",
       "      <td>16</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>25</td>\n",
       "      <td>13y2m</td>\n",
       "      <td>9y6m</td>\n",
       "      <td>10</td>\n",
       "      <td>6y6m</td>\n",
       "      <td>19</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>4</td>\n",
       "      <td>11y</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>7y3m</td>\n",
       "      <td>8y7m</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cohort type</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>...</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>Research and clinical testing</td>\n",
       "      <td>Research cohort</td>\n",
       "      <td>Research cohort</td>\n",
       "      <td>Research cohort</td>\n",
       "      <td>Research cohort</td>\n",
       "      <td>Research cohort</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mutation (NM_)</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>64</td>\n",
       "      <td>65</td>\n",
       "      <td>66</td>\n",
       "      <td>67</td>\n",
       "      <td>68</td>\n",
       "      <td>69</td>\n",
       "      <td>70</td>\n",
       "      <td>71</td>\n",
       "      <td>72</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cDNA change (ENST00000254846.9 or NM_001080424.2)</td>\n",
       "      <td>c.1014delC</td>\n",
       "      <td>c.1085_1088del</td>\n",
       "      <td>c.654_655del</td>\n",
       "      <td>c.1439dup</td>\n",
       "      <td>c.2598delC</td>\n",
       "      <td>c.4500C&gt;A</td>\n",
       "      <td>c.403C&gt;T\\n\\n</td>\n",
       "      <td>c.4737+1G&gt;A</td>\n",
       "      <td>c.3288_3291delTGAG</td>\n",
       "      <td>...</td>\n",
       "      <td>c.4696C&gt;A</td>\n",
       "      <td>c.3762_3764del</td>\n",
       "      <td>c.4118T&gt;C</td>\n",
       "      <td>c.4193C&gt;A</td>\n",
       "      <td>c.4724G&gt;C</td>\n",
       "      <td>c.4174G&gt;A</td>\n",
       "      <td>c.4186T&gt;A</td>\n",
       "      <td>c.4187_4189del</td>\n",
       "      <td>c.4187_4189del</td>\n",
       "      <td>c.4222T&gt;C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 74 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Field      Individual 1  \\\n",
       "0                                                Sex                 F   \n",
       "1                                         Age, years                16   \n",
       "2                                        Cohort type  Clinical testing   \n",
       "3                                     Mutation (NM_)                 1   \n",
       "4  cDNA change (ENST00000254846.9 or NM_001080424.2)        c.1014delC   \n",
       "\n",
       "       Individual 2      Individual 3      Individual 5      Individual 6  \\\n",
       "0                 F                 M                 M                 M   \n",
       "1                10                 9                25             13y2m   \n",
       "2  Clinical testing  Clinical testing  Clinical testing  Clinical testing   \n",
       "3                 2                 3                 4                 5   \n",
       "4    c.1085_1088del      c.654_655del         c.1439dup        c.2598delC   \n",
       "\n",
       "       Individual 7      Individual 8      Individual 9       Individual 11  \\\n",
       "0                 F                 M                 M                   F   \n",
       "1              9y6m                10              6y6m                  19   \n",
       "2  Clinical testing  Clinical testing  Clinical testing    Clinical testing   \n",
       "3                 6                 7                 8                   9   \n",
       "4         c.4500C>A      c.403C>T\\n\\n       c.4737+1G>A  c.3288_3291delTGAG   \n",
       "\n",
       "   ...      Individual 4     Individual 10     Individual 34  \\\n",
       "0  ...                 M                 M                 M   \n",
       "1  ...                14                 4               11y   \n",
       "2  ...  Clinical testing  Clinical testing  Clinical testing   \n",
       "3  ...                64                65                66   \n",
       "4  ...         c.4696C>A    c.3762_3764del         c.4118T>C   \n",
       "\n",
       "      Individual 38     Individual 44 (DDD_286674)  \\\n",
       "0                 M                              F   \n",
       "1                 6                              3   \n",
       "2  Clinical testing  Research and clinical testing   \n",
       "3                67                             68   \n",
       "4         c.4193C>A                      c.4724G>C   \n",
       "\n",
       "  Individual 49 (DEASD_0146_001) Individual 50 (DEASD_0129_001)  \\\n",
       "0                              M                              M   \n",
       "1                           7y3m                           8y7m   \n",
       "2                Research cohort                Research cohort   \n",
       "3                             69                             70   \n",
       "4                      c.4174G>A                      c.4186T>A   \n",
       "\n",
       "  Individual 54 (SSC_13675.p1) Individual 58 (DDD_305030)  \\\n",
       "0                            M                          M   \n",
       "1                          NaN                        NaN   \n",
       "2              Research cohort            Research cohort   \n",
       "3                           71                         72   \n",
       "4               c.4187_4189del             c.4187_4189del   \n",
       "\n",
       "  Individual 59 (DDD_306396)  \n",
       "0                          M  \n",
       "1                        NaN  \n",
       "2            Research cohort  \n",
       "3                         73  \n",
       "4                  c.4222T>C  \n",
       "\n",
       "[5 rows x 74 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b41a3774-6832-41db-b9c4-f6bcfeddbfdd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Field</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age, years</th>\n",
       "      <th>Cohort type</th>\n",
       "      <th>Mutation (NM_)</th>\n",
       "      <th>cDNA change (ENST00000254846.9 or NM_001080424.2)</th>\n",
       "      <th>Amino acid change</th>\n",
       "      <th>Variant Type (PTV or PAV)</th>\n",
       "      <th>Inheritance</th>\n",
       "      <th>Heterozygous/Homozygous</th>\n",
       "      <th>Additional findings of genetic testing</th>\n",
       "      <th>...</th>\n",
       "      <th>Other</th>\n",
       "      <th>Other</th>\n",
       "      <th>Skin hyperlaxity</th>\n",
       "      <th>Genitourinary abnormalities</th>\n",
       "      <th>Cryptorchidism</th>\n",
       "      <th>Other medication received</th>\n",
       "      <th>Other</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Individual 1</th>\n",
       "      <td>F</td>\n",
       "      <td>16</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>1</td>\n",
       "      <td>c.1014delC</td>\n",
       "      <td>p.(Arg340Alafs*147)</td>\n",
       "      <td>PTV</td>\n",
       "      <td>Maternal</td>\n",
       "      <td>Heterozygous</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not reported</td>\n",
       "      <td>Nasal speech</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Individual 2</th>\n",
       "      <td>F</td>\n",
       "      <td>10</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>2</td>\n",
       "      <td>c.1085_1088del</td>\n",
       "      <td>p.(Glu362Alafs*124)</td>\n",
       "      <td>PTV</td>\n",
       "      <td>Paternal</td>\n",
       "      <td>Heterozygous</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>2 Cafe-au-lait spots</td>\n",
       "      <td>Night incontinence</td>\n",
       "      <td>commonly head and abdominal pain</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Individual 3</th>\n",
       "      <td>M</td>\n",
       "      <td>9</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>3</td>\n",
       "      <td>c.654_655del</td>\n",
       "      <td>p.(Glu220Glyfs*16)</td>\n",
       "      <td>PTV</td>\n",
       "      <td>de novo</td>\n",
       "      <td>Heterozygous</td>\n",
       "      <td>Beta-thalasemia carrier</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Melatonine(sleep problems); Prednisolon, budesonide, salbutamol (asthma); macrogol (constipations); esomeprazole (GERD)</td>\n",
       "      <td>1 Cafe-au-lait spot</td>\n",
       "      <td>Adenotomy due to the hyperplasia</td>\n",
       "      <td>Bronchial asthma</td>\n",
       "      <td>Verry common airway infections</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Individual 5</th>\n",
       "      <td>M</td>\n",
       "      <td>25</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>4</td>\n",
       "      <td>c.1439dup</td>\n",
       "      <td>p.(Pro481Thrfs*29)</td>\n",
       "      <td>PTV</td>\n",
       "      <td>de novo</td>\n",
       "      <td>Heterozygous</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>Eats/drinks no cow's milk, no gluten and little soya. No allergy but seems sensitive to these products</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Vitamins and feeding supplements through alternative doctor</td>\n",
       "      <td>At metabolic screening increased essential amino acids alanin amongst others</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Individual 6</th>\n",
       "      <td>M</td>\n",
       "      <td>13y2m</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>5</td>\n",
       "      <td>c.2598delC</td>\n",
       "      <td>p.(Ser867Argfs*27)</td>\n",
       "      <td>PTV</td>\n",
       "      <td>de novo</td>\n",
       "      <td>Heterozygous</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>phimosis</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>tongue frenulum IQ because of dyslalia. Double appical hair whorl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 77 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Field        Sex Age, years       Cohort type Mutation (NM_)  \\\n",
       "Individual 1   F         16  Clinical testing              1   \n",
       "Individual 2   F         10  Clinical testing              2   \n",
       "Individual 3   M          9  Clinical testing              3   \n",
       "Individual 5   M         25  Clinical testing              4   \n",
       "Individual 6   M      13y2m  Clinical testing              5   \n",
       "\n",
       "Field        cDNA change (ENST00000254846.9 or NM_001080424.2)  \\\n",
       "Individual 1                                        c.1014delC   \n",
       "Individual 2                                    c.1085_1088del   \n",
       "Individual 3                                      c.654_655del   \n",
       "Individual 5                                         c.1439dup   \n",
       "Individual 6                                        c.2598delC   \n",
       "\n",
       "Field           Amino acid change Variant Type (PTV or PAV) Inheritance  \\\n",
       "Individual 1  p.(Arg340Alafs*147)                       PTV    Maternal   \n",
       "Individual 2  p.(Glu362Alafs*124)                       PTV    Paternal   \n",
       "Individual 3   p.(Glu220Glyfs*16)                       PTV     de novo   \n",
       "Individual 5   p.(Pro481Thrfs*29)                       PTV     de novo   \n",
       "Individual 6   p.(Ser867Argfs*27)                       PTV     de novo   \n",
       "\n",
       "Field        Heterozygous/Homozygous Additional findings of genetic testing  \\\n",
       "Individual 1            Heterozygous                                     No   \n",
       "Individual 2            Heterozygous                                     No   \n",
       "Individual 3            Heterozygous                Beta-thalasemia carrier   \n",
       "Individual 5            Heterozygous                                     No   \n",
       "Individual 6            Heterozygous                                     No   \n",
       "\n",
       "Field         ...  \\\n",
       "Individual 1  ...   \n",
       "Individual 2  ...   \n",
       "Individual 3  ...   \n",
       "Individual 5  ...   \n",
       "Individual 6  ...   \n",
       "\n",
       "Field                                                                                                          Other  \\\n",
       "Individual 1                                                                                                     NaN   \n",
       "Individual 2                                                                                                     NaN   \n",
       "Individual 3                                                                                                     NaN   \n",
       "Individual 5  Eats/drinks no cow's milk, no gluten and little soya. No allergy but seems sensitive to these products   \n",
       "Individual 6                                                                                                     NaN   \n",
       "\n",
       "Field        Other Skin hyperlaxity Genitourinary abnormalities  \\\n",
       "Individual 1   NaN               No                          No   \n",
       "Individual 2   NaN               No                          No   \n",
       "Individual 3   NaN               No                          No   \n",
       "Individual 5   NaN               No                          No   \n",
       "Individual 6   NaN               No                    phimosis   \n",
       "\n",
       "Field        Cryptorchidism  \\\n",
       "Individual 1            NaN   \n",
       "Individual 2            NaN   \n",
       "Individual 3             No   \n",
       "Individual 5             No   \n",
       "Individual 6             No   \n",
       "\n",
       "Field                                                                                                       Other medication received  \\\n",
       "Individual 1                                                                                                             Not reported   \n",
       "Individual 2                                                                                                                       No   \n",
       "Individual 3  Melatonine(sleep problems); Prednisolon, budesonide, salbutamol (asthma); macrogol (constipations); esomeprazole (GERD)   \n",
       "Individual 5                                                              Vitamins and feeding supplements through alternative doctor   \n",
       "Individual 6                                                                                                                       No   \n",
       "\n",
       "Field                                                                                Other  \\\n",
       "Individual 1                                                                  Nasal speech   \n",
       "Individual 2                                                          2 Cafe-au-lait spots   \n",
       "Individual 3                                                           1 Cafe-au-lait spot   \n",
       "Individual 5  At metabolic screening increased essential amino acids alanin amongst others   \n",
       "Individual 6             tongue frenulum IQ because of dyslalia. Double appical hair whorl   \n",
       "\n",
       "Field                                      NaN  \\\n",
       "Individual 1                               NaN   \n",
       "Individual 2                Night incontinence   \n",
       "Individual 3  Adenotomy due to the hyperplasia   \n",
       "Individual 5                               NaN   \n",
       "Individual 6                               NaN   \n",
       "\n",
       "Field                                      NaN                             NaN  \n",
       "Individual 1                               NaN                             NaN  \n",
       "Individual 2  commonly head and abdominal pain                             NaN  \n",
       "Individual 3                  Bronchial asthma  Verry common airway infections  \n",
       "Individual 5                               NaN                             NaN  \n",
       "Individual 6                               NaN                             NaN  \n",
       "\n",
       "[5 rows x 77 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft = df.transpose()\n",
    "dft.columns = dft.iloc[0]\n",
    "dft.drop(dft.index[0], inplace=True)\n",
    "dft.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ca522a2e-b23c-42ae-80dd-9300864d3821",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Field</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Age, years</th>\n",
       "      <th>Cohort type</th>\n",
       "      <th>Mutation (NM_)</th>\n",
       "      <th>cDNA change (ENST00000254846.9 or NM_001080424.2)</th>\n",
       "      <th>Amino acid change</th>\n",
       "      <th>Variant Type (PTV or PAV)</th>\n",
       "      <th>Inheritance</th>\n",
       "      <th>Heterozygous/Homozygous</th>\n",
       "      <th>Additional findings of genetic testing</th>\n",
       "      <th>...</th>\n",
       "      <th>Other</th>\n",
       "      <th>Skin hyperlaxity</th>\n",
       "      <th>Genitourinary abnormalities</th>\n",
       "      <th>Cryptorchidism</th>\n",
       "      <th>Other medication received</th>\n",
       "      <th>Other</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>NaN</th>\n",
       "      <th>patient_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Individual 1</th>\n",
       "      <td>F</td>\n",
       "      <td>16</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>1</td>\n",
       "      <td>c.1014delC</td>\n",
       "      <td>p.(Arg340Alafs*147)</td>\n",
       "      <td>PTV</td>\n",
       "      <td>Maternal</td>\n",
       "      <td>Heterozygous</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not reported</td>\n",
       "      <td>Nasal speech</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Individual 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Individual 2</th>\n",
       "      <td>F</td>\n",
       "      <td>10</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>2</td>\n",
       "      <td>c.1085_1088del</td>\n",
       "      <td>p.(Glu362Alafs*124)</td>\n",
       "      <td>PTV</td>\n",
       "      <td>Paternal</td>\n",
       "      <td>Heterozygous</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>2 Cafe-au-lait spots</td>\n",
       "      <td>Night incontinence</td>\n",
       "      <td>commonly head and abdominal pain</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Individual 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Individual 3</th>\n",
       "      <td>M</td>\n",
       "      <td>9</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>3</td>\n",
       "      <td>c.654_655del</td>\n",
       "      <td>p.(Glu220Glyfs*16)</td>\n",
       "      <td>PTV</td>\n",
       "      <td>de novo</td>\n",
       "      <td>Heterozygous</td>\n",
       "      <td>Beta-thalasemia carrier</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Melatonine(sleep problems); Prednisolon, budesonide, salbutamol (asthma); macrogol (constipations); esomeprazole (GERD)</td>\n",
       "      <td>1 Cafe-au-lait spot</td>\n",
       "      <td>Adenotomy due to the hyperplasia</td>\n",
       "      <td>Bronchial asthma</td>\n",
       "      <td>Verry common airway infections</td>\n",
       "      <td>Individual 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Individual 5</th>\n",
       "      <td>M</td>\n",
       "      <td>25</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>4</td>\n",
       "      <td>c.1439dup</td>\n",
       "      <td>p.(Pro481Thrfs*29)</td>\n",
       "      <td>PTV</td>\n",
       "      <td>de novo</td>\n",
       "      <td>Heterozygous</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Vitamins and feeding supplements through alternative doctor</td>\n",
       "      <td>At metabolic screening increased essential amino acids alanin amongst others</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Individual 5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Individual 6</th>\n",
       "      <td>M</td>\n",
       "      <td>13y2m</td>\n",
       "      <td>Clinical testing</td>\n",
       "      <td>5</td>\n",
       "      <td>c.2598delC</td>\n",
       "      <td>p.(Ser867Argfs*27)</td>\n",
       "      <td>PTV</td>\n",
       "      <td>de novo</td>\n",
       "      <td>Heterozygous</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>phimosis</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>tongue frenulum IQ because of dyslalia. Double appical hair whorl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Individual 6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 78 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Field        Sex Age, years       Cohort type Mutation (NM_)  \\\n",
       "Individual 1   F         16  Clinical testing              1   \n",
       "Individual 2   F         10  Clinical testing              2   \n",
       "Individual 3   M          9  Clinical testing              3   \n",
       "Individual 5   M         25  Clinical testing              4   \n",
       "Individual 6   M      13y2m  Clinical testing              5   \n",
       "\n",
       "Field        cDNA change (ENST00000254846.9 or NM_001080424.2)  \\\n",
       "Individual 1                                        c.1014delC   \n",
       "Individual 2                                    c.1085_1088del   \n",
       "Individual 3                                      c.654_655del   \n",
       "Individual 5                                         c.1439dup   \n",
       "Individual 6                                        c.2598delC   \n",
       "\n",
       "Field           Amino acid change Variant Type (PTV or PAV) Inheritance  \\\n",
       "Individual 1  p.(Arg340Alafs*147)                       PTV    Maternal   \n",
       "Individual 2  p.(Glu362Alafs*124)                       PTV    Paternal   \n",
       "Individual 3   p.(Glu220Glyfs*16)                       PTV     de novo   \n",
       "Individual 5   p.(Pro481Thrfs*29)                       PTV     de novo   \n",
       "Individual 6   p.(Ser867Argfs*27)                       PTV     de novo   \n",
       "\n",
       "Field        Heterozygous/Homozygous Additional findings of genetic testing  \\\n",
       "Individual 1            Heterozygous                                     No   \n",
       "Individual 2            Heterozygous                                     No   \n",
       "Individual 3            Heterozygous                Beta-thalasemia carrier   \n",
       "Individual 5            Heterozygous                                     No   \n",
       "Individual 6            Heterozygous                                     No   \n",
       "\n",
       "Field         ... Other Skin hyperlaxity Genitourinary abnormalities  \\\n",
       "Individual 1  ...   NaN               No                          No   \n",
       "Individual 2  ...   NaN               No                          No   \n",
       "Individual 3  ...   NaN               No                          No   \n",
       "Individual 5  ...   NaN               No                          No   \n",
       "Individual 6  ...   NaN               No                    phimosis   \n",
       "\n",
       "Field        Cryptorchidism  \\\n",
       "Individual 1            NaN   \n",
       "Individual 2            NaN   \n",
       "Individual 3             No   \n",
       "Individual 5             No   \n",
       "Individual 6             No   \n",
       "\n",
       "Field                                                                                                       Other medication received  \\\n",
       "Individual 1                                                                                                             Not reported   \n",
       "Individual 2                                                                                                                       No   \n",
       "Individual 3  Melatonine(sleep problems); Prednisolon, budesonide, salbutamol (asthma); macrogol (constipations); esomeprazole (GERD)   \n",
       "Individual 5                                                              Vitamins and feeding supplements through alternative doctor   \n",
       "Individual 6                                                                                                                       No   \n",
       "\n",
       "Field                                                                                Other  \\\n",
       "Individual 1                                                                  Nasal speech   \n",
       "Individual 2                                                          2 Cafe-au-lait spots   \n",
       "Individual 3                                                           1 Cafe-au-lait spot   \n",
       "Individual 5  At metabolic screening increased essential amino acids alanin amongst others   \n",
       "Individual 6             tongue frenulum IQ because of dyslalia. Double appical hair whorl   \n",
       "\n",
       "Field                                      NaN  \\\n",
       "Individual 1                               NaN   \n",
       "Individual 2                Night incontinence   \n",
       "Individual 3  Adenotomy due to the hyperplasia   \n",
       "Individual 5                               NaN   \n",
       "Individual 6                               NaN   \n",
       "\n",
       "Field                                      NaN  \\\n",
       "Individual 1                               NaN   \n",
       "Individual 2  commonly head and abdominal pain   \n",
       "Individual 3                  Bronchial asthma   \n",
       "Individual 5                               NaN   \n",
       "Individual 6                               NaN   \n",
       "\n",
       "Field                                    NaN    patient_id  \n",
       "Individual 1                             NaN  Individual 1  \n",
       "Individual 2                             NaN  Individual 2  \n",
       "Individual 3  Verry common airway infections  Individual 3  \n",
       "Individual 5                             NaN  Individual 5  \n",
       "Individual 6                             NaN  Individual 6  \n",
       "\n",
       "[5 rows x 78 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft['patient_id'] = dft.index  # Set the new column 'patient_id' to be identical to the contents of the index\n",
    "dft.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "16a36bc8-1ad8-421a-ba39-8d1e45dcd9c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "generator = SimpleColumnMapperGenerator(df=dft, observed=\"Yes\", excluded=\"No\", hpo_cr=hpo_cr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a6c7acc5-9486-4272-9f34-6df9190e38a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "column_d = generator.try_mapping_columns()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "235ca20f-294d-4506-9af2-a319d77d89c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table style=\"border: 2px solid black;\">\n",
       "\n",
       "<tr>\n",
       "            <th>Result</th>\n",
       "            <th>Columns</th>\n",
       "        </tr>\n",
       "      \n",
       "<tr><td>Mapped</td><td>Motor delay; Intellectual disability; Autism spectrum disorder; Sleep disturbances; Hypotonia; Spasticity; Joint hypermobility; Syndactyly; Pectus excavatum; Strabismus; Recurrent ear infections; Constipation; Cryptorchidism</td></tr>\n",
       "<tr><td>Unmapped</td><td><q>Sex</q>; <q>Age, years</q>; <q>Cohort type</q>; <q>Mutation (NM_)</q>; <q>cDNA change (ENST00000254846.9 or NM_001080424.2)</q>; <q>Amino acid change</q>; <q>Variant Type (PTV or PAV)</q>; <q>Inheritance</q>; <q>Heterozygous/Homozygous</q>; <q>Additional findings of genetic testing</q>; <q>Other affected relatives</q>; <q>Pregnancy/delivery</q>; <q>Complications of Pregnancy/Delivery</q>; <q>Gestational age, weeks</q>; <q>Birth weight, g (SD)</q>; <q>Growth</q>; <q>Height, cm (SD)</q>; <q>Weight, kg(SD)</q>; <q>Head circumference, cm(SD)</q>; <q>Age at folow-up/measurements, years</q>; <q>Neurodevelopment</q>; <q>Language/speech delay</q>; <q>First words, months</q>; <q>First steps, months</q>; <q>IQ profile</q>; <q>nan</q>; <q>nan</q>; <q>Behavior problems</q>; <q>Psychosis / Schizophrenia</q>; <q>Use of psychiatric drugs</q>; <q>Other</q>; <q>Neurological</q>; <q>Seizures / Epilepsy</q>; <q>Dystonia, if present - type and age of onset</q>; <q>Other neurological/movement issues</q>; <q>Brain MRI findings</q>; <q>Musculoskeletal/extremities</q>; <q>Vertebral abnormalities (Scoliosis, kyphosis etc).</q>; <q>Hand /foot/ finger abnormalities</q>; <q>Other</q>; <q>Dysmorphism</q>; <q>Dysmorphic features</q>; <q>Lip/palate cleft</q>; <q>Eyes/visual problems</q>; <q>Hypermetropia/myopia</q>; <q>Other</q>; <q>Ear/ hearing problems</q>; <q>Hearing</q>; <q>Other</q>; <q>Cardiovascular</q>; <q>Congenital heart disease</q>; <q>Other</q>; <q>Gastrointestinal</q>; <q>Neonatal feeding difficulties</q>; <q>Yes</q>; <q>Other</q>; <q>Other</q>; <q>Skin hyperlaxity</q>; <q>Genitourinary abnormalities</q>; <q>Other medication received</q>; <q>Other</q>; <q>nan</q>; <q>nan</q>; <q>nan</q>; <q>patient_id</q></td></tr>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(HTML(generator.to_html()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "8ad7183c-4200-462c-8c5d-2629c1f99b3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Bossy, Agressive (verbally)', 'No',\n",
       "       'Irritability, Anger, anxiety (associated with obstipation periods); No contact except parents and physician',\n",
       "       'Yes', 'ADHD', 'AHDS; Aggression, problems in social interaction ',\n",
       "       'Tantrums and inattention',\n",
       "       'ADHD, aggression, problems in social interaction ',\n",
       "       'AHDS, aggressive, impulsive behaivior ', 'Anxiety, aggression',\n",
       "       'probable ADHD', 'Stubborn, \"lazy\"', 'Agitation,agressivity',\n",
       "       'Aggressive behavior, noncompliance, physical aggression, poor play skills ',\n",
       "       'Anxiety',\n",
       "       'Early and atypical depression, attention deficit, anxiety, atypical sensory',\n",
       "       'Yes, Aggression', 'ADHD, anxiety',\n",
       "       'Poor social skills, stereotypic behaviour.',\n",
       "       'stubborn, aggressive, tantrums', 'Short attention span',\n",
       "       'Rigid behaviour', 'ADHD, aggressive behavior', nan,\n",
       "       'hyperactivity', 'attention deficit', ' impulsive', 'ADHD  ',\n",
       "       'Hyperactivity'], dtype=object)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft[\"Behavior problems\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "efcc72f2-9e4f-491e-ac0f-027c430a29fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>terms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HP:0000718 (Aggressive behavior/observed)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HP:0000708 (Atypical behavior/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HP:0000708 (Atypical behavior/observed)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>HP:0000708 (Atypical behavior/excluded)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>n/a</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        terms\n",
       "0   HP:0000718 (Aggressive behavior/observed)\n",
       "1     HP:0000708 (Atypical behavior/excluded)\n",
       "2                                         n/a\n",
       "3     HP:0000708 (Atypical behavior/observed)\n",
       "4                                         n/a\n",
       "..                                        ...\n",
       "68    HP:0000708 (Atypical behavior/excluded)\n",
       "69                                        n/a\n",
       "70                                        n/a\n",
       "71                                        n/a\n",
       "72                                        n/a\n",
       "\n",
       "[73 rows x 1 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "option_d = {'Bossy, Agressive (verbally)':'Aggressive behavior',\n",
    "           'Yes': 'Atypical behavior'}\n",
    "excluded_d = {'No':'Atypical behavior'}\n",
    "behaviorMapper = OptionColumnMapper(concept_recognizer=hpo_cr, option_d=option_d, excluded_d=excluded_d)\n",
    "behaviorMapper.preview_column(dft[\"Behavior problems\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "61a187af-1f91-4408-a5eb-76d8822e966f",
   "metadata": {},
   "outputs": [],
   "source": [
    "result = OptionColumnMapper.autoformat(df=dft, concept_recognizer=hpo_cr, delimiter=\";,\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "48393b10-b4cb-499f-80e0-ccf28028d525",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['F', 'M'], dtype=object)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft['Sex'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "519d5fb5-55bb-4d4b-bcfe-b9062342ac34",
   "metadata": {},
   "outputs": [],
   "source": [
    "sexMapper = SexColumnMapper(male_symbol=\"M\", female_symbol=\"F\", column_name=\"Sex\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b4c395e7-c599-4879-923a-e908a29550f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>original column contents</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>F</td>\n",
       "      <td>FEMALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>F</td>\n",
       "      <td>FEMALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>M</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>M</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>M</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>M</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>M</td>\n",
       "      <td>MALE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   original column contents     sex\n",
       "0                         F  FEMALE\n",
       "1                         F  FEMALE\n",
       "2                         M    MALE\n",
       "3                         M    MALE\n",
       "4                         M    MALE\n",
       "..                      ...     ...\n",
       "68                        M    MALE\n",
       "69                        M    MALE\n",
       "70                        M    MALE\n",
       "71                        M    MALE\n",
       "72                        M    MALE\n",
       "\n",
       "[73 rows x 2 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sexMapper.preview_column(dft['Sex'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "5cf19ef8-a24c-4d4a-bd8a-aa1ee8dd086d",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_list = dft[\"cDNA change (ENST00000254846.9 or NM_001080424.2)\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "88ea5b74-24f2-4b93-b25d-b9222832f5cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001080424.2%3Ac.1014delC/NM_001080424.2?content-type=application%2Fjson\n"
     ]
    },
    {
     "ename": "ConnectionError",
     "evalue": "HTTPSConnectionPool(host='rest.variantvalidator.org', port=443): Max retries exceeded with url: /VariantValidator/variantvalidator/hg38/NM_001080424.2%3Ac.1014delC/NM_001080424.2?content-type=application%2Fjson (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x7ff97cfa8f40>: Failed to resolve 'rest.variantvalidator.org' ([Errno -3] Temporary failure in name resolution)\"))",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mgaierror\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/GIT/phenopacket-store/ps_venv/lib/python3.10/site-packages/urllib3/connection.py:203\u001b[0m, in \u001b[0;36mHTTPConnection._new_conn\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    202\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 203\u001b[0m     sock \u001b[38;5;241m=\u001b[39m \u001b[43mconnection\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcreate_connection\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    204\u001b[0m \u001b[43m        \u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_dns_host\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mport\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    205\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    206\u001b[0m \u001b[43m        \u001b[49m\u001b[43msource_address\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msource_address\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    207\u001b[0m \u001b[43m        \u001b[49m\u001b[43msocket_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msocket_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    208\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    209\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m socket\u001b[38;5;241m.\u001b[39mgaierror \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/ps_venv/lib/python3.10/site-packages/urllib3/util/connection.py:60\u001b[0m, in \u001b[0;36mcreate_connection\u001b[0;34m(address, timeout, source_address, socket_options)\u001b[0m\n\u001b[1;32m     58\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m LocationParseError(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mhost\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m, label empty or too long\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m---> 60\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m res \u001b[38;5;129;01min\u001b[39;00m \u001b[43msocket\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetaddrinfo\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhost\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mport\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfamily\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msocket\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mSOCK_STREAM\u001b[49m\u001b[43m)\u001b[49m:\n\u001b[1;32m     61\u001b[0m     af, socktype, proto, canonname, sa \u001b[38;5;241m=\u001b[39m res\n",
      "File \u001b[0;32m/usr/lib/python3.10/socket.py:955\u001b[0m, in \u001b[0;36mgetaddrinfo\u001b[0;34m(host, port, family, type, proto, flags)\u001b[0m\n\u001b[1;32m    954\u001b[0m addrlist \u001b[38;5;241m=\u001b[39m []\n\u001b[0;32m--> 955\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m res \u001b[38;5;129;01min\u001b[39;00m \u001b[43m_socket\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetaddrinfo\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhost\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mport\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfamily\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mtype\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mproto\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mflags\u001b[49m\u001b[43m)\u001b[49m:\n\u001b[1;32m    956\u001b[0m     af, socktype, proto, canonname, sa \u001b[38;5;241m=\u001b[39m res\n",
      "\u001b[0;31mgaierror\u001b[0m: [Errno -3] Temporary failure in name resolution",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mNameResolutionError\u001b[0m                       Traceback (most recent call last)",
      "File \u001b[0;32m~/GIT/phenopacket-store/ps_venv/lib/python3.10/site-packages/urllib3/connectionpool.py:790\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, preload_content, decode_content, **response_kw)\u001b[0m\n\u001b[1;32m    789\u001b[0m \u001b[38;5;66;03m# Make the request on the HTTPConnection object\u001b[39;00m\n\u001b[0;32m--> 790\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    791\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    792\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    793\u001b[0m \u001b[43m    \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    794\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout_obj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    795\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    796\u001b[0m \u001b[43m    \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    797\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    798\u001b[0m \u001b[43m    \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mretries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    799\u001b[0m \u001b[43m    \u001b[49m\u001b[43mresponse_conn\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mresponse_conn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    800\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpreload_content\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    801\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdecode_content\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    802\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mresponse_kw\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    803\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    805\u001b[0m \u001b[38;5;66;03m# Everything went great!\u001b[39;00m\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/ps_venv/lib/python3.10/site-packages/urllib3/connectionpool.py:491\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, body, headers, retries, timeout, chunked, response_conn, preload_content, decode_content, enforce_content_length)\u001b[0m\n\u001b[1;32m    490\u001b[0m         new_e \u001b[38;5;241m=\u001b[39m _wrap_proxy_error(new_e, conn\u001b[38;5;241m.\u001b[39mproxy\u001b[38;5;241m.\u001b[39mscheme)\n\u001b[0;32m--> 491\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m new_e\n\u001b[1;32m    493\u001b[0m \u001b[38;5;66;03m# conn.request() calls http.client.*.request, not the method in\u001b[39;00m\n\u001b[1;32m    494\u001b[0m \u001b[38;5;66;03m# urllib3.request. It also calls makefile (recv) on the socket.\u001b[39;00m\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/ps_venv/lib/python3.10/site-packages/urllib3/connectionpool.py:467\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, body, headers, retries, timeout, chunked, response_conn, preload_content, decode_content, enforce_content_length)\u001b[0m\n\u001b[1;32m    466\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 467\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_conn\u001b[49m\u001b[43m(\u001b[49m\u001b[43mconn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    468\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (SocketTimeout, BaseSSLError) \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/ps_venv/lib/python3.10/site-packages/urllib3/connectionpool.py:1092\u001b[0m, in \u001b[0;36mHTTPSConnectionPool._validate_conn\u001b[0;34m(self, conn)\u001b[0m\n\u001b[1;32m   1091\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m conn\u001b[38;5;241m.\u001b[39mis_closed:\n\u001b[0;32m-> 1092\u001b[0m     \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1094\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m conn\u001b[38;5;241m.\u001b[39mis_verified:\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/ps_venv/lib/python3.10/site-packages/urllib3/connection.py:611\u001b[0m, in \u001b[0;36mHTTPSConnection.connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    610\u001b[0m sock: socket\u001b[38;5;241m.\u001b[39msocket \u001b[38;5;241m|\u001b[39m ssl\u001b[38;5;241m.\u001b[39mSSLSocket\n\u001b[0;32m--> 611\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msock \u001b[38;5;241m=\u001b[39m sock \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_new_conn\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    612\u001b[0m server_hostname: \u001b[38;5;28mstr\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhost\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/ps_venv/lib/python3.10/site-packages/urllib3/connection.py:210\u001b[0m, in \u001b[0;36mHTTPConnection._new_conn\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    209\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m socket\u001b[38;5;241m.\u001b[39mgaierror \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m--> 210\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m NameResolutionError(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhost, \u001b[38;5;28mself\u001b[39m, e) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01me\u001b[39;00m\n\u001b[1;32m    211\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m SocketTimeout \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "\u001b[0;31mNameResolutionError\u001b[0m: <urllib3.connection.HTTPSConnection object at 0x7ff97cfa8f40>: Failed to resolve 'rest.variantvalidator.org' ([Errno -3] Temporary failure in name resolution)",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mMaxRetryError\u001b[0m                             Traceback (most recent call last)",
      "File \u001b[0;32m~/GIT/phenopacket-store/ps_venv/lib/python3.10/site-packages/requests/adapters.py:486\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    485\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 486\u001b[0m     resp \u001b[38;5;241m=\u001b[39m \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43murlopen\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    487\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    488\u001b[0m \u001b[43m        \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    489\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    490\u001b[0m \u001b[43m        \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    491\u001b[0m \u001b[43m        \u001b[49m\u001b[43mredirect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    492\u001b[0m \u001b[43m        \u001b[49m\u001b[43massert_same_host\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    493\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    494\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    495\u001b[0m \u001b[43m        \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmax_retries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    496\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    497\u001b[0m \u001b[43m        \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    498\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    500\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (ProtocolError, \u001b[38;5;167;01mOSError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/ps_venv/lib/python3.10/site-packages/urllib3/connectionpool.py:844\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, preload_content, decode_content, **response_kw)\u001b[0m\n\u001b[1;32m    842\u001b[0m     new_e \u001b[38;5;241m=\u001b[39m ProtocolError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mConnection aborted.\u001b[39m\u001b[38;5;124m\"\u001b[39m, new_e)\n\u001b[0;32m--> 844\u001b[0m retries \u001b[38;5;241m=\u001b[39m \u001b[43mretries\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mincrement\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    845\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43merror\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnew_e\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m_pool\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m_stacktrace\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msys\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexc_info\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m2\u001b[39;49m\u001b[43m]\u001b[49m\n\u001b[1;32m    846\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    847\u001b[0m retries\u001b[38;5;241m.\u001b[39msleep()\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/ps_venv/lib/python3.10/site-packages/urllib3/util/retry.py:515\u001b[0m, in \u001b[0;36mRetry.increment\u001b[0;34m(self, method, url, response, error, _pool, _stacktrace)\u001b[0m\n\u001b[1;32m    514\u001b[0m     reason \u001b[38;5;241m=\u001b[39m error \u001b[38;5;129;01mor\u001b[39;00m ResponseError(cause)\n\u001b[0;32m--> 515\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m MaxRetryError(_pool, url, reason) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mreason\u001b[39;00m  \u001b[38;5;66;03m# type: ignore[arg-type]\u001b[39;00m\n\u001b[1;32m    517\u001b[0m log\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mIncremented Retry for (url=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m): \u001b[39m\u001b[38;5;132;01m%r\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, url, new_retry)\n",
      "\u001b[0;31mMaxRetryError\u001b[0m: HTTPSConnectionPool(host='rest.variantvalidator.org', port=443): Max retries exceeded with url: /VariantValidator/variantvalidator/hg38/NM_001080424.2%3Ac.1014delC/NM_001080424.2?content-type=application%2Fjson (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x7ff97cfa8f40>: Failed to resolve 'rest.variantvalidator.org' ([Errno -3] Temporary failure in name resolution)\"))",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mConnectionError\u001b[0m                           Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[46], line 4\u001b[0m\n\u001b[1;32m      2\u001b[0m variant_d \u001b[38;5;241m=\u001b[39m {}\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m v \u001b[38;5;129;01min\u001b[39;00m var_list:\n\u001b[0;32m----> 4\u001b[0m     vvar \u001b[38;5;241m=\u001b[39m \u001b[43mvvalidator\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencode_hgvs\u001b[49m\u001b[43m(\u001b[49m\u001b[43mv\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      5\u001b[0m     \u001b[38;5;28mprint\u001b[39m(v)\n\u001b[1;32m      6\u001b[0m     \u001b[38;5;28mprint\u001b[39m(vvar)\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/ps_venv/lib/python3.10/site-packages/pyphetools/creation/variant_validator.py:53\u001b[0m, in \u001b[0;36mVariantValidator.encode_hgvs\u001b[0;34m(self, hgvs, custom_transcript)\u001b[0m\n\u001b[1;32m     50\u001b[0m \u001b[38;5;66;03m#f\"https://rest.variantvalidator.org/VariantValidator/variantvalidator/{self._genome_assembly}/{transcript}%3A{hgvs}/{transcript}\"\u001b[39;00m\n\u001b[1;32m     51\u001b[0m \u001b[38;5;66;03m#\u001b[39;00m\n\u001b[1;32m     52\u001b[0m \u001b[38;5;28mprint\u001b[39m(api_url)\n\u001b[0;32m---> 53\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[43mrequests\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43mapi_url\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     54\u001b[0m \u001b[38;5;66;03m# We expect to get a dictionary with three keys. The first is the name of the variant, e.g., ACC:HGVS, then we\u001b[39;00m\n\u001b[1;32m     55\u001b[0m \u001b[38;5;66;03m# get flag and metadata\u001b[39;00m\n\u001b[1;32m     56\u001b[0m vv_dict \u001b[38;5;241m=\u001b[39m response\u001b[38;5;241m.\u001b[39mjson() \n",
      "File \u001b[0;32m~/GIT/phenopacket-store/ps_venv/lib/python3.10/site-packages/requests/api.py:73\u001b[0m, in \u001b[0;36mget\u001b[0;34m(url, params, **kwargs)\u001b[0m\n\u001b[1;32m     62\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget\u001b[39m(url, params\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m     63\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Sends a GET request.\u001b[39;00m\n\u001b[1;32m     64\u001b[0m \n\u001b[1;32m     65\u001b[0m \u001b[38;5;124;03m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     70\u001b[0m \u001b[38;5;124;03m    :rtype: requests.Response\u001b[39;00m\n\u001b[1;32m     71\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 73\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mget\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparams\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/ps_venv/lib/python3.10/site-packages/requests/api.py:59\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[38;5;66;03m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     56\u001b[0m \u001b[38;5;66;03m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;66;03m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m sessions\u001b[38;5;241m.\u001b[39mSession() \u001b[38;5;28;01mas\u001b[39;00m session:\n\u001b[0;32m---> 59\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43msession\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/ps_venv/lib/python3.10/site-packages/requests/sessions.py:589\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    584\u001b[0m send_kwargs \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    585\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m\"\u001b[39m: timeout,\n\u001b[1;32m    586\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m\"\u001b[39m: allow_redirects,\n\u001b[1;32m    587\u001b[0m }\n\u001b[1;32m    588\u001b[0m send_kwargs\u001b[38;5;241m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 589\u001b[0m resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprep\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43msend_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    591\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/ps_venv/lib/python3.10/site-packages/requests/sessions.py:703\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    700\u001b[0m start \u001b[38;5;241m=\u001b[39m preferred_clock()\n\u001b[1;32m    702\u001b[0m \u001b[38;5;66;03m# Send the request\u001b[39;00m\n\u001b[0;32m--> 703\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43madapter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    705\u001b[0m \u001b[38;5;66;03m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    706\u001b[0m elapsed \u001b[38;5;241m=\u001b[39m preferred_clock() \u001b[38;5;241m-\u001b[39m start\n",
      "File \u001b[0;32m~/GIT/phenopacket-store/ps_venv/lib/python3.10/site-packages/requests/adapters.py:519\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    515\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(e\u001b[38;5;241m.\u001b[39mreason, _SSLError):\n\u001b[1;32m    516\u001b[0m         \u001b[38;5;66;03m# This branch is for urllib3 v1.22 and later.\u001b[39;00m\n\u001b[1;32m    517\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m SSLError(e, request\u001b[38;5;241m=\u001b[39mrequest)\n\u001b[0;32m--> 519\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m(e, request\u001b[38;5;241m=\u001b[39mrequest)\n\u001b[1;32m    521\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m ClosedPoolError \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    522\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m(e, request\u001b[38;5;241m=\u001b[39mrequest)\n",
      "\u001b[0;31mConnectionError\u001b[0m: HTTPSConnectionPool(host='rest.variantvalidator.org', port=443): Max retries exceeded with url: /VariantValidator/variantvalidator/hg38/NM_001080424.2%3Ac.1014delC/NM_001080424.2?content-type=application%2Fjson (Caused by NameResolutionError(\"<urllib3.connection.HTTPSConnection object at 0x7ff97cfa8f40>: Failed to resolve 'rest.variantvalidator.org' ([Errno -3] Temporary failure in name resolution)\"))"
     ]
    }
   ],
   "source": [
    "vvalidator = VariantValidator(genome_build=\"hg38\", transcript=\"NM_001080424.2\" )\n",
    "variant_d = {}\n",
    "for v in var_list:\n",
    "    vvar = vvalidator.encode_hgvs(v)\n",
    "    print(v)\n",
    "    print(vvar)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33d8769e-3c46-494e-95fa-5c0adbe60d51",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ps_venv",
   "language": "python",
   "name": "ps_venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
