{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "34284130",
   "metadata": {},
   "source": [
    "<h1>landscape of STXBP1-related disorders </h1>\n",
    "<p>Extract the clinical data from <a href=\"https://pubmed.ncbi.nlm.nih.gov/35190816/\"target=\"__blank\">Xian et al. (2022) Assessing the landscape of STXBP1-related disorders in 534 individuals. Brain.</a>.<p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "0ecc4b45",
   "metadata": {},
   "outputs": [],
   "source": [
    "import phenopackets as php\n",
    "from google.protobuf.json_format import MessageToDict, MessageToJson\n",
    "from google.protobuf.json_format import Parse, ParseDict\n",
    "import pandas as pd\n",
    "from csv import DictReader\n",
    "pd.set_option('display.max_colwidth', None) # show entire column contents, important!\n",
    "from collections import defaultdict\n",
    "import re\n",
    "from pyphetools.creation import *\n",
    "# last tested with pyphetools version 0.2.20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9e8891ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = HpoParser()\n",
    "hpo_cr = parser.get_hpo_concept_recognizer()\n",
    "hpo_version = parser.get_version()\n",
    "metadata = MetaData(created_by=\"ORCID:0000-0002-0736-9199\")\n",
    "metadata.default_versions_with_hpo(version=hpo_version)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9dc8c7f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinical_df = pd.read_table(\"input/brain-2021-00642-File011.tsv\");\n",
    "genotype_df = pd.read_table(\"input/brain-2021-00642-File011-genotype.tsv\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "717ba1d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PatID</th>\n",
       "      <th>Source_Journal</th>\n",
       "      <th>Source_PMID*</th>\n",
       "      <th>Year</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Phenotypic_group**</th>\n",
       "      <th>age_onset_m</th>\n",
       "      <th>age_offset_m</th>\n",
       "      <th>age_eval_y</th>\n",
       "      <th>Base_HPO***</th>\n",
       "      <th>HPO_term</th>\n",
       "      <th>Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STX_18469812_Subject_11</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>18469812</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>M</td>\n",
       "      <td>EOEE</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>HP:0003593</td>\n",
       "      <td>Infantile onset</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STX_18469812_Subject_11</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>18469812</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>M</td>\n",
       "      <td>EOEE</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>HP:0010818</td>\n",
       "      <td>Generalized tonic seizures</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STX_18469812_Subject_11</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>18469812</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>M</td>\n",
       "      <td>EOEE</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>HP:0002069</td>\n",
       "      <td>Generalized tonic-clonic seizures</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STX_18469812_Subject_11</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>18469812</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>M</td>\n",
       "      <td>EOEE</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>HP:0010851</td>\n",
       "      <td>EEG with burst suppression</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STX_18469812_Subject_11</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>18469812</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>M</td>\n",
       "      <td>EOEE</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>HP:0002521</td>\n",
       "      <td>Hypsarrhythmia</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     PatID Source_Journal Source_PMID*    Year Sex   \n",
       "0  STX_18469812_Subject_11      Nat Genet     18469812  2008.0   M  \\\n",
       "1  STX_18469812_Subject_11      Nat Genet     18469812  2008.0   M   \n",
       "2  STX_18469812_Subject_11      Nat Genet     18469812  2008.0   M   \n",
       "3  STX_18469812_Subject_11      Nat Genet     18469812  2008.0   M   \n",
       "4  STX_18469812_Subject_11      Nat Genet     18469812  2008.0   M   \n",
       "\n",
       "  Phenotypic_group**  age_onset_m  age_offset_m  age_eval_y Base_HPO***   \n",
       "0               EOEE          2.0           NaN         8.0  HP:0003593  \\\n",
       "1               EOEE          2.0           NaN         8.0  HP:0010818   \n",
       "2               EOEE          2.0           NaN         8.0  HP:0002069   \n",
       "3               EOEE          2.0           NaN         8.0  HP:0010851   \n",
       "4               EOEE          2.0           NaN         8.0  HP:0002521   \n",
       "\n",
       "                            HPO_term Notes  \n",
       "0                    Infantile onset   NaN  \n",
       "1         Generalized tonic seizures   NaN  \n",
       "2  Generalized tonic-clonic seizures   NaN  \n",
       "3         EEG with burst suppression   NaN  \n",
       "4                     Hypsarrhythmia   NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a938b319",
   "metadata": {},
   "outputs": [],
   "source": [
    "patient_d = defaultdict(list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c644123b",
   "metadata": {},
   "outputs": [],
   "source": [
    "class HpoTerm:\n",
    "    def __init__(self, row):\n",
    "        try: \n",
    "            self.age_onset_m = int(row[\"age_onset_m\"])\n",
    "        except:\n",
    "            self.age_onset_m = None\n",
    "        try:\n",
    "            self.age_offset_m = int(row[\"age_offset_m\"])\n",
    "        except:\n",
    "            self.age_offset_m = None\n",
    "        try:\n",
    "            self.age_eval_y = int(row[\"age_eval_y\"])\n",
    "        except:\n",
    "            self.age_eval_y = None\n",
    "        self.hpo_id = row[\"Base_HPO***\"]\n",
    "        self.hpo_label = row[\"HPO_term\"]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "class PatientRow:\n",
    "    def __init__(self, row):\n",
    "        self.patID = row[\"PatID\"]\n",
    "        self.sex = row[\"Sex\"]\n",
    "        self.phenogroup = row[\"Phenotypic_group**\"]\n",
    "        self.hpo_term = HpoTerm(row=row)\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c657801d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We extracted data on 534 individuals\n"
     ]
    }
   ],
   "source": [
    "with open(\"input/brain-2021-00642-File011.tsv\") as f:\n",
    "    reader = DictReader(f, delimiter=\"\\t\")\n",
    "    for row in reader:\n",
    "        prow = PatientRow(row=row)\n",
    "        patient_d[prow.patID].append(prow)\n",
    "print(f\"We extracted data on {len(patient_d)} individuals\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "55a565a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PatID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Func.refGeneWithVer</th>\n",
       "      <th>Gene.refGeneWithVer</th>\n",
       "      <th>GeneDetail.refGeneWithVer</th>\n",
       "      <th>ExonicFunc.refGeneWithVer</th>\n",
       "      <th>...</th>\n",
       "      <th>Otherinfo3</th>\n",
       "      <th>Otherinfo4</th>\n",
       "      <th>Otherinfo5</th>\n",
       "      <th>Otherinfo7</th>\n",
       "      <th>Otherinfo8</th>\n",
       "      <th>Otherinfo9</th>\n",
       "      <th>Otherinfo10</th>\n",
       "      <th>Otherinfo11</th>\n",
       "      <th>bed</th>\n",
       "      <th>Unnamed: 134</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STX_18469812_Subject_11</td>\n",
       "      <td>9.0</td>\n",
       "      <td>130422313.0</td>\n",
       "      <td>130422313.0</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>STXBP1</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>9</td>\n",
       "      <td>130422313</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>Name=70.695764689</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STX_18469812_Subject_3</td>\n",
       "      <td>9.0</td>\n",
       "      <td>130444768.0</td>\n",
       "      <td>130444768.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>STXBP1</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>9</td>\n",
       "      <td>130444768</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>Name=75.862552050</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STX_18469812_Subject_6</td>\n",
       "      <td>9.0</td>\n",
       "      <td>130425593.0</td>\n",
       "      <td>130425593.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>STXBP1</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>9</td>\n",
       "      <td>130425593</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>Name=89.193399398</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STX_18469812_Subject_7</td>\n",
       "      <td>9.0</td>\n",
       "      <td>130439001.0</td>\n",
       "      <td>130439001.0</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>exonic</td>\n",
       "      <td>STXBP1</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>9</td>\n",
       "      <td>130439001</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>Name=20.833326089</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STX_19557857_Patient_1</td>\n",
       "      <td>9.0</td>\n",
       "      <td>130416076.0</td>\n",
       "      <td>130416076.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>splicing</td>\n",
       "      <td>STXBP1</td>\n",
       "      <td>NM_001032221.3:exon3:c.169+1G&gt;A;NM_003165.3:exon3:c.169+1G&gt;A</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>9</td>\n",
       "      <td>130416076</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 135 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     PatID  Chr        Start          End Ref Alt   \n",
       "0  STX_18469812_Subject_11  9.0  130422313.0  130422313.0   T   A  \\\n",
       "1   STX_18469812_Subject_3  9.0  130444768.0  130444768.0   G   A   \n",
       "2   STX_18469812_Subject_6  9.0  130425593.0  130425593.0   G   A   \n",
       "3   STX_18469812_Subject_7  9.0  130439001.0  130439001.0   T   G   \n",
       "4   STX_19557857_Patient_1  9.0  130416076.0  130416076.0   G   A   \n",
       "\n",
       "  Func.refGeneWithVer Gene.refGeneWithVer   \n",
       "0              exonic              STXBP1  \\\n",
       "1              exonic              STXBP1   \n",
       "2              exonic              STXBP1   \n",
       "3              exonic              STXBP1   \n",
       "4            splicing              STXBP1   \n",
       "\n",
       "                                      GeneDetail.refGeneWithVer   \n",
       "0                                                             .  \\\n",
       "1                                                             .   \n",
       "2                                                             .   \n",
       "3                                                             .   \n",
       "4  NM_001032221.3:exon3:c.169+1G>A;NM_003165.3:exon3:c.169+1G>A   \n",
       "\n",
       "  ExonicFunc.refGeneWithVer  ... Otherinfo3 Otherinfo4 Otherinfo5 Otherinfo7   \n",
       "0         nonsynonymous SNV  ...          .          9  130422313          T  \\\n",
       "1         nonsynonymous SNV  ...          .          9  130444768          G   \n",
       "2         nonsynonymous SNV  ...          .          9  130425593          G   \n",
       "3         nonsynonymous SNV  ...          .          9  130439001          T   \n",
       "4                         .  ...          .          9  130416076          G   \n",
       "\n",
       "  Otherinfo8 Otherinfo9 Otherinfo10 Otherinfo11                bed   \n",
       "0          A          .        PASS           .  Name=70.695764689  \\\n",
       "1          A          .        PASS           .  Name=75.862552050   \n",
       "2          A          .        PASS           .  Name=89.193399398   \n",
       "3          G          .        PASS           .  Name=20.833326089   \n",
       "4          A          .        PASS           .                  .   \n",
       "\n",
       "  Unnamed: 134  \n",
       "0          NaN  \n",
       "1          NaN  \n",
       "2          NaN  \n",
       "3          NaN  \n",
       "4          NaN  \n",
       "\n",
       "[5 rows x 135 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "b667810e",
   "metadata": {},
   "outputs": [],
   "source": [
    "class GenotypeEntry:\n",
    "    def __init__(self, row):\n",
    "        self.patID = row[\"PatID\"]\n",
    "        self.chrom = row[\"Chr\"]\n",
    "        self.start = row[\"Start\"]\n",
    "        self.end = row[\"End\"]\n",
    "        self.ref = row[\"Ref\"]\n",
    "        self.alt = row[\"Alt\"]\n",
    "        func = row[\"Func.refGeneWithVer\"]\n",
    "        genenot = row[\"Gene.refGeneWithVer\"]\n",
    "        aachange = row[\"AAChange.refGeneWithVer\"]\n",
    "        if func == \"exonic\":\n",
    "            fields = aachange.split(\":\")\n",
    "            index = 0\n",
    "            i = 0\n",
    "            for f in fields:\n",
    "                if f == \"NM_001032221.3\":\n",
    "                    index = i\n",
    "                    break\n",
    "                i += 1\n",
    "            if (i + 2) < len(fields):\n",
    "                variant = fields[i+2]\n",
    "            else:\n",
    "                raise ValueError(f\"Could not get variant from row: {row}\")\n",
    "            regex_del = r\"c.\\d+_\\d+del\"\n",
    "            regex_dup = r\"c.(\\d+)dup([A-Z]+)\"\n",
    "            regex_sub = r\"c.([A-Z]+)(\\d+)([A-Z]+)\"\n",
    "            result = re.search(regex_sub, variant)\n",
    "            result_dup = re.search(regex_dup, variant)\n",
    "            if re.match(regex_del, variant):\n",
    "                self.hgvs = variant\n",
    "            elif result:\n",
    "                ref = result.group(1)\n",
    "                position = result.group(2)\n",
    "                alt = result.group(3)\n",
    "                hgvs = f\"c.{position}{ref}>{alt}\"\n",
    "                self.hgvs = hgvs\n",
    "            elif result_dup:\n",
    "                position=result_dup.group(1)\n",
    "                hgvs = f\"c.{position}dup\"\n",
    "                self.hgvs = hgvs\n",
    "            else:\n",
    "                raise ValueError(f\"Could not parse variant {variant}\")\n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "198cbd90",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Could not get variant from row: {'PatID': 'STX_21770924_Patient_3', 'Chr': '9', 'Start': '130446664', 'End': '130446664', 'Ref': 'A', 'Alt': 'C', 'Func.refGeneWithVer': 'exonic', 'Gene.refGeneWithVer': 'STXBP1', 'GeneDetail.refGeneWithVer': '.', 'ExonicFunc.refGeneWithVer': 'nonsynonymous SNV', 'AAChange.refGeneWithVer': 'STXBP1:NM_003165.3:exon19:c.A1720C:p.T574P', 'Func.ensGene': 'exonic', 'Gene.ensGene': 'ENSG00000136854', 'GeneDetail.ensGene': '.', 'ExonicFunc.ensGene': 'nonsynonymous SNV', 'AAChange.ensGene': 'ENSG00000136854:ENST00000373302:exon19:c.A1720C:p.T574P', 'Gnomad_exome_AF': '.', 'Gnomad_exome_AF_popmax': '.', 'Gnomad_exome_AF_male': '.', 'Gnomad_exome_AF_female': '.', 'Gnomad_exome_AF_raw': '.', 'Gnomad_exome_AF_afr': '.', 'Gnomad_exome_AF_sas': '.', 'Gnomad_exome_AF_amr': '.', 'Gnomad_exome_AF_eas': '.', 'Gnomad_exome_AF_nfe': '.', 'Gnomad_exome_AF_fin': '.', 'Gnomad_exome_AF_asj': '.', 'Gnomad_exome_AF_oth': '.', 'non_topmed_Gnomad_exome_AF_popmax': '.', 'non_neuro_Gnomad_exome_AF_popmax': '.', 'non_cancer_Gnomad_exome_AF_popmax': '.', 'controls_Gnomad_exome_AF_popmax': '.', 'PopFreqMax': '.', '1000G_ALL': '.', '1000G_AFR': '.', '1000G_AMR': '.', '1000G_EAS': '.', '1000G_EUR': '.', '1000G_SAS': '.', 'ExAC_ALL': '.', 'ExAC_AFR': '.', 'ExAC_AMR': '.', 'ExAC_EAS': '.', 'ExAC_FIN': '.', 'ExAC_NFE': '.', 'ExAC_OTH': '.', 'ExAC_SAS': '.', 'ESP6500siv2_ALL': '.', 'ESP6500siv2_AA': '.', 'ESP6500siv2_EA': '.', 'CG46': '.', 'CLNALLELEID': '.', 'CLNDN': '.', 'CLNDISDB': '.', 'CLNREVSTAT': '.', 'CLNSIG': '.', 'DamagePredCount': '16.19', 'SIFT_pred': 'D', 'SIFT4G_pred': 'D', 'Polyphen2_HDIV_pred': 'D', 'Polyphen2_HVAR_pred': 'D', 'LRT_pred': 'D', 'MutationTaster_pred': 'D', 'MutationAssessor_pred': '.', 'FATHMM_pred': 'T', 'PROVEAN_pred': 'D', 'VEST4_score': '0.879', 'MetaSVM_pred': 'D', 'MetaLR_pred': 'D', 'M-CAP_pred': 'D', 'REVEL_score': '0.918', 'MutPred_score': '0.652', 'MVP_score': '0.877', 'MPC_score': '2.626', 'PrimateAI_pred': 'T', 'DEOGEN2_pred': 'T', 'BayesDel_addAF_pred': 'D', 'BayesDel_noAF_pred': 'D', 'ClinPred_pred': 'D', 'LIST-S2_pred': 'D', 'CADD_raw': '3.958', 'CADD_phred': '27.3', 'DANN_score': '0.996', 'fathmm-MKL_coding_pred': 'D', 'fathmm-XF_coding_pred': 'D', 'Eigen-raw_coding': '0.847', 'Eigen-phred_coding': '9.779', 'Eigen-PC-raw_coding': '0.799', 'Eigen-PC-phred_coding': '10.102', 'GenoCanyon_score': '1', 'integrated_fitCons_score': '0.497', 'GM12878_fitCons_score': '0.59', 'H1-hESC_fitCons_score': '0.545', 'HUVEC_fitCons_score': '0.492', 'LINSIGHT': '.', 'GERP++_NR': '5.27', 'GERP++_RS': '5.27', 'phyloP100way_vertebrate': '8.914', 'phyloP30way_mammalian': '1.312', 'phyloP17way_primate': '0.756', 'phastCons100way_vertebrate': '1', 'phastCons30way_mammalian': '0.994', 'phastCons17way_primate': '0.994', 'bStatistic': '301', 'Interpro_domain': '.', 'GTEx_V8_gene': '.', 'GTEx_V8_tissue': '.', 'dbscSNV_ADA_SCORE': '.', 'dbscSNV_RF_SCORE': '.', 'score_zscore_gene': '11;0.89;STXBP3:STXBP1:STXBP2', 'hgmd_class': 'NA', 'hgmd_prot': 'NA', 'hgmd_phen': 'NA', 'hgmd_rankscore': 'NA', 'Otherinfo1': '.', 'Otherinfo2': '.', 'Otherinfo3': '.', 'Otherinfo4': '9', 'Otherinfo5': '130446664', 'Otherinfo7': 'A', 'Otherinfo8': 'C', 'Otherinfo9': '.', 'Otherinfo10': 'PASS', 'Otherinfo11': '.', 'bed': 'Name=98.760813298', '': ''}",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[44], line 6\u001b[0m\n\u001b[1;32m      3\u001b[0m reader \u001b[38;5;241m=\u001b[39m DictReader(f, delimiter\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m row \u001b[38;5;129;01min\u001b[39;00m reader:\n\u001b[1;32m      5\u001b[0m     \u001b[38;5;66;03m#print(row)\u001b[39;00m\n\u001b[0;32m----> 6\u001b[0m     ge \u001b[38;5;241m=\u001b[39m \u001b[43mGenotypeEntry\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrow\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrow\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      7\u001b[0m     cc \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m      8\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m cc \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m50\u001b[39m:\n",
      "Cell \u001b[0;32mIn[43], line 24\u001b[0m, in \u001b[0;36mGenotypeEntry.__init__\u001b[0;34m(self, row)\u001b[0m\n\u001b[1;32m     22\u001b[0m     variant \u001b[38;5;241m=\u001b[39m fields[i\u001b[38;5;241m+\u001b[39m\u001b[38;5;241m2\u001b[39m]\n\u001b[1;32m     23\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m---> 24\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCould not get variant from row: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mrow\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     25\u001b[0m regex_del \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mr\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mc.\u001b[39m\u001b[38;5;124m\\\u001b[39m\u001b[38;5;124md+_\u001b[39m\u001b[38;5;124m\\\u001b[39m\u001b[38;5;124md+del\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     26\u001b[0m regex_dup \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mr\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mc.(\u001b[39m\u001b[38;5;124m\\\u001b[39m\u001b[38;5;124md+)dup([A-Z]+)\u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
      "\u001b[0;31mValueError\u001b[0m: Could not get variant from row: {'PatID': 'STX_21770924_Patient_3', 'Chr': '9', 'Start': '130446664', 'End': '130446664', 'Ref': 'A', 'Alt': 'C', 'Func.refGeneWithVer': 'exonic', 'Gene.refGeneWithVer': 'STXBP1', 'GeneDetail.refGeneWithVer': '.', 'ExonicFunc.refGeneWithVer': 'nonsynonymous SNV', 'AAChange.refGeneWithVer': 'STXBP1:NM_003165.3:exon19:c.A1720C:p.T574P', 'Func.ensGene': 'exonic', 'Gene.ensGene': 'ENSG00000136854', 'GeneDetail.ensGene': '.', 'ExonicFunc.ensGene': 'nonsynonymous SNV', 'AAChange.ensGene': 'ENSG00000136854:ENST00000373302:exon19:c.A1720C:p.T574P', 'Gnomad_exome_AF': '.', 'Gnomad_exome_AF_popmax': '.', 'Gnomad_exome_AF_male': '.', 'Gnomad_exome_AF_female': '.', 'Gnomad_exome_AF_raw': '.', 'Gnomad_exome_AF_afr': '.', 'Gnomad_exome_AF_sas': '.', 'Gnomad_exome_AF_amr': '.', 'Gnomad_exome_AF_eas': '.', 'Gnomad_exome_AF_nfe': '.', 'Gnomad_exome_AF_fin': '.', 'Gnomad_exome_AF_asj': '.', 'Gnomad_exome_AF_oth': '.', 'non_topmed_Gnomad_exome_AF_popmax': '.', 'non_neuro_Gnomad_exome_AF_popmax': '.', 'non_cancer_Gnomad_exome_AF_popmax': '.', 'controls_Gnomad_exome_AF_popmax': '.', 'PopFreqMax': '.', '1000G_ALL': '.', '1000G_AFR': '.', '1000G_AMR': '.', '1000G_EAS': '.', '1000G_EUR': '.', '1000G_SAS': '.', 'ExAC_ALL': '.', 'ExAC_AFR': '.', 'ExAC_AMR': '.', 'ExAC_EAS': '.', 'ExAC_FIN': '.', 'ExAC_NFE': '.', 'ExAC_OTH': '.', 'ExAC_SAS': '.', 'ESP6500siv2_ALL': '.', 'ESP6500siv2_AA': '.', 'ESP6500siv2_EA': '.', 'CG46': '.', 'CLNALLELEID': '.', 'CLNDN': '.', 'CLNDISDB': '.', 'CLNREVSTAT': '.', 'CLNSIG': '.', 'DamagePredCount': '16.19', 'SIFT_pred': 'D', 'SIFT4G_pred': 'D', 'Polyphen2_HDIV_pred': 'D', 'Polyphen2_HVAR_pred': 'D', 'LRT_pred': 'D', 'MutationTaster_pred': 'D', 'MutationAssessor_pred': '.', 'FATHMM_pred': 'T', 'PROVEAN_pred': 'D', 'VEST4_score': '0.879', 'MetaSVM_pred': 'D', 'MetaLR_pred': 'D', 'M-CAP_pred': 'D', 'REVEL_score': '0.918', 'MutPred_score': '0.652', 'MVP_score': '0.877', 'MPC_score': '2.626', 'PrimateAI_pred': 'T', 'DEOGEN2_pred': 'T', 'BayesDel_addAF_pred': 'D', 'BayesDel_noAF_pred': 'D', 'ClinPred_pred': 'D', 'LIST-S2_pred': 'D', 'CADD_raw': '3.958', 'CADD_phred': '27.3', 'DANN_score': '0.996', 'fathmm-MKL_coding_pred': 'D', 'fathmm-XF_coding_pred': 'D', 'Eigen-raw_coding': '0.847', 'Eigen-phred_coding': '9.779', 'Eigen-PC-raw_coding': '0.799', 'Eigen-PC-phred_coding': '10.102', 'GenoCanyon_score': '1', 'integrated_fitCons_score': '0.497', 'GM12878_fitCons_score': '0.59', 'H1-hESC_fitCons_score': '0.545', 'HUVEC_fitCons_score': '0.492', 'LINSIGHT': '.', 'GERP++_NR': '5.27', 'GERP++_RS': '5.27', 'phyloP100way_vertebrate': '8.914', 'phyloP30way_mammalian': '1.312', 'phyloP17way_primate': '0.756', 'phastCons100way_vertebrate': '1', 'phastCons30way_mammalian': '0.994', 'phastCons17way_primate': '0.994', 'bStatistic': '301', 'Interpro_domain': '.', 'GTEx_V8_gene': '.', 'GTEx_V8_tissue': '.', 'dbscSNV_ADA_SCORE': '.', 'dbscSNV_RF_SCORE': '.', 'score_zscore_gene': '11;0.89;STXBP3:STXBP1:STXBP2', 'hgmd_class': 'NA', 'hgmd_prot': 'NA', 'hgmd_phen': 'NA', 'hgmd_rankscore': 'NA', 'Otherinfo1': '.', 'Otherinfo2': '.', 'Otherinfo3': '.', 'Otherinfo4': '9', 'Otherinfo5': '130446664', 'Otherinfo7': 'A', 'Otherinfo8': 'C', 'Otherinfo9': '.', 'Otherinfo10': 'PASS', 'Otherinfo11': '.', 'bed': 'Name=98.760813298', '': ''}"
     ]
    }
   ],
   "source": [
    "cc= 0\n",
    "with open(\"input/brain-2021-00642-File011-genotype.tsv\") as f:\n",
    "    reader = DictReader(f, delimiter=\"\\t\")\n",
    "    for row in reader:\n",
    "        #print(row)\n",
    "        ge = GenotypeEntry(row=row)\n",
    "        cc += 1\n",
    "        if cc > 50:\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d2a0632",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "c2c2venv",
   "language": "python",
   "name": "c2c2venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
