{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "34284130",
   "metadata": {},
   "source": [
    "<h1>landscape of STXBP1-related disorders </h1>\n",
    "<p>Extract the clinical data from <a href=\"https://pubmed.ncbi.nlm.nih.gov/35190816/\"target=\"__blank\">Xian et al. (2022) Assessing the landscape of STXBP1-related disorders in 534 individuals. Brain.</a>.<p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0ecc4b45",
   "metadata": {},
   "outputs": [],
   "source": [
    "import phenopackets as php\n",
    "from google.protobuf.json_format import MessageToDict, MessageToJson\n",
    "from google.protobuf.json_format import Parse, ParseDict\n",
    "import pandas as pd\n",
    "from csv import DictReader\n",
    "pd.set_option('display.max_colwidth', None) # show entire column contents, important!\n",
    "from collections import defaultdict\n",
    "import re\n",
    "from pyphetools.creation import *\n",
    "# last tested with pyphetools version 0.2.23"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9e8891ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = HpoParser()\n",
    "hpo_cr = parser.get_hpo_concept_recognizer()\n",
    "hpo_version = parser.get_version()\n",
    "metadata = MetaData(created_by=\"ORCID:0000-0002-0736-9199\")\n",
    "metadata.default_versions_with_hpo(version=hpo_version)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "9dc8c7f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinical_df = pd.read_table(\"input/brain-2021-00642-File011.tsv\");\n",
    "genotype_df = pd.read_table(\"input/brain-2021-00642-File011-genotype.tsv\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "717ba1d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PatID</th>\n",
       "      <th>Source_Journal</th>\n",
       "      <th>Source_PMID*</th>\n",
       "      <th>Year</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Phenotypic_group**</th>\n",
       "      <th>age_onset_m</th>\n",
       "      <th>age_offset_m</th>\n",
       "      <th>age_eval_y</th>\n",
       "      <th>Base_HPO***</th>\n",
       "      <th>HPO_term</th>\n",
       "      <th>Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STX_18469812_Subject_11</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>18469812</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>M</td>\n",
       "      <td>EOEE</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>HP:0003593</td>\n",
       "      <td>Infantile onset</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STX_18469812_Subject_11</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>18469812</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>M</td>\n",
       "      <td>EOEE</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>HP:0010818</td>\n",
       "      <td>Generalized tonic seizures</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STX_18469812_Subject_11</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>18469812</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>M</td>\n",
       "      <td>EOEE</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>HP:0002069</td>\n",
       "      <td>Generalized tonic-clonic seizures</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STX_18469812_Subject_11</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>18469812</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>M</td>\n",
       "      <td>EOEE</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>HP:0010851</td>\n",
       "      <td>EEG with burst suppression</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STX_18469812_Subject_11</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>18469812</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>M</td>\n",
       "      <td>EOEE</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>HP:0002521</td>\n",
       "      <td>Hypsarrhythmia</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     PatID Source_Journal Source_PMID*    Year Sex   \n",
       "0  STX_18469812_Subject_11      Nat Genet     18469812  2008.0   M  \\\n",
       "1  STX_18469812_Subject_11      Nat Genet     18469812  2008.0   M   \n",
       "2  STX_18469812_Subject_11      Nat Genet     18469812  2008.0   M   \n",
       "3  STX_18469812_Subject_11      Nat Genet     18469812  2008.0   M   \n",
       "4  STX_18469812_Subject_11      Nat Genet     18469812  2008.0   M   \n",
       "\n",
       "  Phenotypic_group**  age_onset_m  age_offset_m  age_eval_y Base_HPO***   \n",
       "0               EOEE          2.0           NaN         8.0  HP:0003593  \\\n",
       "1               EOEE          2.0           NaN         8.0  HP:0010818   \n",
       "2               EOEE          2.0           NaN         8.0  HP:0002069   \n",
       "3               EOEE          2.0           NaN         8.0  HP:0010851   \n",
       "4               EOEE          2.0           NaN         8.0  HP:0002521   \n",
       "\n",
       "                            HPO_term Notes  \n",
       "0                    Infantile onset   NaN  \n",
       "1         Generalized tonic seizures   NaN  \n",
       "2  Generalized tonic-clonic seizures   NaN  \n",
       "3         EEG with burst suppression   NaN  \n",
       "4                     Hypsarrhythmia   NaN  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "a938b319",
   "metadata": {},
   "outputs": [],
   "source": [
    "patient_d = defaultdict(list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aadcdbea",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "c657801d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We extracted data on 534 individuals\n"
     ]
    }
   ],
   "source": [
    "with open(\"input/brain-2021-00642-File011.tsv\") as f:\n",
    "    reader = DictReader(f, delimiter=\"\\t\")\n",
    "    for row in reader:\n",
    "        prow = PatientRow(row=row)\n",
    "        patient_d[prow.patID].append(prow)\n",
    "print(f\"We extracted data on {len(patient_d)} individuals\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "55a565a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PatID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Func.refGeneWithVer</th>\n",
       "      <th>Gene.refGeneWithVer</th>\n",
       "      <th>GeneDetail.refGeneWithVer</th>\n",
       "      <th>ExonicFunc.refGeneWithVer</th>\n",
       "      <th>...</th>\n",
       "      <th>Otherinfo3</th>\n",
       "      <th>Otherinfo4</th>\n",
       "      <th>Otherinfo5</th>\n",
       "      <th>Otherinfo7</th>\n",
       "      <th>Otherinfo8</th>\n",
       "      <th>Otherinfo9</th>\n",
       "      <th>Otherinfo10</th>\n",
       "      <th>Otherinfo11</th>\n",
       "      <th>bed</th>\n",
       "      <th>Unnamed: 134</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STX_18469812_Subject_11</td>\n",
       "      <td>9.0</td>\n",
       "      <td>130422313.0</td>\n",
       "      <td>130422313.0</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>STXBP1</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>9</td>\n",
       "      <td>130422313</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>Name=70.695764689</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STX_18469812_Subject_3</td>\n",
       "      <td>9.0</td>\n",
       "      <td>130444768.0</td>\n",
       "      <td>130444768.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>STXBP1</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>9</td>\n",
       "      <td>130444768</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>Name=75.862552050</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STX_18469812_Subject_6</td>\n",
       "      <td>9.0</td>\n",
       "      <td>130425593.0</td>\n",
       "      <td>130425593.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>STXBP1</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>9</td>\n",
       "      <td>130425593</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>Name=89.193399398</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STX_18469812_Subject_7</td>\n",
       "      <td>9.0</td>\n",
       "      <td>130439001.0</td>\n",
       "      <td>130439001.0</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>exonic</td>\n",
       "      <td>STXBP1</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>9</td>\n",
       "      <td>130439001</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>Name=20.833326089</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STX_19557857_Patient_1</td>\n",
       "      <td>9.0</td>\n",
       "      <td>130416076.0</td>\n",
       "      <td>130416076.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>splicing</td>\n",
       "      <td>STXBP1</td>\n",
       "      <td>NM_001032221.3:exon3:c.169+1G&gt;A:NM_003165.3:exon3:c.169+1G&gt;A</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>9</td>\n",
       "      <td>130416076</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 135 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     PatID  Chr        Start          End Ref Alt   \n",
       "0  STX_18469812_Subject_11  9.0  130422313.0  130422313.0   T   A  \\\n",
       "1   STX_18469812_Subject_3  9.0  130444768.0  130444768.0   G   A   \n",
       "2   STX_18469812_Subject_6  9.0  130425593.0  130425593.0   G   A   \n",
       "3   STX_18469812_Subject_7  9.0  130439001.0  130439001.0   T   G   \n",
       "4   STX_19557857_Patient_1  9.0  130416076.0  130416076.0   G   A   \n",
       "\n",
       "  Func.refGeneWithVer Gene.refGeneWithVer   \n",
       "0              exonic              STXBP1  \\\n",
       "1              exonic              STXBP1   \n",
       "2              exonic              STXBP1   \n",
       "3              exonic              STXBP1   \n",
       "4            splicing              STXBP1   \n",
       "\n",
       "                                      GeneDetail.refGeneWithVer   \n",
       "0                                                             .  \\\n",
       "1                                                             .   \n",
       "2                                                             .   \n",
       "3                                                             .   \n",
       "4  NM_001032221.3:exon3:c.169+1G>A:NM_003165.3:exon3:c.169+1G>A   \n",
       "\n",
       "  ExonicFunc.refGeneWithVer  ... Otherinfo3 Otherinfo4 Otherinfo5 Otherinfo7   \n",
       "0         nonsynonymous SNV  ...          .          9  130422313          T  \\\n",
       "1         nonsynonymous SNV  ...          .          9  130444768          G   \n",
       "2         nonsynonymous SNV  ...          .          9  130425593          G   \n",
       "3         nonsynonymous SNV  ...          .          9  130439001          T   \n",
       "4                         .  ...          .          9  130416076          G   \n",
       "\n",
       "  Otherinfo8 Otherinfo9 Otherinfo10 Otherinfo11                bed   \n",
       "0          A          .        PASS           .  Name=70.695764689  \\\n",
       "1          A          .        PASS           .  Name=75.862552050   \n",
       "2          A          .        PASS           .  Name=89.193399398   \n",
       "3          G          .        PASS           .  Name=20.833326089   \n",
       "4          A          .        PASS           .                  .   \n",
       "\n",
       "  Unnamed: 134  \n",
       "0          NaN  \n",
       "1          NaN  \n",
       "2          NaN  \n",
       "3          NaN  \n",
       "4          NaN  \n",
       "\n",
       "[5 rows x 135 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "b667810e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_var_inf(aachange):\n",
    "    \"\"\"\n",
    "    aachange: e.g., STXBP1:NM_001032221.3:exon6:c.T353G:p.L118R,STXBP1:NM_003165.3:exon6:c.T353G:p.L118R\n",
    "    \"\"\"\n",
    "    fields = aachange.split(\":\")\n",
    "    index = 0\n",
    "    i = 0\n",
    "    transcript = \"?\"\n",
    "    for f in fields:\n",
    "        if f == \"NM_001032221.3\" or f == \"NM_003165.3\":\n",
    "            index = i\n",
    "            transcript = f\n",
    "            break\n",
    "        i += 1\n",
    "    if (i + 2) < len(fields):\n",
    "        variant = fields[i+2]\n",
    "    else:\n",
    "        raise ValueError(f\"Could not get variant because of fields: {fields} and i={i}\")   \n",
    "    return transcript, variant\n",
    "\n",
    "def extract_splice_var(genedetail):\n",
    "    \"\"\"\n",
    "    genedetail: e.g., NM_001032221.3:exon3:c.169+1G>A;NM_003165.3:exon3:c.169+1G>A\n",
    "    \"\"\"\n",
    "    fields = genedetail.split(\":\")\n",
    "    index = 0\n",
    "    i = 0\n",
    "    transcript = \"?\"\n",
    "    for f in fields:\n",
    "        if f == \"NM_001032221.3\" or f == \"NM_003165.3\":\n",
    "            index = i\n",
    "            transcript = f\n",
    "            break\n",
    "        i += 1\n",
    "    if (i + 2) < len(fields):\n",
    "        variant = fields[i+2]\n",
    "    else:\n",
    "        raise ValueError(f\"Could not get variant because of fields: {fields} and i={i} from genedetail \\\"{genedetail}\\\"\")   \n",
    "    return transcript, variant\n",
    "\n",
    "\n",
    "class GenotypeEntry:\n",
    "    def __init__(self, row):\n",
    "        self.patID = row[\"PatID\"]\n",
    "        self.chrom = row[\"Chr\"]\n",
    "        self.start = row[\"Start\"]\n",
    "        self.end = row[\"End\"]\n",
    "        self.ref = row[\"Ref\"]\n",
    "        self.alt = row[\"Alt\"]\n",
    "        self.transcript = \"?\"\n",
    "        func = row[\"Func.refGeneWithVer\"]\n",
    "        self.category = func\n",
    "        genenot = row[\"Gene.refGeneWithVer\"]\n",
    "        aachange = row[\"AAChange.refGeneWithVer\"]\n",
    "        if func == \"exonic\":\n",
    "            transcript, variant = extract_var_inf(aachange)\n",
    "            self.transcript = transcript\n",
    "            regex_del = r\"c.\\d+_\\d+del\"\n",
    "            regex_single_nt_del = r\"(c.\\d+del)[ACGT]\"\n",
    "            regex_dup = r\"c.(\\d+)dup([A-Z]+)\"\n",
    "            regex_sub = r\"c.([A-Z]+)(\\d+)([A-Z]+)\"\n",
    "            regex_ins = r\"c.(\\d+)_(\\d+)ins([A-Z]+)\"  # e.g., 1372_1373insGCCGGAGCAA\n",
    "            regex_delins = r\"(c.\\d+_\\d+delins[A-Z]+)\"\n",
    "            result = re.search(regex_sub, variant)\n",
    "            result_dup = re.search(regex_dup, variant)\n",
    "            result_single_nt_del = re.search(regex_single_nt_del, variant)\n",
    "            result_ins = re.search(regex_ins, variant)\n",
    "            result_delins = re.search(regex_delins, variant)\n",
    "            if re.match(regex_del, variant):\n",
    "                self.hgvs = variant\n",
    "            elif result:\n",
    "                ref = result.group(1)\n",
    "                position = result.group(2)\n",
    "                alt = result.group(3)\n",
    "                hgvs = f\"c.{position}{ref}>{alt}\"\n",
    "                self.hgvs = hgvs\n",
    "            elif result_dup:\n",
    "                position=result_dup.group(1)\n",
    "                hgvs = f\"c.{position}dup\"\n",
    "                self.hgvs = hgvs\n",
    "            elif result_single_nt_del:\n",
    "                self.hgvs = result_single_nt_del.group(1)\n",
    "            elif result_delins:\n",
    "                self.hgvs = result_delins.group(1)\n",
    "            elif result_ins:\n",
    "                pos1 = result_ins.group(1)\n",
    "                pos2 = result_ins.group(2)\n",
    "                seq = result_ins.group(3)\n",
    "                self.hgvs = f\"c.{pos1}_{pos2}ins{seq}\"\n",
    "            else:\n",
    "                raise ValueError(f\"Could not parse variant {variant}\")\n",
    "        elif func == 'splicing':\n",
    "            geneDetail = row[\"GeneDetail.refGeneWithVer\"]\n",
    "            transcript, variant = extract_splice_var(geneDetail)\n",
    "            self.transcript = transcript\n",
    "            self.hgvs = variant\n",
    "        elif func == \"NA\":\n",
    "            pass\n",
    "        elif func == \"intronic\":\n",
    "            pass  \n",
    "        else:\n",
    "            print(f\"{self.patID}---function {func}\")\n",
    "            raise ValueError(f\"Could not parse variant  for func {func}\\n{row}\")\n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "198cbd90",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We got 463 usable genotypes\n",
      "We got 46 NAs, and 25 intronic genotypes - both were skipped\n"
     ]
    }
   ],
   "source": [
    "na_genotype = 0\n",
    "intronic_genotype = 0\n",
    "genotype_d = defaultdict(GenotypeEntry)\n",
    "with open(\"input/brain-2021-00642-File011-genotype.tsv\") as f:\n",
    "    reader = DictReader(f, delimiter=\"\\t\")\n",
    "    for row in reader:\n",
    "        ge = GenotypeEntry(row=row)\n",
    "        if ge.category == \"intronic\":\n",
    "            intronic_genotype = intronic_genotype + 1\n",
    "        elif ge.category == \"NA\":\n",
    "            na_genotype = na_genotype +1\n",
    "        else:\n",
    "            patient_id = ge.patID\n",
    "            genotype_d[patient_id] = ge\n",
    "print(f\"We got {len(genotype_d)} usable genotypes\")\n",
    "print(f\"We got {na_genotype} NAs, and {intronic_genotype} intronic genotypes - both were skipped\")\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "665ed413",
   "metadata": {},
   "source": [
    "<h2>HpoTerm</h2>\n",
    "<p>We create classes called HpoTerm and PatientRow to help process the data from the input file. Note that\n",
    "the data for one patient is distributed across multiple rows of the input Excel file.</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "09762f2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "class PatientRow:\n",
    "    def __init__(self, row):\n",
    "        self.patID = row[\"PatID\"]\n",
    "        self.sex = row[\"Sex\"]\n",
    "        self.phenogroup = row[\"Phenotypic_group**\"]\n",
    "        \n",
    "def row_to_hpo(row):\n",
    "    \"\"\"Transform a row of the dataframe to an HPO term\n",
    "    \"\"\"\n",
    "    try: \n",
    "        age_onset_m = int(row[\"age_onset_m\"])\n",
    "        onset = f\"P{age_onset_m}M\"\n",
    "    except:\n",
    "        onset = None\n",
    "    try:\n",
    "        age_offset_m = int(row[\"age_offset_m\"])\n",
    "        offset = f\"P{age_offset_m}M\"\n",
    "    except:\n",
    "        offset = None\n",
    "    hpo_id = row[\"Base_HPO***\"]\n",
    "    hpo_label = row[\"HPO_term\"]\n",
    "    excluded = False\n",
    "    # excluded terms are coded with NP:0001234 instead of HP:0001234\n",
    "    if hpo_id.startswith(\"NP\"):\n",
    "        excluded = True\n",
    "    hpo_id = \"H\" + hpo_id[1:]\n",
    "    return HpTerm(hpo_id=hpo_id, label=hpo_label, onset=onset, resolution=offset) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "7134daee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We got 463 patients and 463 demographics\n"
     ]
    }
   ],
   "source": [
    "patient_d = defaultdict(list)\n",
    "patient_demographic_d = defaultdict(PatientRow)\n",
    "for _, row in clinical_df.iterrows():\n",
    "    patID = row[\"PatID\"]\n",
    "    if patID not in genotype_d:\n",
    "        continue\n",
    "    if patID not in patient_demographic_d:\n",
    "        patient_demographic_d[patID] = PatientRow(row=row)\n",
    "    hpo = row_to_hpo(row=row)\n",
    "    patient_d[patID].append(hpo)\n",
    "print(f\"We got {len(patient_d)} patients and {len(patient_demographic_d)} demographics\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63b25dc7",
   "metadata": {},
   "source": [
    "<H2>Putting it all together</h2>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "60ffc759",
   "metadata": {},
   "outputs": [],
   "source": [
    "disease_label = \"Developmental and epileptic encephalopathy 4\"\n",
    "disease_id = \"OMIM:612164\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "2959fd90",
   "metadata": {},
   "outputs": [],
   "source": [
    "#  def __init__(self, individual_id, hpo_terms, sex, age=Constants.NOT_PROVIDED, variant_list=None, disease_id=None, disease_label=None):\n",
    "individual_list = []\n",
    "for pat_id, patRow in patient_demographic_d.items():\n",
    "    hpo_list = patient_d.get(pat_id)\n",
    "    if hpo_list is None:\n",
    "        print(f\"Could not find hpo list for {pat_id}\")\n",
    "        continue\n",
    "    if len(hpo_list) == 0:\n",
    "        print(f\"warning, empty HPO list for {pat_id}\")\n",
    "    sex = patRow.sex\n",
    "    gtype = genotype_d.get(pat_id)\n",
    "    if gtype is None:\n",
    "        print(f\"Could not find genotype for {pat_id} (should never happen)\")\n",
    "        continue\n",
    "    #(self, assembly, vcf_d, symbol=None, hgnc=None, hgvs=None, transcript=None, g_hgvs=None) -> None:\n",
    "    \n",
    "    genome = 'hg38'\n",
    "    transcript='NM_015133.4'\n",
    "    varMapper = VariantColumnMapper(assembly=genome,\n",
    "                                column_name='Transcript\\nNM_015133.4\\nc.', \n",
    "                                transcript=transcript, \n",
    "                                default_genotype='heterozygous')\n",
    "\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "dbee90b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "validator = VariantValidator(genome_build='hg38')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "4c4a9f42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NM_001032221.6:c.251T>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.251T>A/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127660034T>A\n",
      "NM_001032221.6:c.1631G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1631G>A/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127682489G>A\n",
      "NM_001032221.6:c.539G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.539G>A/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127663314G>A\n",
      "NM_001032221.6:c.1328T>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1328T>G/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127676722T>G\n",
      "NM_001032221.6:c.169+1G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.169+1G>A/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127653797G>A\n",
      "NM_001032221.6:c.1162C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1162C>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127675855C>T\n",
      "NM_001032221.6:c.1434G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1434G>A/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127678505G>A\n",
      "NM_001032221.6:c.893_894del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.893_894del/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127668175CAG>C\n",
      "NM_001032221.6:c.1029+1G>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1029+1G>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127672117G>T\n",
      "NM_001032221.6:c.429+1G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.429+1G>A/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127661206G>A\n",
      "NM_001032221.6:c.703C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.703C>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127666205C>T\n",
      "NM_001032221.6:c.388_389del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.388_389del/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127661161ACT>A\n",
      "NM_001032221.6:c.1217G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1217G>A/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127675910G>A\n",
      "NM_001032221.6:c.157G>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.157G>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127653784G>T\n",
      "NM_001032221.6:c.961A>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.961A>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127669956A>T\n",
      "NM_001032221.6:c.747dup\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.747dup/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127666245C>CT\n",
      "NM_001032221.6:c.232_235del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.232_235del/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127658434ATCAC>A\n",
      "NM_001032221.6:c.1654T>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1654T>C/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127682512T>C\n",
      "NM_001032221.6:c.902+1G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.902+1G>A/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127668188G>A\n",
      "NM_001032221.6:c.1439C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1439C>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127678510C>T\n",
      "NM_003165.6:c.1720A>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003165.6%3Ac.1720A>C/NM_003165.6?content-type=application%2Fjson\n",
      "chr9:127684385A>C\n",
      "NM_001032221.6:c.548T>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.548T>G/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127663323T>G\n",
      "NM_001032221.6:c.931dup\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.931dup/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127669921A>AT\n",
      "NM_001032221.6:c.301G>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.301G>C/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127660084G>C\n",
      "NM_001032221.6:c.247-1del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.247-1del/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127660028AG>A\n",
      "NM_001032221.6:c.175G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.175G>A/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127658380G>A\n",
      "NM_001032221.6:c.607C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.607C>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127665275C>T\n",
      "NM_001032221.6:c.1631G>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1631G>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127682489G>T\n",
      "NM_001032221.6:c.230T>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.230T>A/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127658435T>A\n",
      "NM_001032221.6:c.1651C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1651C>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127682509C>T\n",
      "NM_001032221.6:c.1630G>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1630G>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127682488G>T\n",
      "NM_001032221.6:c.1157dup\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1157dup/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127675847A>AC\n",
      "NM_001032221.6:c.1060T>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1060T>C/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127673211T>C\n",
      "NM_001032221.6:c.568C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.568C>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127663343C>T\n",
      "NM_001032221.6:c.125C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.125C>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127653752C>T\n",
      "NM_003165.6:c.1708A>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003165.6%3Ac.1708A>A/NM_003165.6?content-type=application%2Fjson\n",
      "chr9:127684373A>A\n",
      "NM_001032221.6:c.238T>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.238T>C/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127658443T>C\n",
      "NM_001032221.6:c.1004C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1004C>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127672091C>T\n",
      "NM_001032221.6:c.585C>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.585C>G/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127665253C>G\n",
      "NM_001032221.6:c.1130dup\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1130dup/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127675822G>GA\n",
      "NM_001032221.6:c.875G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.875G>A/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127668160G>A\n",
      "NM_001032221.6:c.795-1G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.795-1G>A/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127668079G>A\n",
      "NM_001032221.6:c.1462-2A>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1462-2A>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127680155A>T\n",
      "NM_001032221.6:c.444del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.444del/NM_001032221.6?content-type=application%2Fjson\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr9:127663218AC>A\n",
      "NM_001032221.6:c.416C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.416C>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127661192C>T\n",
      "NM_001032221.6:c.901C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.901C>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127668186C>T\n",
      "NM_001032221.6:c.148dup\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.148dup/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127653774C>CA\n",
      "NM_001032221.6:c.704G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.704G>A/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127666206G>A\n",
      "NM_001032221.6:c.437_438del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.437_438del/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127663211TCC>T\n",
      "NM_001032221.6:c.778G>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.778G>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127666280G>T\n",
      "NM_001032221.6:c.1099C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1099C>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127673250C>T\n",
      "NM_001032221.6:c.734A>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.734A>G/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127666236A>G\n",
      "NM_001032221.6:c.754_755del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.754_755del/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127666254CTA>C\n",
      "NM_001032221.6:c.1614_1616del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1614_1616del/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127682467CTCA>C\n",
      "NM_001032221.6:c.364C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.364C>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127661140C>T\n",
      "NM_001032221.6:c.79del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.79del/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127651639AG>A\n",
      "NM_001032221.6:c.795-2A>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.795-2A>G/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127668078A>G\n",
      "NM_001032221.6:c.1651C>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1651C>G/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127682509C>G\n",
      "NM_001032221.6:c.15del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.15del/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127612417GC>G\n",
      "NM_001032221.6:c.734A>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.734A>C/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127666236A>C\n",
      "NM_001032221.6:c.1095_1096del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1095_1096del/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127673243ACT>A\n",
      "NM_001032221.6:c.1630G>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1630G>C/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127682488G>C\n",
      "NM_001032221.6:c.83G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.83G>A/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127651648G>A\n",
      "NM_001032221.6:c.661G>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.661G>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127665329G>T\n",
      "NM_001032221.6:c.771dup\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.771dup/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127666272T>TG\n",
      "NM_001032221.6:c.308A>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.308A>C/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127660091A>C\n",
      "NM_001032221.6:c.1249+2T>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1249+2T>C/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127675944T>C\n",
      "NM_001032221.6:c.922A>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.922A>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127669917A>T\n",
      "NM_001032221.6:c.170-12_173del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.170-12_173del/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127658357GGTTGTTTTGTTGTCTA>G\n",
      "NM_001032221.6:c.663+1G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.663+1G>A/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127665332G>A\n",
      "NM_001032221.6:c.1303G>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1303G>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127676697G>T\n",
      "NM_001032221.6:c.1347del\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1347del/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127676738GC>G\n",
      "NM_001032221.6:c.1607G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1607G>A/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127682465G>A\n",
      "NM_001032221.6:c.296A>G\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.296A>G/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127660079A>G\n",
      "NM_001032221.6:c.1216C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1216C>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127675909C>T\n",
      "NM_001032221.6:c.236C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.236C>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127658441C>T\n",
      "NM_001032221.6:c.538T>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.538T>C/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127663313T>C\n",
      "NM_001032221.6:c.1022T>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1022T>C/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127672109T>C\n",
      "NM_001032221.6:c.1438C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1438C>T/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127678509C>T\n",
      "NM_001032221.6:c.1359+1G>A\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1359+1G>A/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127676754G>A\n",
      "NM_001032221.6:c.430-1G>C\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.430-1G>C/NM_001032221.6?content-type=application%2Fjson\n",
      "chr9:127663204G>C\n",
      "NM_003165.6:c.1723C>T\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_003165.6%3Ac.1723C>T/NM_003165.6?content-type=application%2Fjson\n",
      "chr9:127684388C>T\n",
      "NM_001032221.6:c.1548_1559delinsAT,STXBP1\n",
      "https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg38/NM_001032221.6%3Ac.1548_1559delinsAT,STXBP1/NM_001032221.6?content-type=application%2Fjson\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Expecting to get a gene_variant from Variant Validator but got warning",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[80], line 16\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m     15\u001b[0m     \u001b[38;5;28mprint\u001b[39m(total_hgvs)\n\u001b[0;32m---> 16\u001b[0m     v \u001b[38;5;241m=\u001b[39m \u001b[43mvalidator\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencode_hgvs\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhgvs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgtype\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhgvs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcustom_transcript\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtranscript\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     17\u001b[0m     \u001b[38;5;28mprint\u001b[39m(v)\n\u001b[1;32m     18\u001b[0m     validated_var_d[total_hgvs] \u001b[38;5;241m=\u001b[39m v\n",
      "File \u001b[0;32m~/GIT/c2s2-clin/c2c2venv/lib/python3.10/site-packages/pyphetools/creation/variant_validator.py:36\u001b[0m, in \u001b[0;36mVariantValidator.encode_hgvs\u001b[0;34m(self, hgvs, custom_transcript)\u001b[0m\n\u001b[1;32m     34\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m vv_dict[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mflag\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgene_variant\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[1;32m     35\u001b[0m         flag \u001b[38;5;241m=\u001b[39m vv_dict[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mflag\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[0;32m---> 36\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mExpecting to get a gene_variant from Variant Validator but got \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mflag\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     37\u001b[0m \u001b[38;5;66;03m# This is the easiest way to get the variant key, which may contain difficult characters\u001b[39;00m\n\u001b[1;32m     38\u001b[0m variant_key \u001b[38;5;241m=\u001b[39m [k \u001b[38;5;28;01mfor\u001b[39;00m k \u001b[38;5;129;01min\u001b[39;00m vv_dict\u001b[38;5;241m.\u001b[39mkeys() \u001b[38;5;28;01mif\u001b[39;00m k \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mflag\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmetadata\u001b[39m\u001b[38;5;124m'\u001b[39m}][\u001b[38;5;241m0\u001b[39m]\n",
      "\u001b[0;31mValueError\u001b[0m: Expecting to get a gene_variant from Variant Validator but got warning"
     ]
    }
   ],
   "source": [
    "validated_var_d = defaultdict()\n",
    "c = 0\n",
    "for patid, gtype in genotype_d.items():\n",
    "    #print(f\"{patid} - {gtype.hgvs}\")\n",
    "    if gtype.transcript == 'NM_001032221.3':\n",
    "        transcript = 'NM_001032221.6'\n",
    "    elif gtype.transcript == 'NM_003165.3':\n",
    "        transcript = 'NM_003165.6'\n",
    "    else:\n",
    "        raise ValueError(f\"Unexpected transcript: {gtype.transcript}\")\n",
    "    total_hgvs = f\"{transcript}:{gtype.hgvs}\"\n",
    "    if total_hgvs in validated_var_d:\n",
    "        pass\n",
    "    else:\n",
    "        print(total_hgvs)\n",
    "        v = validator.encode_hgvs(hgvs=gtype.hgvs, custom_transcript=transcript)\n",
    "        print(v)\n",
    "        validated_var_d[total_hgvs] = v"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71995c79",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "c2c2venv",
   "language": "python",
   "name": "c2c2venv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
