{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CYP21A2\n",
    "[this form of congenital adrenal hyperplasia](https://omim.org/entry/201910) is caused by homozygous or compound heterozygous mutation in the CYP21A2 gene.\n",
    "This notebook shows how to create phenopackets using GA4GH classes from pyphetools directly to flexibly model elements not yet included in the Template.\n",
    "\n",
    "Here, we use this to import data specifically from the publication [Xu C, et al. (2019) Genotype-phenotype correlation study and mutational and hormonal analysis in a Chinese cohort with 21-hydroxylase deficiency. Mol Genet Genomic Med.](https://pubmed.ncbi.nlm.nih.gov/30968594/). We processed the data in the supplemental Word file and created an Excel file that\n",
    "contains the data.\n",
    "\n",
    "We did not include individuals 70-72, because only one variant was identified.\n",
    "\n",
    "In some cases, the alleles were compound. Since we cannot currently represent this in a simple way, we chose the most clearly pathogenic variant in such cases. To\n",
    "see our choices, see the original_allele column of the Excel file.\n",
    "\n",
    "We create Measurement objects for many of the lab values. We did not include progesterone, FSH, and LH because of the fact that the normal values for these analytes can depend on the\n",
    "stage of the menstrual cycle, which was not available in the input data.\n",
    "\n",
    "Note that this notebook includes Python code that is specific to the format of the input Excel file, but the approach can be adapted to other cohorts mutatis mutandis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pyphetools version 0.9.105\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "pd.set_option('display.max_colwidth', None) # show entire column contents, important!\n",
    "import typing\n",
    "import os\n",
    "import re\n",
    "from google.protobuf.json_format import MessageToJson\n",
    "from pyphetools.creation import Measurements, VariantManager, PromoterVariant, StructuralVariant, OntologyTerms\n",
    "from pyphetools.pp.v202 import *\n",
    "import pyphetools\n",
    "print(f\"pyphetools version {pyphetools.__version__}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_at_dx</th>\n",
       "      <th>17ohp</th>\n",
       "      <th>androgen</th>\n",
       "      <th>testosterone</th>\n",
       "      <th>other</th>\n",
       "      <th>allele_1</th>\n",
       "      <th>allele_2</th>\n",
       "      <th>allele_original</th>\n",
       "      <th>...</th>\n",
       "      <th>Clitoral hypertrophy</th>\n",
       "      <th>Irregular menstruation</th>\n",
       "      <th>Diarrhea</th>\n",
       "      <th>Arrhythmia</th>\n",
       "      <th>Congenital hypothyroidism</th>\n",
       "      <th>Growth delay</th>\n",
       "      <th>Acne</th>\n",
       "      <th>Moon facies</th>\n",
       "      <th>Hyperpigmentation of the skin</th>\n",
       "      <th>Original:symptoms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>str</td>\n",
       "      <td>sex</td>\n",
       "      <td>age_at_dx</td>\n",
       "      <td>17ohp</td>\n",
       "      <td>androgen</td>\n",
       "      <td>ng/dl</td>\n",
       "      <td>other</td>\n",
       "      <td>gt1</td>\n",
       "      <td>gt2</td>\n",
       "      <td>gt1</td>\n",
       "      <td>...</td>\n",
       "      <td>HP:0008665</td>\n",
       "      <td>HP:0000858</td>\n",
       "      <td>HP:0002014</td>\n",
       "      <td>HP:0011675</td>\n",
       "      <td>HP:0000851</td>\n",
       "      <td>HP:0001510</td>\n",
       "      <td>HP:0001061</td>\n",
       "      <td>HP:0500011</td>\n",
       "      <td>HP:0000953</td>\n",
       "      <td>Original</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>individual 1</td>\n",
       "      <td>M</td>\n",
       "      <td>P20D</td>\n",
       "      <td>800</td>\n",
       "      <td>T: 1.27</td>\n",
       "      <td>127</td>\n",
       "      <td>na</td>\n",
       "      <td>c.955C&gt;T</td>\n",
       "      <td>DEL exon1-7</td>\n",
       "      <td>c.955C&gt;T</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>observed</td>\n",
       "      <td>Hyperpigmentation, milk regurgitation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>individual 2</td>\n",
       "      <td>M</td>\n",
       "      <td>Congenital onset</td>\n",
       "      <td>526</td>\n",
       "      <td>T: 1.68</td>\n",
       "      <td>168</td>\n",
       "      <td>na</td>\n",
       "      <td>c.332_339del8</td>\n",
       "      <td>DEL exon7-10</td>\n",
       "      <td>c.332_339del8</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>observed</td>\n",
       "      <td>Hyperpigmentation, undescended testis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>individual 3</td>\n",
       "      <td>M</td>\n",
       "      <td>P1M</td>\n",
       "      <td>685</td>\n",
       "      <td>T: 3.4</td>\n",
       "      <td>340</td>\n",
       "      <td>na</td>\n",
       "      <td>c.332_339del</td>\n",
       "      <td>gene deletion</td>\n",
       "      <td>c.290-13A/C&gt;G, c.332-339del8, c.518 T&gt;A</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>milk regurgitation, advanced penile growth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>individual 4</td>\n",
       "      <td>M</td>\n",
       "      <td>Congenital onset</td>\n",
       "      <td>488</td>\n",
       "      <td>T: 0.78</td>\n",
       "      <td>78</td>\n",
       "      <td>na</td>\n",
       "      <td>c.1069C&gt;T</td>\n",
       "      <td>gene deletion</td>\n",
       "      <td>c.1069C&gt;T</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>observed</td>\n",
       "      <td>Hyperpigmentation, advanced penile growth</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     individual  sex         age_at_dx  17ohp  androgen testosterone  other  \\\n",
       "0           str  sex         age_at_dx  17ohp  androgen        ng/dl  other   \n",
       "1  individual 1    M              P20D    800   T: 1.27          127     na   \n",
       "2  individual 2    M  Congenital onset    526   T: 1.68          168     na   \n",
       "3  individual 3    M               P1M    685    T: 3.4          340     na   \n",
       "4  individual 4    M  Congenital onset    488   T: 0.78           78     na   \n",
       "\n",
       "        allele_1       allele_2                          allele_original  ...  \\\n",
       "0            gt1            gt2                                      gt1  ...   \n",
       "1       c.955C>T    DEL exon1-7                                 c.955C>T  ...   \n",
       "2  c.332_339del8   DEL exon7-10                            c.332_339del8  ...   \n",
       "3   c.332_339del  gene deletion  c.290-13A/C>G, c.332-339del8, c.518 T>A  ...   \n",
       "4      c.1069C>T  gene deletion                                c.1069C>T  ...   \n",
       "\n",
       "  Clitoral hypertrophy Irregular menstruation    Diarrhea  Arrhythmia  \\\n",
       "0           HP:0008665             HP:0000858  HP:0002014  HP:0011675   \n",
       "1                   na                     na          na          na   \n",
       "2                   na                     na          na          na   \n",
       "3                   na                     na          na          na   \n",
       "4                   na                     na          na          na   \n",
       "\n",
       "  Congenital hypothyroidism Growth delay        Acne Moon facies  \\\n",
       "0                HP:0000851   HP:0001510  HP:0001061  HP:0500011   \n",
       "1                        na           na          na          na   \n",
       "2                        na           na          na          na   \n",
       "3                        na           na          na          na   \n",
       "4                        na           na          na          na   \n",
       "\n",
       "  Hyperpigmentation of the skin                           Original:symptoms  \n",
       "0                    HP:0000953                                    Original  \n",
       "1                      observed      Hyperpigmentation, milk regurgitation   \n",
       "2                      observed       Hyperpigmentation, undescended testis  \n",
       "3                            na  milk regurgitation, advanced penile growth  \n",
       "4                      observed   Hyperpigmentation, advanced penile growth  \n",
       "\n",
       "[5 rows x 38 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_excel(\"input/CYP21A2_Xu_2019.xlsx\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "PMID = \"PMID:30968594\"\n",
    "title = \"Genotype-phenotype correlation study and mutational and hormonal analysis in a Chinese cohort with 21-hydroxylase deficiency\"\n",
    "created_by=\"ORCID:0000-0002-5648-2155\"\n",
    "metadata = MetaData.metadata_for_pmid(created_by=created_by, pmid=PMID, citation_title=title, include_loinc=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Measurements and Phenotypic Features\n",
    "The PhenotypicFeature element (\"message\") of the Phenopacket Schema represents a qualitivate observation that is indicate by an ontology term as observed or excluded. In contracts, a\n",
    "Measurement element represents a (usually) numerical result (for instance of a laboratory test). Measurements can include a reference range, and if the observed value is above the upper limit of normal (e.g., blood sodium of 152) this may also correspond to an ontology term (in this case, Hypernatremia, i.e., elevated blood sodium). In version 2 of the Phenopacket Schema, it is not possible to relate Measurement and PhenotypicFeature objects, but this feature may be added to version 3. Here, we show how to create both Measurement and PhenotypicFeature messages based on laboratory values. Note that if a Measurement in in the normal range (e.g., a blood sodium value of 136), then we can exclude the parent term (Abnormal circulation sodium concentration -- excluded)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def measurement_17php(row: pd.Series, \n",
    "                      measurement_list: typing.List[Measurement]) -> None:\n",
    "    \"\"\"\n",
    "    In CYP21A2 deficiency, 17-hydroxyprogesterone (17-OHP) is not converted to 11-deoxycortisol.\n",
    "    Normal values:\n",
    "    Babies more than 24 hours old - less than 400 to 600 nanograms per deciliter (ng/dL) or 12.12 to 18.18 nanomoles per liter (nmol/L)\n",
    "    Children before puberty around 100 ng/dL or 3.03 nmol/L\n",
    "    Adults - less than 200 ng/dL or 6.06 nmol/L\n",
    "    The units used here are nanogram per deciliter. We do not have the age at measurement, so we cannot apply the normal range (and thus we do not call HPO terms)\n",
    "    Add a measurement to the list if we can successfully parse  \n",
    "    \"\"\"\n",
    "    value = row[\"17ohp\"]\n",
    "    if value == \"na\":\n",
    "        return None\n",
    "    try:\n",
    "        concentration = float(value)\n",
    "        m = Measurements.nanogram_per_deciliter(code=\"LOINC:1668-3\",\n",
    "                                      label=\"17-Hydroxyprogesterone[Mass/Vol]\",\n",
    "                                      concentration=concentration)\n",
    "        measurement_list.append(m)\n",
    "    except ValueError:\n",
    "        print(f\"Could not parse \\\"{value}\\\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def testosterone_assessment(row: pd.Series, \n",
    "                            measurement_list: typing.List[Measurement],\n",
    "                            phenotypic_feature_list: typing.List[PhenotypicFeature]) -> None:\n",
    "    \"\"\"\n",
    "    Male: 300 to 1,000 nanograms per deciliter (ng/dL) or 10 to 35 nanomoles per liter (nmol/L)\n",
    "    Female: 15 to 70 ng/dL or 0.5 to 2.4 nmol/L\n",
    "    in this paper, testosterone is reported as (ng/ml), we converted to ng/dL (conversion factor 100)\n",
    "    The LOINC code is 2986-8; Testosterone [Mass/volume] in Serum or Plasma     \n",
    "    Add a measurement to the list if we can successfully parse and check if the value is abnormal.\n",
    "    \"\"\"\n",
    "    value = row[\"testosterone\"]\n",
    "    if value == \"na\":\n",
    "        return None\n",
    "    sex = row[\"sex\"]\n",
    "    if sex == \"M\":\n",
    "        low = 300\n",
    "        high = 1000\n",
    "    elif sex == \"F\":\n",
    "        low = 15\n",
    "        high = 70\n",
    "    else:\n",
    "        print(\"Warning: did not recognize sex\")\n",
    "        return None\n",
    "    try:\n",
    "        concentration = float(value)\n",
    "        m = Measurements.nanogram_per_deciliter(code=\"LOINC:2986-8\",\n",
    "                                      label=\"Testosterone[Mass/Vol]\",\n",
    "                                      concentration=concentration,\n",
    "                                      low=low,\n",
    "                                      high=high)\n",
    "        measurement_list.append(m)\n",
    "        if concentration > high:\n",
    "            pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0030088\", label=\"Increased serum testosterone level\"))\n",
    "        elif concentration < low:\n",
    "            pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0040171\", label=\"Decreased serum testosterone concentration\"))\n",
    "        else:\n",
    "            pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0030087\", label=\"Abnormal circulating testosterone concentration\"), excluded=True)\n",
    "        phenotypic_feature_list.append(pf)\n",
    "    except ValueError:\n",
    "        print(f\"Could not parse \\\"{value}\\\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ACTH_assesssment(row:pd.Series, \n",
    "                     measurement_list: typing.List[Measurement],\n",
    "                     phenotypic_feature_list: typing.List[PhenotypicFeature]) -> None:\n",
    "    \"\"\"\n",
    "    ACTH is measured in picograms per milliliter (pg/mL). Test results are influenced by the time of day the test was done. Normal results are: \n",
    "    Adults: 10-60 pg/ml (1.3-16.7 pmol/L) for an early morning sample (8 a.m.); less than 20 pg/ml (4.5 pmol/L) for a late afternoon sample (4 p.m.)\n",
    "    We will choose the 10-60 range as the test normally should be done in early morning\n",
    "    LOINC: 2141-0, Corticotropin [Mass/volume] in Plasma\n",
    "    \"\"\"\n",
    "    value = row[\"ACTH\"]\n",
    "    if value == \"na\":\n",
    "        return None\n",
    "    if isinstance(value,str) and value.endswith(\" \"):\n",
    "        raise ValueError(f\"Maformed ACTH: {value}\")\n",
    "    lower_limit_of_normal = 10\n",
    "    upper_limit_of_normal = 60\n",
    "    try:\n",
    "        concentration = float(value)\n",
    "        m = Measurements.picogram_per_liter(code=\"LOINC:2141-0\",\n",
    "                                      label=\"Corticotropin (P) [Mass/Vol]\",\n",
    "                                      concentration=concentration,\n",
    "                                      low=lower_limit_of_normal,\n",
    "                                      high=upper_limit_of_normal)\n",
    "        measurement_list.append(m)\n",
    "        if concentration > upper_limit_of_normal:\n",
    "            pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0003154\", label=\"Increased circulating ACTH level\"))\n",
    "        elif concentration < lower_limit_of_normal:\n",
    "            pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0002920\", label=\"Decreased circulating ACTH concentration\"))\n",
    "        else:\n",
    "            pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0011043\", label=\"Abnormal circulating adrenocorticotropin concentration\"), excluded=True)\n",
    "        phenotypic_feature_list.append(pf)\n",
    "    except ValueError:\n",
    "        print(f\"Could not parse \\\"{value}\\\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cortisol_assessment(row:pd.Series, \n",
    "                        measurement_list: typing.List[Measurement],\n",
    "                        phenotypic_feature_list: typing.List[PhenotypicFeature]) -> None:\n",
    "    \"\"\"\n",
    "    Normal values for a blood sample taken at 8 in the morning are 5 to 25 mcg/dL or 140 to 690 nmol/L.\n",
    "    LOINC: 2143-6, Cortisol [Mass/volume] in Serum or Plasma\n",
    "    \"\"\"\n",
    "    value = row[\"cortisol\"]\n",
    "    if value == \"na\":\n",
    "        return None\n",
    "    if isinstance(value,str) and value.endswith(\" \"):\n",
    "        raise ValueError(f\"Maformed cortisol: {value}\")\n",
    "    lower_limit_of_normal = 140\n",
    "    upper_limit_of_normal = 690\n",
    "    try:\n",
    "        concentration = float(value)\n",
    "        m = Measurements.nanomole_per_liter(code=\"LOINC:2143-6\",\n",
    "                                      label=\"Cortisol [Mass/Vol]\",\n",
    "                                      concentration=concentration,\n",
    "                                      low=lower_limit_of_normal,\n",
    "                                      high=upper_limit_of_normal)\n",
    "        measurement_list.append(m)\n",
    "        if concentration > upper_limit_of_normal:\n",
    "            pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0003118\", label=\"Increased circulating cortisol level\"))\n",
    "        elif concentration < lower_limit_of_normal:\n",
    "            pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0008163\", label=\"Decreased circulating cortisol level\"))\n",
    "        else:\n",
    "            pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0011731\", label=\"Abnormality of circulating cortisol level\"), excluded=True)\n",
    "        phenotypic_feature_list.append(pf)\n",
    "    except ValueError:\n",
    "        print(f\"Could not parse \\\"{value}\\\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prolactin_assessment(row:pd.Series, \n",
    "                        measurement_list: typing.List[Measurement],\n",
    "                        phenotypic_feature_list: typing.List[PhenotypicFeature]) -> None:\n",
    "    \"\"\" \n",
    "    Normal prolactin levels typically range from 2 to 17 nanograms per millilitre (ng/mL) in non-pregnant women and 2 to 15 ng/mL in men.\n",
    "    Pregnant women: 80 to 400 ng/mL (80 to 400 µg/L) # we will assume patients tested in non-pregnant state\n",
    "    LOINC: 2842-3 Prolactin [Mass/volume] in Serum or Plasma\n",
    "    \"\"\"\n",
    "    value = row[\"prolactin\"]\n",
    "    if value == \"na\":\n",
    "        return None\n",
    "    sex = row[\"sex\"]\n",
    "    if sex == \"M\":\n",
    "        lower_limit_of_normal = 2\n",
    "        upper_limit_of_normal = 15\n",
    "    elif sex == \"F\":\n",
    "        lower_limit_of_normal = 2\n",
    "        upper_limit_of_normal = 17\n",
    "    else:\n",
    "        print(\"Warning: did not recognize sex\")\n",
    "        return None\n",
    "    if isinstance(value,str) and value.endswith(\" \"):\n",
    "        raise ValueError(f\"Maformed cortisol: {value}\")\n",
    "    try:\n",
    "        concentration = float(value)\n",
    "        m = Measurements.nanogram_per_milliliter(code=\"LOINC:2842-3\",\n",
    "                                      label=\"Prolactin [Mass/Vol]\",\n",
    "                                      concentration=concentration,\n",
    "                                      low=lower_limit_of_normal,\n",
    "                                      high=upper_limit_of_normal)\n",
    "        measurement_list.append(m)\n",
    "        if concentration > upper_limit_of_normal:\n",
    "            pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0000870\", label=\"Increased circulating prolactin concentration\"))\n",
    "        else:\n",
    "            pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0000870\", label=\"Increased circulating prolactin concentration\"), excluded=True)\n",
    "        # Note there is no medically relevant \"Reduced\" term\n",
    "        phenotypic_feature_list.append(pf)\n",
    "    except ValueError:\n",
    "        print(f\"Could not parse \\\"{value}\\\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def estradiol_assessment(row:pd.Series, \n",
    "                         measurement_list: typing.List[Measurement],\n",
    "                         phenotypic_feature_list: typing.List[PhenotypicFeature]) -> None:\n",
    "    \"\"\" \n",
    "    30 to 400 pg/mL for premenopausal women\n",
    "    0 to 30 pg/mL for postmenopausal women\n",
    "    10 to 50 pg/mL for men\n",
    "    LOINC: 2243-4 Estradiol (E2) [Mass/volume] in Serum or Plasma\n",
    "    \"\"\"\n",
    "    value = row[\"estradiol (E2)\"]\n",
    "    if value == \"na\":\n",
    "        return None\n",
    "    sex = row[\"sex\"]\n",
    "    if sex == \"M\":\n",
    "        lower_limit_of_normal = 10\n",
    "        upper_limit_of_normal = 50\n",
    "    elif sex == \"F\":\n",
    "        lower_limit_of_normal = 30\n",
    "        upper_limit_of_normal = 400\n",
    "    else:\n",
    "        print(\"Warning: did not recognize sex\")\n",
    "        return None\n",
    "    \n",
    "    if isinstance(value,str) and value.endswith(\" \"):\n",
    "        raise ValueError(f\"Maformed estradiol: {value}\")\n",
    "    try:\n",
    "        concentration = float(value)\n",
    "        m = Measurements.picogram_per_milliliter(code=\"LOINC:2243-4\",\n",
    "                                      label=\"Estradiol (E2) [Mass/Vol]\",\n",
    "                                      concentration=concentration,\n",
    "                                      low=lower_limit_of_normal,\n",
    "                                      high=upper_limit_of_normal)\n",
    "        measurement_list.append(m)\n",
    "        if concentration > upper_limit_of_normal:\n",
    "            pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0025134\", label=\"Increased serum estradiol\"))\n",
    "        elif concentration < lower_limit_of_normal:\n",
    "            pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0008214\", label=\"Decreased serum estradiol\"))\n",
    "        else:\n",
    "            pf = PhenotypicFeature(type=OntologyClass(id=\"HP:0025133\", label=\"Abnormal serum estradiol\"), excluded=True)\n",
    "        phenotypic_feature_list.append(pf)\n",
    "    except ValueError:\n",
    "        print(f\"Could not parse \\\"{value}\\\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HPO terms\n",
    "The original input table had a list of symptoms that we encode as HPO terms. If a symptom was not mentioned in the original file, we assume it was not available rather than explicitly excluded.\n",
    "The following function is used to extract observed HPO terms for each individual based on the corresponding table row. Note that we additionally generate HPO terms based on the numerical measurements as shown above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_hpo_terms(row:pd):\n",
    "    hpo_term_d = {\n",
    "        \"Secondary amenorrhea\": \"HP:0000869\",\n",
    "        \"Infertility\": \"HP:0000789\",\n",
    "        \"Testicular neoplasm\": \"HP:0010788\",\n",
    "        \"Cryptorchidism\": \"HP:0000028\",\n",
    "        \"Hirsutism\": \"HP:0001007\",\n",
    "        \"Hypertension\": \"HP:0000822\",\n",
    "        \"Clitoral hypertrophy\": \"HP:0008665\",\n",
    "        \"Irregular menstruation\": \"HP:0000858\",\n",
    "        \"Diarrhea\": \"HP:0002014\",\n",
    "        \"Arrhythmia\": \"HP:0011675\",\n",
    "        \"Congenital hypothyroidism\": \"HP:0000851\",\n",
    "        \"Growth delay\": \"HP:0001510\",\n",
    "        \"Acne\": \"HP:0001061\",\n",
    "        \"Moon facies\": \"HP:0500011\",\n",
    "        \"Hyperpigmentation of the skin\": \"HP:0000953\"\n",
    "    }\n",
    "    phenotypic_feature_list = list()\n",
    "    for label, hpo_id in hpo_term_d.items():\n",
    "        status = row[label]\n",
    "        if not status == \"observed\":\n",
    "            continue\n",
    "        term = OntologyClass(id=hpo_id, label=label)\n",
    "        phenotypic_feature_list.append(PhenotypicFeature(type=term))\n",
    "    return phenotypic_feature_list\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CYP21A2 variants\n",
    "The following two cells leverage VariantValidator to obtain representations of HGVS-encoded variants as well as pyphetools classes for structural and promoter variants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "cyp21a2_symbol = \"CYP21A2\"\n",
    "cyp21a2_id = \"HGNC:2600\"\n",
    "cyp21a2_MANE_transcript = \"NM_000500.9\"\n",
    "vmanager = VariantManager(df=df, \n",
    "                          individual_column_name=\"individual\", \n",
    "                          transcript=cyp21a2_MANE_transcript, \n",
    "                          gene_id=cyp21a2_id, \n",
    "                          gene_symbol=cyp21a2_symbol, \n",
    "                          allele_1_column_name=\"allele_1\", \n",
    "                          allele_2_column_name=\"allele_2\")\n",
    "variant_d = vmanager.get_variant_d()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_to_object_d = dict()\n",
    "all_variant_set = set()\n",
    "deletion_set = {\"DEL exon7-10\",\"DEL exon1-7\", \"DEL exon1-3\", \"gene deletion\"}\n",
    "duplication_set = {\"gene duplication\"}\n",
    "firstRow = None\n",
    "for _,row in df.iterrows():\n",
    "    if firstRow is None:\n",
    "        firstRow = row\n",
    "        continue ## first row has additional definitions such as HPO identifiers\n",
    "    allele_1 = row[\"allele_1\"]\n",
    "    allele_2 = row[\"allele_2\"]\n",
    "    all_variant_set.add(allele_1)\n",
    "    all_variant_set.add(allele_2)\n",
    "for allele in all_variant_set:\n",
    "    #print(f\"\\\"{allele}\\\"\")\n",
    "    if allele.startswith(\"PROMOTER\"):\n",
    "        fields = allele.split(\":\")\n",
    "        description = \":\".join(fields[1:])\n",
    "        var_obj = PromoterVariant.two_KB_upstream_variant(description=description,gene_symbol=cyp21a2_symbol, gene_id=cyp21a2_id)\n",
    "        variant_to_object_d[allele] = var_obj.to_variant_interpretation()\n",
    "    elif allele in duplication_set:\n",
    "        dup = StructuralVariant.chromosomal_duplication(cell_contents=allele, gene_id=cyp21a2_id, gene_symbol=cyp21a2_symbol)\n",
    "        variant_to_object_d[allele] = dup.to_variant_interpretation_202()\n",
    "    elif allele in deletion_set:\n",
    "        delet = StructuralVariant.chromosomal_deletion(cell_contents=allele, gene_id=cyp21a2_id, gene_symbol=cyp21a2_symbol)\n",
    "        variant_to_object_d[allele] = delet.to_variant_interpretation_202()\n",
    "    elif allele in variant_d:\n",
    "        var_obj = variant_d.get(allele)\n",
    "        variant_to_object_d[allele] = var_obj.to_variant_interpretation_202()\n",
    "    else:\n",
    "        raise ValueError(f\"Did not recognize allele \\\"{allele}\\\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def row_to_phenopacket(row:pd.Series):\n",
    "    individual_id = row[\"individual\"]\n",
    "    phenopacket_id = \"PMID_30968594_{}\".format(individual_id.replace(\" \", \"_\"))\n",
    "    sex = row[\"sex\"]\n",
    "    age_at_dx = row[\"age_at_dx\"]\n",
    "    age_of_onset = Age(iso8601duration=age_at_dx)\n",
    "    i = Individual(id=individual_id)\n",
    "    if sex == \"M\":\n",
    "        i.sex = Sex.MALE\n",
    "    elif sex == \"F\":\n",
    "        i.sex = Sex.FEMALE\n",
    "    else:\n",
    "        print(f\"Warning count not identify sex for {individual_id}\")\n",
    "    allele_1 = row[\"allele_1\"]\n",
    "    allele_2 = row[\"allele_2\"]\n",
    "    var_list = list()\n",
    "    if allele_1 == allele_2:\n",
    "        var = variant_to_object_d.get(allele_1)\n",
    "        var.variation_descriptor.allelic_state = OntologyTerms.homozygous()\n",
    "        var_list.append(var)\n",
    "    else:\n",
    "        var1 = variant_to_object_d.get(allele_1)\n",
    "        var2 = variant_to_object_d.get(allele_2)\n",
    "        var1.variation_descriptor.allelic_state = OntologyTerms.heterozygous()\n",
    "        var2.variation_descriptor.allelic_state = OntologyTerms.heterozygous()\n",
    "        var_list.append(var1)\n",
    "        var_list.append(var2)\n",
    "    ## create genomic interpretation\n",
    "    interpretation_list = list()\n",
    "    for var in var_list:\n",
    "        genomic_interpretation = GenomicInterpretation(subject_or_biosample_id=individual_id, \n",
    "                                                       interpretation_status=GenomicInterpretation.InterpretationStatus.CAUSATIVE,\n",
    "                                                       call=var)\n",
    "        interpretation_list.append(genomic_interpretation)\n",
    "    ## Disease is always OMIM:201910 for this cohort.\n",
    "    diseaseClass = OntologyClass(id=\"OMIM:201910\", label=\"Adrenal hyperplasia, congenital, due to 21-hydroxylase deficiency\")\n",
    "    ## all individuals have Congenital onset or an iso8601 onset\n",
    "    if age_at_dx == \"Congenital onset\":\n",
    "        onset = OntologyTerms.congenital_onset()\n",
    "    elif age_at_dx.startswith(\"P\"):\n",
    "        onset = Age(iso8601duration=age_at_dx)\n",
    "    else:\n",
    "        raise ValueError(f\"Did not recognize onset \\\"{age_at_dx}\\\"\")\n",
    "    disease = Disease(term=diseaseClass, onset=TimeElement(element=onset))\n",
    "    diagnosis = Diagnosis(disease=diseaseClass, genomic_interpretations=interpretation_list)\n",
    "    interpretation = Interpretation(id=individual_id, progress_status=Interpretation.ProgressStatus.SOLVED, diagnosis=diagnosis)\n",
    "    phenotypic_features = get_hpo_terms(row)\n",
    "    measurements = list()\n",
    "    measurement_17php(row=row, measurement_list=measurements) # we have no applicable normal range, no phenotypic features possible.\n",
    "    testosterone_assessment(row=row, measurement_list=measurements, phenotypic_feature_list=phenotypic_features)\n",
    "    ACTH_assesssment(row=row, measurement_list=measurements, phenotypic_feature_list=phenotypic_features)\n",
    "    cortisol_assessment(row=row, measurement_list=measurements, phenotypic_feature_list=phenotypic_features)\n",
    "    prolactin_assessment(row=row, measurement_list=measurements, phenotypic_feature_list=phenotypic_features)\n",
    "    estradiol_assessment(row=row, measurement_list=measurements, phenotypic_feature_list=phenotypic_features)\n",
    "    ppkt = Phenopacket(id=phenopacket_id, \n",
    "                       subject=i, \n",
    "                       diseases=[disease], \n",
    "                       phenotypic_features=phenotypic_features, \n",
    "                       measurements=measurements, \n",
    "                       interpretations=[interpretation], \n",
    "                       meta_data=metadata)\n",
    "    return ppkt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenopacket_list = list()\n",
    "firstRow = None\n",
    "for _,row in df.iterrows():\n",
    "    if firstRow is None:\n",
    "        firstRow = row\n",
    "        continue ## first row has additional definitions such as HPO identifiers\n",
    "    item = row_to_phenopacket(row)\n",
    "    phenopacket_list.append(item.to_message())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We output 69 GA4GH phenopackets to the directory phenopackets\n"
     ]
    }
   ],
   "source": [
    "outdir = \"phenopackets\"\n",
    "if not os.path.isdir(outdir):\n",
    "    os.makedirs(outdir)\n",
    "written = 0\n",
    "json_list = list()\n",
    "for ppkt in phenopacket_list:    \n",
    "    json_string = MessageToJson(ppkt)\n",
    "    fname = re.sub('[^A-Za-z0-9_-]', '', ppkt.id)  # remove any illegal characters from filename\n",
    "    fname = fname.replace(\" \", \"_\") + \".json\"\n",
    "    outpth = os.path.join(outdir, fname)\n",
    "    with open(outpth, \"wt\") as fh:\n",
    "        fh.write(json_string)\n",
    "        json_list.append(json_string)\n",
    "        written += 1\n",
    "print(f\"We output {written} GA4GH phenopackets to the directory {outdir}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ps24venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
